<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />


<meta name="author" content="Junming Guan" />


<title>Initial exploration of the human brain dataset: committed OPC cells</title>

<script src="site_libs/header-attrs-2.28/header-attrs.js"></script>
<script src="site_libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/readable.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<style>h1 {font-size: 34px;}
       h1.title {font-size: 38px;}
       h2 {font-size: 30px;}
       h3 {font-size: 24px;}
       h4 {font-size: 18px;}
       h5 {font-size: 16px;}
       h6 {font-size: 12px;}
       code {color: inherit; background-color: rgba(0, 0, 0, 0.04);}
       pre:not([class]) { background-color: white }</style>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<link href="site_libs/highlightjs-9.12.0/textmate.css" rel="stylesheet" />
<script src="site_libs/highlightjs-9.12.0/highlight.js"></script>

<link rel="icon" href="https://github.com/workflowr/workflowr-assets/raw/main/img/reproducible.png">
<!-- Add a small amount of space between sections. -->
<style type="text/css">
div.section {
  padding-top: 12px;
}
</style>



<style type="text/css">
  code{white-space: pre-wrap;}
  span.smallcaps{font-variant: small-caps;}
  span.underline{text-decoration: underline;}
  div.column{display: inline-block; vertical-align: top; width: 50%;}
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  ul.task-list{list-style: none;}
    </style>

<style type="text/css">code{white-space: pre;}</style>
<script type="text/javascript">
if (window.hljs) {
  hljs.configure({languages: []});
  hljs.initHighlightingOnLoad();
  if (document.readyState && document.readyState === "complete") {
    window.setTimeout(function() { hljs.initHighlighting(); }, 0);
  }
}
</script>









<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
img {
  max-width:100%;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
details > summary > p:only-child {
  display: inline;
}
pre code {
  padding: 0;
}
</style>


<style type="text/css">
.dropdown-submenu {
  position: relative;
}
.dropdown-submenu>.dropdown-menu {
  top: 0;
  left: 100%;
  margin-top: -6px;
  margin-left: -1px;
  border-radius: 0 6px 6px 6px;
}
.dropdown-submenu:hover>.dropdown-menu {
  display: block;
}
.dropdown-submenu>a:after {
  display: block;
  content: " ";
  float: right;
  width: 0;
  height: 0;
  border-color: transparent;
  border-style: solid;
  border-width: 5px 0 5px 5px;
  border-left-color: #cccccc;
  margin-top: 5px;
  margin-right: -10px;
}
.dropdown-submenu:hover>a:after {
  border-left-color: #adb5bd;
}
.dropdown-submenu.pull-left {
  float: none;
}
.dropdown-submenu.pull-left>.dropdown-menu {
  left: -100%;
  margin-left: 10px;
  border-radius: 6px 0 6px 6px;
}
</style>

<script type="text/javascript">
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark the anchor link active (and if it's in a dropdown, also mark that active)
  var dropdown = menuAnchor.closest('li.dropdown');
  if (window.bootstrap) { // Bootstrap 4+
    menuAnchor.addClass('active');
    dropdown.find('> .dropdown-toggle').addClass('active');
  } else { // Bootstrap 3
    menuAnchor.parent().addClass('active');
    dropdown.addClass('active');
  }

  // Navbar adjustments
  var navHeight = $(".navbar").first().height() + 15;
  var style = document.createElement('style');
  var pt = "padding-top: " + navHeight + "px; ";
  var mt = "margin-top: -" + navHeight + "px; ";
  var css = "";
  // offset scroll position for anchor links (for fixed navbar)
  for (var i = 1; i <= 6; i++) {
    css += ".section h" + i + "{ " + pt + mt + "}\n";
  }
  style.innerHTML = "body {" + pt + "padding-bottom: 40px; }\n" + css;
  document.head.appendChild(style);
});
</script>

<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before, .tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "\e259";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "\e258";
  font-family: 'Glyphicons Halflings';
  border: none;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
  background-color: transparent;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<!-- code folding -->




</head>

<body>


<div class="container-fluid main-container">




<div class="navbar navbar-default  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-bs-toggle="collapse" data-target="#navbar" data-bs-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html">single-cell jamboree</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="index.html">Overview</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/stephenslab/single-cell-jamboree">source</a>
</li>
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<div id="header">



<h1 class="title toc-ignore">Initial exploration of the human brain
dataset: committed OPC cells</h1>
<h4 class="author">Junming Guan</h4>

</div>


<p>
<button type="button" class="btn btn-default btn-workflowr btn-workflowr-report" data-toggle="collapse" data-target="#workflowr-report">
<span class="glyphicon glyphicon-list" aria-hidden="true"></span>
workflowr <span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span>
</button>
</p>
<div id="workflowr-report" class="collapse">
<ul class="nav nav-tabs">
<li class="active">
<a data-toggle="tab" href="#summary">Summary</a>
</li>
<li>
<a data-toggle="tab" href="#checks"> Checks <span
class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span>
</a>
</li>
<li>
<a data-toggle="tab" href="#versions">Past versions</a>
</li>
</ul>
<div class="tab-content">
<div id="summary" class="tab-pane fade in active">
<p>
<strong>Last updated:</strong> 2024-11-09
</p>
<p>
<strong>Checks:</strong> <span
class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> 7
<span class="glyphicon glyphicon-exclamation-sign text-danger"
aria-hidden="true"></span> 0
</p>
<p>
<strong>Knit directory:</strong>
<code>single-cell-jamboree/analysis/</code> <span
class="glyphicon glyphicon-question-sign" aria-hidden="true"
title="This is the local directory in which the code in this file was executed.">
</span>
</p>
<p>
This reproducible <a href="https://rmarkdown.rstudio.com">R Markdown</a>
analysis was created with <a
  href="https://github.com/workflowr/workflowr">workflowr</a> (version
1.7.1). The <em>Checks</em> tab describes the reproducibility checks
that were applied when the results were created. The <em>Past
versions</em> tab lists the development history.
</p>
<hr>
</div>
<div id="checks" class="tab-pane fade">
<div id="workflowr-checks" class="panel-group">
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongRMarkdownfilestronguptodate">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>R Markdown file:</strong> up-to-date
</a>
</p>
</div>
<div id="strongRMarkdownfilestronguptodate"
class="panel-collapse collapse">
<div class="panel-body">
<p>Great! Since the R Markdown file has been committed to the Git
repository, you know the exact version of the code that produced these
results.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongEnvironmentstrongempty">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Environment:</strong> empty </a>
</p>
</div>
<div id="strongEnvironmentstrongempty" class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! The global environment was empty. Objects defined in the
global environment can affect the analysis in your R Markdown file in
unknown ways. For reproduciblity it’s best to always run the code in an
empty environment.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongSeedstrongcodesetseed1code">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Seed:</strong>
<code>set.seed(1)</code> </a>
</p>
</div>
<div id="strongSeedstrongcodesetseed1code"
class="panel-collapse collapse">
<div class="panel-body">
<p>The command <code>set.seed(1)</code> was run prior to running the
code in the R Markdown file. Setting a seed ensures that any results
that rely on randomness, e.g. subsampling or permutations, are
reproducible.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongSessioninformationstrongrecorded">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Session information:</strong>
recorded </a>
</p>
</div>
<div id="strongSessioninformationstrongrecorded"
class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! Recording the operating system, R version, and package
versions is critical for reproducibility.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongCachestrongnone">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Cache:</strong> none </a>
</p>
</div>
<div id="strongCachestrongnone" class="panel-collapse collapse">
<div class="panel-body">
<p>Nice! There were no cached chunks for this analysis, so you can be
confident that you successfully produced the results during this
run.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongFilepathsstrongrelative">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>File paths:</strong> relative </a>
</p>
</div>
<div id="strongFilepathsstrongrelative" class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! Using relative paths to the files within your workflowr
project makes it easier to run your code on other machines.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongRepositoryversionstrongahrefhttpsgithubcomstephenslabsinglecelljamboreetreee8fc53fcf42b23566f0a564256ae34cb44060a95targetblanke8fc53fa">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Repository version:</strong>
<a href="https://github.com/stephenslab/single-cell-jamboree/tree/e8fc53fcf42b23566f0a564256ae34cb44060a95" target="_blank">e8fc53f</a>
</a>
</p>
</div>
<div
id="strongRepositoryversionstrongahrefhttpsgithubcomstephenslabsinglecelljamboreetreee8fc53fcf42b23566f0a564256ae34cb44060a95targetblanke8fc53fa"
class="panel-collapse collapse">
<div class="panel-body">
<p>
Great! You are using Git for version control. Tracking code development
and connecting the code version to the results is critical for
reproducibility.
</p>
<p>
The results in this page were generated with repository version
<a href="https://github.com/stephenslab/single-cell-jamboree/tree/e8fc53fcf42b23566f0a564256ae34cb44060a95" target="_blank">e8fc53f</a>.
See the <em>Past versions</em> tab to see a history of the changes made
to the R Markdown and HTML files.
</p>
<p>
Note that you need to be careful to ensure that all relevant files for
the analysis have been committed to Git prior to generating the results
(you can use <code>wflow_publish</code> or
<code>wflow_git_commit</code>). workflowr only checks the R Markdown
file, but you know if there are other scripts or data files that it
depends on. Below is the status of the Git repository when the results
were generated:
</p>
<pre><code>
Untracked files:
    Untracked:  .DS_Store
    Untracked:  analysis/.DS_Store
    Untracked:  analysis/.Rhistory
    Untracked:  analysis/human_brain_COP_cells_fit.RData
    Untracked:  code/read_gene_set_data.R
    Untracked:  data/.DS_Store
    Untracked:  data/.Rhistory
    Untracked:  data/Homo_sapiens.gene_info.gz
    Untracked:  data/human_brain_COP_cells.rds
    Untracked:  data/human_brain_COP_cells_fit.RData
    Untracked:  data/human_brain_COP_cells_fit_protein_coding.RData

Unstaged changes:
    Deleted:    data/human_brain_OPC_cells.rds

</code></pre>
<p>
Note that any generated files, e.g. HTML, png, CSS, etc., are not
included in this status report because it is ok for generated content to
have uncommitted changes.
</p>
</div>
</div>
</div>
</div>
<hr>
</div>
<div id="versions" class="tab-pane fade">
<p>
These are the previous versions of the repository in which changes were
made to the R Markdown (<code>analysis/human_brain_COP_cells.Rmd</code>)
and HTML (<code>docs/human_brain_COP_cells.html</code>) files. If you’ve
configured a remote Git repository (see <code>?wflow_git_remote</code>),
click on the hyperlinks in the table below to view the files as they
were in that past version.
</p>
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
File
</th>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
<th>
Message
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/stephenslab/single-cell-jamboree/blob/e8fc53fcf42b23566f0a564256ae34cb44060a95/analysis/human_brain_COP_cells.Rmd" target="_blank">e8fc53f</a>
</td>
<td>
junmingguan
</td>
<td>
2024-11-09
</td>
<td>
workflowr::wflow_publish("human_brain_COP_cells.Rmd")
</td>
</tr>
</tbody>
</table>
</div>
<hr>
</div>
</div>
</div>
<p>TODO: UMAP by factors</p>
<div id="introduction" class="section level3">
<h3>Introduction</h3>
<p>This single-nucleus RNA seq dataset is from the paper “Transcriptomic
diversity of cell types across the adult human brain” (Siletti, 2023).
around 3 millions nuclei were collected from around 100 dissections from
the following areas of brains of 3 donors:</p>
<ul>
<li><p>Forebrain:</p>
<ul>
<li><p>cerebral cortex</p></li>
<li><p>cerebral nuclei</p></li>
<li><p>hypothalamus</p></li>
<li><p>hippocampus formation</p></li>
<li><p>thalamus complex</p></li>
</ul></li>
<li><p>Midbrain</p></li>
<li><p>Hindbrain:</p>
<ul>
<li><p>pons</p></li>
<li><p>cerebellum</p></li>
<li><p>Myelencephalon (Medulla)</p></li>
</ul></li>
<li><p>Spinal cord</p></li>
</ul>
<p>The authors performed hierarchical graph-based clustering, grouping
the cells into superclusters, clusters, and subclusters. The data can be
accessed <a
href="https://data.humancellatlas.org/hca-bio-networks/nervous-system/atlases/brain-v1-0">here</a>,
with files organized by supercluster or by dissection.</p>
<p>This exploratory analysis focuses on committed oligodendrocyte
precursor (COP) cells (see <a
href="https://cellxgene.cziscience.com/e/f16f4108-7873-4035-9989-3748da1a7ff1.cxg/"
class="uri">https://cellxgene.cziscience.com/e/f16f4108-7873-4035-9989-3748da1a7ff1.cxg/</a>).
There are 4,720 cells and 2,1462 genes (after QC) in the dataset.</p>
<pre class="r"><code>library(Matrix)
# library(MatrixExtra)
library(flashier)
library(fastTopics)
library(ggplot2)
library(cowplot)
library(dplyr)
library(readr)</code></pre>
<p>Load data and fit</p>
<pre class="r"><code># Taken from https://github.com/stephenslab/pathways/blob/master/inst/code/read_gene_set_data.R
read_gene_info &lt;- function (file) {
  
  # Read the data into a data frame.
  out &lt;- suppressMessages(read_delim(file,delim = &quot;\t&quot;,col_names = TRUE))
  class(out) &lt;- &quot;data.frame&quot;
  dbXrefs    &lt;- out$dbXrefs
  out        &lt;- out[c(&quot;GeneID&quot;,&quot;Symbol&quot;,&quot;Synonyms&quot;,&quot;chromosome&quot;)]
  
  # Set any entries with a single hyphen to NA, and convert the
  # &quot;chromosome&quot; column to a factor.
  out$chromosome[out$chromosome == &quot;-&quot;] &lt;- NA
  out$Synonyms[out$Synonyms == &quot;-&quot;]     &lt;- NA
  dbXrefs[dbXrefs == &quot;-&quot;]               &lt;- NA
  out &lt;- transform(out,chromosome = factor(chromosome))
  
  # Extract the Ensembl ids. Note that a small number of genes map to
  # more than one Ensembl id; in those cases, we retain the first
  # Ensembl id only.
  dbXrefs &lt;- strsplit(dbXrefs,&quot;|&quot;,fixed = TRUE)
  out$Ensembl &lt;- sapply(dbXrefs,function (x) {
    i &lt;- which(substr(x,1,8) == &quot;Ensembl:&quot;)
    if (length(i) &gt; 0)
      return(substr(x[i[1]],9,nchar(x[i[1]])))
    else
      return(as.character(NA))
  })
  
  # For human genes, extract the HGNC (HUGO Gene Nomenclature
  # Committee) ids.
  out$HGNC &lt;- sapply(dbXrefs,function (x) {
    i &lt;- which(substr(x,1,10) == &quot;HGNC:HGNC:&quot;)
    if (length(i) &gt; 0)
      return(substr(x[i[1]],6,nchar(x[i[1]])))
    else
      return(as.character(NA))
  })
  
  # Return the processed gene data.
  return(out)
}

homo_sapien_geno_info &lt;- read_gene_info(&#39;../data/Homo_sapiens.gene_info.gz&#39;)
data &lt;- readRDS(&#39;../data/human_brain_COP_cells.rds&#39;)
counts &lt;- t(data$RNA$data)
# Warning: package &#39;SeuratObject&#39; was built under R version 4.3.3

# Keep only protein-coding genes and remove those without nonzero counts
reduced_counts &lt;- 
  counts[, colnames(counts) %in% homo_sapien_geno_info$Ensembl]
cols_to_keep &lt;- colSums(reduced_counts != 0, na.rm = TRUE) &gt; 0
reduced_counts &lt;- reduced_counts[, cols_to_keep]
load(&#39;../data/human_brain_COP_cells_fit_protein_coding.RData&#39;)</code></pre>
<pre class="r"><code>map_tissue &lt;- function(tissue) {
  if (tissue %in% c(&quot;cerebral cortex&quot;, &quot;cerebral nuclei&quot;, &quot;hypothalamus&quot;, 
                    &quot;hippocampal formation&quot;, &quot;thalamic complex&quot;)) {
    return(&quot;forebrain&quot;)
  } else if (tissue == &quot;midbrain&quot;) {
    return(&quot;midbrain&quot;)
  } else if (tissue %in% c(&quot;pons&quot;, &quot;cerebellum&quot;, &quot;myelencephalon&quot;)) {
    return(&quot;hindbrain&quot;)
  } else if (tissue == &quot;spinal cord&quot;) {
    return(&quot;spinal cord&quot;)
  } else {
    return(NA)
  }
}

regions &lt;- sapply(data$tissue, map_tissue)
</code></pre>
</div>
<div id="t-sne-and-umap" class="section level3">
<h3>t-SNE and UMAP</h3>
<p>The dataset includes precomputed tSNE and UMAP embeddings, allowing
us to plot them directly. We can color the cells by tissue, by region,
or by cluster.</p>
<div id="t-sne" class="section level4">
<h4>t-SNE</h4>
<pre class="r"><code># colors &lt;- brewer.pal(length(unique(data$tissue)), &quot;Paired&quot;)
colors &lt;-  c(&#39;#756bb1&#39;, &#39;#1c9099&#39;, &#39;#d95f0e&#39;, &#39;#edf8b1&#39;, &#39;#dd1c77&#39;, &#39;#636363&#39;, &#39;#a1d99b&#39;, &#39;#fa9fb5&#39;, &#39;#fec44f&#39;, &#39;#de2d26&#39;)
ggplot(Embeddings(data$tSNE) , aes(x = TSNE_1, y = TSNE_2, color = data$tissue)) +
  geom_point(alpha = 0.7) +
  labs(title = &quot;t-SNE Plot Colored by Tissue Type&quot;, x = &quot;t-SNE 1&quot;, y = &quot;t-SNE 2&quot;) +
  theme_minimal() +
  scale_color_manual(values = colors)</code></pre>
<p><img src="figure/human_brain_COP_cells.Rmd/tsne-1.png" width="840" style="display: block; margin: auto;" /></p>
<pre class="r"><code>
# colors &lt;- brewer.pal(length(unique(regions)), &quot;Set1&quot;)
# ggplot(Embeddings(data$tSNE) , aes(x = TSNE_1, y = TSNE_2, color = regions)) +
#   geom_point(alpha = 0.7) +
#   labs(title = &quot;t-SNE Plot Colored by Tissue Type&quot;, x = &quot;t-SNE 1&quot;, y = &quot;t-SNE 2&quot;) +
#   theme_minimal() +
#   scale_color_manual(values = colors)

# colors &lt;- brewer.pal(length(unique(data$cluster_id)), &quot;Paired&quot;)
colors &lt;-  c(&#39;#756bb1&#39;, &#39;#1c9099&#39;, &#39;#d95f0e&#39;, &#39;#edf8b1&#39;, &#39;#dd1c77&#39;, &#39;#636363&#39;, &#39;#a1d99b&#39;)
ggplot(Embeddings(data$tSNE) , aes(x = TSNE_1, y = TSNE_2, color = data$cluster_id)) +
  geom_point(alpha = 0.7) +
  labs(title = &quot;t-SNE Plot Colored by Cluster&quot;, x = &quot;t-SNE 1&quot;, y = &quot;t-SNE 2&quot;) +
  theme_minimal() +
  scale_color_manual(values = colors)</code></pre>
<p><img src="figure/human_brain_COP_cells.Rmd/tsne-2.png" width="840" style="display: block; margin: auto;" />
#### UMAP</p>
<pre class="r"><code># colors &lt;- brewer.pal(length(unique(data$tissue)), &quot;Paired&quot;)
colors &lt;-  c(&#39;#756bb1&#39;, &#39;#1c9099&#39;, &#39;#d95f0e&#39;, &#39;#edf8b1&#39;, &#39;#dd1c77&#39;, &#39;#636363&#39;, &#39;#a1d99b&#39;, &#39;#fa9fb5&#39;, &#39;#fec44f&#39;, &#39;#de2d26&#39;)
ggplot(Embeddings(data$UMAP) , aes(x = UMAP_1, y = UMAP_2, color = data$tissue)) +
  geom_point(alpha = 0.7) +
  labs(title = &quot;UMAP Plot Colored by Tissue Type&quot;, x = &quot;UMAP 1&quot;, y = &quot;UMAP 2&quot;) +
  theme_minimal() +
  scale_color_manual(values = colors)</code></pre>
<p><img src="figure/human_brain_COP_cells.Rmd/umap-1.png" width="840" style="display: block; margin: auto;" /></p>
<pre class="r"><code>
# ggplot(Embeddings(data$UMAP) , aes(x = UMAP_1, y = UMAP_2, color = regions)) +
#   geom_point(alpha = 0.7) +
#   labs(title = &quot;UMAP Plot Colored by Tissue Type&quot;, x = &quot;UMAP 1&quot;, y = &quot;UMAP 2&quot;) +
#   theme_minimal() +
#   scale_color_manual(values = rainbow(length(unique(regions))))

# colors &lt;- brewer.pal(length(unique(data$cluster_id)), &quot;Paired&quot;)
colors &lt;-  c(&#39;#756bb1&#39;, &#39;#1c9099&#39;, &#39;#d95f0e&#39;, &#39;#edf8b1&#39;, &#39;#dd1c77&#39;, &#39;#636363&#39;, &#39;#a1d99b&#39;)
ggplot(Embeddings(data$UMAP) , aes(x = UMAP_1, y = UMAP_2, color = data$cluster_id)) +
  geom_point(alpha = 0.7) +
  labs(title = &quot;UMAP Plot Colored by Cluster&quot;, x = &quot;UMAP 1&quot;, y = &quot;UMAP 2&quot;) +
  theme_minimal() +
  scale_color_manual(values = colors)</code></pre>
<p><img src="figure/human_brain_COP_cells.Rmd/umap-2.png" width="840" style="display: block; margin: auto;" /></p>
</div>
</div>
<div id="flashier" class="section level3">
<h3>Flashier</h3>
<div id="nmf" class="section level4">
<h4>NMF</h4>
<p>Redefine subcluster ids so that the hierarchical structure is
clearer</p>
<pre class="r"><code># M &lt;- table(data$subcluster_id, data$cluster_id)
# M
subcluster_id &lt;- as.vector(data$subcluster_id)
subcluster_id[subcluster_id == &#39;3033&#39;] &lt;- &#39;37.1&#39;
subcluster_id[subcluster_id == &#39;3035&#39;] &lt;- &#39;37.2&#39;
subcluster_id[subcluster_id == &#39;3036&#39;] &lt;- &#39;37.3&#39;
subcluster_id[subcluster_id == &#39;3037&#39;] &lt;- &#39;37.4&#39;
subcluster_id[subcluster_id == &#39;3038&#39;] &lt;- &#39;37.5&#39;

subcluster_id[subcluster_id == &#39;3030&#39;] &lt;- &#39;38.1&#39;
subcluster_id[subcluster_id == &#39;3031&#39;] &lt;- &#39;38.2&#39;
subcluster_id[subcluster_id == &#39;3032&#39;] &lt;- &#39;38.3&#39;
subcluster_id[subcluster_id == &#39;3034&#39;] &lt;- &#39;38.4&#39;

subcluster_id[subcluster_id == &#39;3027&#39;] &lt;- &#39;39.1&#39;
subcluster_id[subcluster_id == &#39;3028&#39;] &lt;- &#39;39.2&#39;
subcluster_id[subcluster_id == &#39;3029&#39;] &lt;- &#39;39.3&#39;

subcluster_id[subcluster_id == &#39;3014&#39;] &lt;- &#39;41.1&#39;
subcluster_id[subcluster_id == &#39;3015&#39;] &lt;- &#39;42.2&#39;

subcluster_id[subcluster_id == &#39;3007&#39;] &lt;- &#39;42.1&#39;
subcluster_id[subcluster_id == &#39;3008&#39;] &lt;- &#39;42.2&#39;
subcluster_id[subcluster_id == &#39;3009&#39;] &lt;- &#39;42.3&#39;
subcluster_id[subcluster_id == &#39;3010&#39;] &lt;- &#39;42.4&#39;

subcluster_id[subcluster_id == &#39;3006&#39;] &lt;- &#39;43.1&#39;

subcluster_id[subcluster_id == &#39;3195&#39;] &lt;- &#39;75.1&#39;
subcluster_id[subcluster_id == &#39;3196&#39;] &lt;- &#39;75.2&#39;
subcluster_id &lt;- factor(subcluster_id)</code></pre>
<pre class="r"><code>flash_plot_scree(flashier_fit)</code></pre>
<p><img src="figure/human_brain_COP_cells.Rmd/flashier-plot-1.png" width="840" style="display: block; margin: auto;" /></p>
<pre class="r"><code>

plot(flashier_fit, 
     plot_type = &quot;heatmap&quot;,
     pm_which = &quot;loadings&quot;, 
     pm_groups = data$cluster_id,
     gap = 25)</code></pre>
<p><img src="figure/human_brain_COP_cells.Rmd/flashier-plot-2.png" width="840" style="display: block; margin: auto;" /></p>
<pre class="r"><code>
plot(flashier_fit, 
     plot_type = &quot;histogram&quot;,
     pm_which = &quot;loadings&quot;, 
     pm_groups = data$cluster_id,
     bins = 25)</code></pre>
<p><img src="figure/human_brain_COP_cells.Rmd/flashier-plot-3.png" width="840" style="display: block; margin: auto;" /></p>
<pre class="r"><code>plot(flashier_fit, 
     plot_type = &quot;structure&quot;,
     pm_which = &quot;loadings&quot;, 
     pm_groups = data$cluster_id,
     bins = 30, gap = 70)</code></pre>
<p><img src="figure/human_brain_COP_cells.Rmd/flashier-struc-plot-1.png" width="840" style="display: block; margin: auto;" /></p>
<pre class="r"><code>
plot(flashier_fit, 
     plot_type = &quot;structure&quot;,
     pm_which = &quot;loadings&quot;, 
     pm_groups = subcluster_id,
     bins = 60, gap = 100)</code></pre>
<p><img src="figure/human_brain_COP_cells.Rmd/flashier-struc-plot-2.png" width="840" style="display: block; margin: auto;" /></p>
<pre class="r"><code>vals &lt;- ldf(flashier_fit,type=&quot;m&quot;)
ncells &lt;- colSums(vals$L&gt;0.1)
ncells
#  [1] 4720 3101 1557   79 2060 1641 1281    1    5    1    4    2 1795</code></pre>
<p>Several factors are load on only a small number of cells.</p>
<p>Removing those factors from the structure plots:</p>
<pre class="r"><code>plot(flashier_fit, 
     kset = which(ncells &gt; 5),
     plot_type = &quot;structure&quot;,
     pm_which = &quot;loadings&quot;, 
     pm_groups = data$cluster_id,
     bins = 30, gap = 70)</code></pre>
<p><img src="figure/human_brain_COP_cells.Rmd/flashier-struc-plot-reduced-1.png" width="840" style="display: block; margin: auto;" /></p>
<pre class="r"><code>
plot(flashier_fit, 
     kset = which(ncells &gt; 5),
     plot_type = &quot;structure&quot;,
     pm_which = &quot;loadings&quot;, 
     pm_groups = subcluster_id,
     bins = 60, gap = 100)</code></pre>
<p><img src="figure/human_brain_COP_cells.Rmd/flashier-struc-plot-reduced-2.png" width="840" style="display: block; margin: auto;" /></p>
<p>Showing those factors in the structure plots:</p>
<pre class="r"><code>plot(flashier_fit, 
     kset = which(ncells &lt;= 5),
     plot_type = &quot;structure&quot;,
     pm_which = &quot;loadings&quot;, 
     pm_groups = data$cluster_id,
     bins = 30, gap = 70)</code></pre>
<p><img src="figure/human_brain_COP_cells.Rmd/flashier-struc-plot-reduced-outlier-1.png" width="840" style="display: block; margin: auto;" /></p>
<pre class="r"><code>
plot(flashier_fit, 
     kset = which(ncells &lt;= 5),
     plot_type = &quot;structure&quot;,
     pm_which = &quot;loadings&quot;, 
     pm_groups = subcluster_id,
     bins = 60, gap = 100)</code></pre>
<p><img src="figure/human_brain_COP_cells.Rmd/flashier-struc-plot-reduced-outlier-2.png" width="840" style="display: block; margin: auto;" /></p>
<div id="umap-by-factor" class="section level5">
<h5>UMAP by factor</h5>
<pre class="r"><code># library(gridExtra)
# plot_list &lt;- list()
# i &lt;- 1
# for (f in 1:flashier_fit$n_factors) {
# plot_list[[i]] &lt;- 
  # ggplot(Embeddings(data$UMAP) , aes(x = UMAP_1, y = UMAP_2, size = vals$L[,f])) +
  # geom_point(alpha = 0.7) +
  # labs(title = paste(&quot;factor &quot;, f) , x = &quot;UMAP 1&quot;, y = &quot;UMAP 2&quot;) +
  # theme_minimal() +
  # scale_color_manual(values = &#39;#f0f0f0&#39;)
# i &lt;- i + 1
# }
# grid.arrange(grobs=plot_list)</code></pre>
</div>
</div>
<div id="semi-nmf" class="section level4">
<h4>Semi-NMF</h4>
<pre class="r"><code>flash_plot_scree(flashier_fit_semi)</code></pre>
<p><img src="figure/human_brain_COP_cells.Rmd/flashier-plot-cluster-1.png" width="840" style="display: block; margin: auto;" /></p>
<pre class="r"><code>

plot(flashier_fit_semi, 
     plot_type = &quot;heatmap&quot;,
     pm_which = &quot;loadings&quot;, 
     pm_groups = data$cluster_id,
     gap = 25)</code></pre>
<p><img src="figure/human_brain_COP_cells.Rmd/flashier-plot-cluster-2.png" width="840" style="display: block; margin: auto;" /></p>
<pre class="r"><code>
plot(flashier_fit_semi, 
     plot_type = &quot;histogram&quot;,
     pm_which = &quot;loadings&quot;, 
     pm_groups = data$cluster_id,
     bins = 25)</code></pre>
<p><img src="figure/human_brain_COP_cells.Rmd/flashier-plot-cluster-3.png" width="840" style="display: block; margin: auto;" /></p>
<pre class="r"><code>plot(flashier_fit_semi, 
     plot_type = &quot;structure&quot;,
     pm_which = &quot;loadings&quot;, 
     pm_groups = data$cluster_id,
     bins = 30, gap = 70)</code></pre>
<p><img src="figure/human_brain_COP_cells.Rmd/flashier-seminmf-struc-plot-1.png" width="840" style="display: block; margin: auto;" /></p>
<pre class="r"><code>
plot(flashier_fit_semi, 
     plot_type = &quot;structure&quot;,
     pm_which = &quot;loadings&quot;, 
     pm_groups = subcluster_id,
     bins = 60, gap = 100)</code></pre>
<p><img src="figure/human_brain_COP_cells.Rmd/flashier-seminmf-struc-plot-2.png" width="840" style="display: block; margin: auto;" /></p>
<pre class="r"><code>vals_semi &lt;- ldf(flashier_fit_semi, type=&quot;m&quot;)
ncells_semi &lt;- colSums(vals_semi$L&gt;0.1)
ncells_semi
#  [1] 4720 3121 1700    3  559 1208  769    1  324 2350    1 1469 1799  177 1591
# [16]    1    1 1838 1886    3   68    1    2    1    3    1    2    5    2    1
# [31]    1    2    1    1   54    1    5    1    1    1    2    2    1    1   59
# [46]    2    2    3    1    3</code></pre>
<p>Several factors are load on only a small number of cells.</p>
<p>Removing those factors from the structure plots:</p>
<pre class="r"><code>plot(flashier_fit_semi, 
     kset = which(ncells_semi &gt; 5),
     plot_type = &quot;structure&quot;,
     pm_which = &quot;loadings&quot;, 
     pm_groups = data$cluster_id,
     bins = 30, gap = 70)</code></pre>
<p><img src="figure/human_brain_COP_cells.Rmd/flashier-struc-plot-reduced-semi-1.png" width="840" style="display: block; margin: auto;" /></p>
<pre class="r"><code>
plot(flashier_fit_semi, 
     kset = which(ncells_semi &gt; 5),
     plot_type = &quot;structure&quot;,
     pm_which = &quot;loadings&quot;, 
     pm_groups = subcluster_id,
     bins = 60, gap = 100)</code></pre>
<p><img src="figure/human_brain_COP_cells.Rmd/flashier-struc-plot-reduced-semi-2.png" width="840" style="display: block; margin: auto;" /></p>
<p>Showing factors that are only loaded on a small number of cells:</p>
<pre class="r"><code>plot(flashier_fit_semi, 
     kset = which(ncells_semi &lt;= 5),
     plot_type = &quot;structure&quot;,
     pm_which = &quot;loadings&quot;, 
     pm_groups = data$cluster_id,
     bins = 30, gap = 70)</code></pre>
<p><img src="figure/human_brain_COP_cells.Rmd/flashier-struc-plot-reduced-outlier-semi-1.png" width="840" style="display: block; margin: auto;" /></p>
<pre class="r"><code>
plot(flashier_fit_semi, 
     kset = which(ncells_semi &lt;= 5),
     plot_type = &quot;structure&quot;,
     pm_which = &quot;loadings&quot;, 
     pm_groups = subcluster_id,
     bins = 60, gap = 100)</code></pre>
<p><img src="figure/human_brain_COP_cells.Rmd/flashier-struc-plot-reduced-outlier-semi-2.png" width="840" style="display: block; margin: auto;" /></p>
</div>
<div id="comparison" class="section level4">
<h4>Comparison</h4>
<pre class="r"><code>
plot(flashier_fit, 
     kset = which(ncells &gt; 5),
     plot_type = &quot;structure&quot;,
     pm_which = &quot;loadings&quot;, 
     pm_groups = data$cluster_id,
     bins = 30, gap = 70)</code></pre>
<p><img src="figure/human_brain_COP_cells.Rmd/struc-plot-compare-1.png" width="840" style="display: block; margin: auto;" /></p>
<pre class="r"><code>
plot(flashier_fit_semi, 
     kset = which(ncells_semi &gt; 5),
     plot_type = &quot;structure&quot;,
     pm_which = &quot;loadings&quot;, 
     pm_groups = data$cluster_id,
     bins = 30, gap = 70)</code></pre>
<p><img src="figure/human_brain_COP_cells.Rmd/struc-plot-compare-2.png" width="840" style="display: block; margin: auto;" /></p>
<pre class="r"><code>
# plot_grid(p1, p2, ncol = 1, align = &quot;v&quot;, axis = &quot;lr&quot;, rel_heights = c(1, 1))</code></pre>
</div>
</div>
<div id="fasttopics" class="section level3">
<h3>FastTopics</h3>
<!-- ```{r fit-fasttopics} -->
<!-- # reduced_counts <- as(reduced_counts, "dgCMatrix") -->
<!-- fasttopics_fit <- fit_topic_model(reduced_counts, k = 50) -->
<!-- ``` -->
<pre class="r"><code>plot_progress(fasttopics_fit,x = &quot;iter&quot;,add.point.every = 10,colors = &quot;black&quot;) +
  theme_cowplot(font_size = 10)</code></pre>
<p><img src="figure/human_brain_COP_cells.Rmd/fasttopics-plot-progress-1.png" width="840" style="display: block; margin: auto;" /></p>
<pre class="r"><code>loglik &lt;- loglik_multinom_topic_model(reduced_counts, fasttopics_fit)
pdat &lt;- data.frame(loglik)
ggplot(pdat,aes(loglik)) +
  geom_histogram(bins = 64,color = &quot;white&quot;,fill = &quot;black&quot;,size = 0.25) +
  labs(y = &quot;number of cells&quot;) +
  theme_cowplot(font_size = 10)
# Warning: Using `size` aesthetic for lines was deprecated in ggplot2 3.4.0.
# ℹ Please use `linewidth` instead.
# This warning is displayed once every 8 hours.
# Call `lifecycle::last_lifecycle_warnings()` to see where this warning was
# generated.</code></pre>
<p><img src="figure/human_brain_COP_cells.Rmd/fasttopics-loglik-plot-1.png" width="840" style="display: block; margin: auto;" /></p>
<pre class="r"><code>
# subpop_colors &lt;- c(&quot;dodgerblue&quot;,&quot;forestgreen&quot;,&quot;darkmagenta&quot;,&quot;skyblue&quot;,&quot;gold&quot;, )
subpop_colors &lt;- c(&quot;red&quot;, &quot;blue&quot;, &quot;green&quot;, &quot;purple&quot;, &quot;orange&quot;, &quot;pink&quot;, &quot;cyan&quot;, &quot;brown&quot;, &quot;yellow&quot;, &quot;darkgreen&quot;)

pdat &lt;- data.frame(loglik = loglik,subpop = data$tissue)
ggplot(pdat,aes(x = loglik,fill = subpop)) +
  geom_histogram(bins = 64,color = &quot;white&quot;,size = 0.25) +
  scale_fill_manual(values = subpop_colors) +
  labs(y = &quot;number of cells&quot;) +
  theme_cowplot(font_size = 10)</code></pre>
<p><img src="figure/human_brain_COP_cells.Rmd/fasttopics-loglik-plot-2.png" width="840" style="display: block; margin: auto;" /></p>
<pre class="r"><code>structure_plot(fasttopics_fit, grouping = data$cluster_id, gap = 70)</code></pre>
<p><img src="figure/human_brain_COP_cells.Rmd/fasttopics-struc-plot-1.png" width="840" style="display: block; margin: auto;" /></p>
<pre class="r"><code>structure_plot(fasttopics_fit, grouping = subcluster_id, gap = 70)</code></pre>
<p><img src="figure/human_brain_COP_cells.Rmd/fasttopics-struc-plot-2.png" width="840" style="display: block; margin: auto;" /></p>
<pre class="r"><code>ncells_ft &lt;- colSums(fasttopics_fit$L&gt;0.1)
ncells_ft
#   k1   k2   k3   k4   k5   k6   k7   k8   k9  k10  k11  k12  k13  k14  k15  k16 
#  127   97  876  156 1019  115  405   12  171  112 1734  315  141   20  112   35 
#  k17  k18  k19  k20  k21  k22  k23  k24  k25  k26  k27  k28  k29  k30  k31  k32 
#   46   65  143  484  322  281 1008  127   66  347  107  128   38  612  517  202 
#  k33  k34  k35  k36  k37  k38  k39  k40  k41  k42  k43  k44  k45  k46  k47  k48 
#  196  148  877   24   70   28  204  370  259  189   58   44  120  402   72  433 
#  k49  k50 
#   22  313</code></pre>
<p>No factors that are loaded on a single cells. Still there are a
couple that are loaded on a small number of cells, e.g., k8.</p>
</div>
<div id="gbcd" class="section level3">
<h3>GBCD</h3>
<p>TODO</p>
<pre class="r"><code># source(&quot;../code/fit_cov_ebnmf.R&quot;)
# 
# fit.gbcd &lt;-
#   flash_fit_cov_ebnmf(Y = reduced_counts, Kmax = 7,
#                       prior = flash_ebnm(prior_family = &quot;generalized_binary&quot;,
#                                          scale = 0.04),
#                       extrapolate = FALSE)</code></pre>
<br>
<p>
<button type="button" class="btn btn-default btn-workflowr btn-workflowr-sessioninfo" data-toggle="collapse" data-target="#workflowr-sessioninfo" style="display: block;">
<span class="glyphicon glyphicon-wrench" aria-hidden="true"></span>
Session information
</button>
</p>
<div id="workflowr-sessioninfo" class="collapse">
<pre class="r"><code>sessionInfo()
# R version 4.3.1 (2023-06-16)
# Platform: aarch64-apple-darwin20 (64-bit)
# Running under: macOS Sonoma 14.5
# 
# Matrix products: default
# BLAS:   /Library/Frameworks/R.framework/Versions/4.3-arm64/Resources/lib/libRblas.0.dylib 
# LAPACK: /Library/Frameworks/R.framework/Versions/4.3-arm64/Resources/lib/libRlapack.dylib;  LAPACK version 3.11.0
# 
# locale:
# [1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8
# 
# time zone: America/Chicago
# tzcode source: internal
# 
# attached base packages:
# [1] stats     graphics  grDevices utils     datasets  methods   base     
# 
# other attached packages:
#  [1] SeuratObject_5.0.2 sp_2.1-4           readr_2.1.4        dplyr_1.1.4       
#  [5] cowplot_1.1.3      ggplot2_3.5.1      fastTopics_0.6-192 flashier_1.0.54   
#  [9] ebnm_1.1-34        Matrix_1.6-4      
# 
# loaded via a namespace (and not attached):
#  [1] pbapply_1.7-2        rlang_1.1.4          magrittr_2.0.3      
#  [4] git2r_0.35.0         horseshoe_0.2.0      compiler_4.3.1      
#  [7] vctrs_0.6.5          quadprog_1.5-8       stringr_1.5.1       
# [10] pkgconfig_2.0.3      crayon_1.5.3         fastmap_1.2.0       
# [13] labeling_0.4.3       utf8_1.2.4           promises_1.3.0      
# [16] rmarkdown_2.28       tzdb_0.4.0           purrr_1.0.2         
# [19] bit_4.0.5            xfun_0.48            cachem_1.1.0        
# [22] trust_0.1-8          jsonlite_1.8.9       progress_1.2.3      
# [25] highr_0.11           later_1.3.2          irlba_2.3.5.1       
# [28] parallel_4.3.1       prettyunits_1.2.0    R6_2.5.1            
# [31] bslib_0.8.0          stringi_1.8.4        RColorBrewer_1.1-3  
# [34] SQUAREM_2021.1       parallelly_1.38.0    jquerylib_0.1.4     
# [37] Rcpp_1.0.13          knitr_1.48           future.apply_1.11.3 
# [40] httpuv_1.6.14        splines_4.3.1        tidyselect_1.2.1    
# [43] rstudioapi_0.15.0    yaml_2.3.10          codetools_0.2-19    
# [46] listenv_0.9.1        lattice_0.21-8       tibble_3.2.1        
# [49] withr_3.0.2          evaluate_1.0.1       Rtsne_0.17          
# [52] future_1.34.0        RcppParallel_5.1.9   pillar_1.9.0        
# [55] whisker_0.4.1        plotly_4.10.4        softImpute_1.4-1    
# [58] generics_0.1.3       vroom_1.6.3          rprojroot_2.0.3     
# [61] invgamma_1.1         truncnorm_1.0-9      hms_1.1.3           
# [64] munsell_0.5.1        scales_1.3.0         ashr_2.2-63         
# [67] gtools_3.9.5         globals_0.16.3       RhpcBLASctl_0.23-42 
# [70] glue_1.8.0           scatterplot3d_0.3-44 lazyeval_0.2.2      
# [73] tools_4.3.1          data.table_1.16.2    dotCall64_1.2       
# [76] fs_1.6.4             grid_4.3.1           tidyr_1.3.1         
# [79] colorspace_2.1-1     deconvolveR_1.2-1    cli_3.6.3           
# [82] Polychrome_1.5.1     workflowr_1.7.1      spam_2.11-0         
# [85] fansi_1.0.6          mixsqp_0.3-54        viridisLite_0.4.2   
# [88] uwot_0.1.16          gtable_0.3.6         sass_0.4.9          
# [91] digest_0.6.37        progressr_0.14.0     ggrepel_0.9.6       
# [94] farver_2.1.2         htmlwidgets_1.6.4    htmltools_0.5.8.1   
# [97] lifecycle_1.0.4      httr_1.4.7           bit64_4.0.5</code></pre>
</div>
</div>
</div>
</div>
</div>
</div>
</div>


<!-- Adjust MathJax settings so that all math formulae are shown using
TeX fonts only; see
https://docs.mathjax.org/en/latest/web/configuration.html. This will make
the presentation more consistent at the cost of the webpage sometimes
taking slightly longer to load. Note that this only works because the
footer is added to webpages before the MathJax javascript. -->
<script type="text/x-mathjax-config">
  MathJax.Hub.Config({
    "HTML-CSS": { availableFonts: ["TeX"] }
  });
</script>





</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.odd').parent('tbody').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- tabsets -->

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open');
  });
});
</script>

<!-- code folding -->


<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
