<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />


<meta name="author" content="Peter Carbonetto" />


<title>NMF analysis of the LPS data set</title>

<script src="site_libs/header-attrs-2.29/header-attrs.js"></script>
<script src="site_libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/readable.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<style>h1 {font-size: 34px;}
       h1.title {font-size: 38px;}
       h2 {font-size: 30px;}
       h3 {font-size: 24px;}
       h4 {font-size: 18px;}
       h5 {font-size: 16px;}
       h6 {font-size: 12px;}
       code {color: inherit; background-color: rgba(0, 0, 0, 0.04);}
       pre:not([class]) { background-color: white }</style>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<link href="site_libs/highlightjs-9.12.0/textmate.css" rel="stylesheet" />
<script src="site_libs/highlightjs-9.12.0/highlight.js"></script>

<link rel="icon" href="https://github.com/workflowr/workflowr-assets/raw/main/img/reproducible.png">
<!-- Add a small amount of space between sections. -->
<style type="text/css">
div.section {
  padding-top: 12px;
}
</style>



<style type="text/css">
  code{white-space: pre-wrap;}
  span.smallcaps{font-variant: small-caps;}
  span.underline{text-decoration: underline;}
  div.column{display: inline-block; vertical-align: top; width: 50%;}
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  ul.task-list{list-style: none;}
    </style>

<style type="text/css">code{white-space: pre;}</style>
<script type="text/javascript">
if (window.hljs) {
  hljs.configure({languages: []});
  hljs.initHighlightingOnLoad();
  if (document.readyState && document.readyState === "complete") {
    window.setTimeout(function() { hljs.initHighlighting(); }, 0);
  }
}
</script>









<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
img {
  max-width:100%;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
details > summary > p:only-child {
  display: inline;
}
pre code {
  padding: 0;
}
</style>


<style type="text/css">
.dropdown-submenu {
  position: relative;
}
.dropdown-submenu>.dropdown-menu {
  top: 0;
  left: 100%;
  margin-top: -6px;
  margin-left: -1px;
  border-radius: 0 6px 6px 6px;
}
.dropdown-submenu:hover>.dropdown-menu {
  display: block;
}
.dropdown-submenu>a:after {
  display: block;
  content: " ";
  float: right;
  width: 0;
  height: 0;
  border-color: transparent;
  border-style: solid;
  border-width: 5px 0 5px 5px;
  border-left-color: #cccccc;
  margin-top: 5px;
  margin-right: -10px;
}
.dropdown-submenu:hover>a:after {
  border-left-color: #adb5bd;
}
.dropdown-submenu.pull-left {
  float: none;
}
.dropdown-submenu.pull-left>.dropdown-menu {
  left: -100%;
  margin-left: 10px;
  border-radius: 6px 0 6px 6px;
}
</style>

<script type="text/javascript">
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark the anchor link active (and if it's in a dropdown, also mark that active)
  var dropdown = menuAnchor.closest('li.dropdown');
  if (window.bootstrap) { // Bootstrap 4+
    menuAnchor.addClass('active');
    dropdown.find('> .dropdown-toggle').addClass('active');
  } else { // Bootstrap 3
    menuAnchor.parent().addClass('active');
    dropdown.addClass('active');
  }

  // Navbar adjustments
  var navHeight = $(".navbar").first().height() + 15;
  var style = document.createElement('style');
  var pt = "padding-top: " + navHeight + "px; ";
  var mt = "margin-top: -" + navHeight + "px; ";
  var css = "";
  // offset scroll position for anchor links (for fixed navbar)
  for (var i = 1; i <= 6; i++) {
    css += ".section h" + i + "{ " + pt + mt + "}\n";
  }
  style.innerHTML = "body {" + pt + "padding-bottom: 40px; }\n" + css;
  document.head.appendChild(style);
});
</script>

<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before, .tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "\e259";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "\e258";
  font-family: 'Glyphicons Halflings';
  border: none;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
  background-color: transparent;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<!-- code folding -->




</head>

<body>


<div class="container-fluid main-container">




<div class="navbar navbar-default  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-bs-toggle="collapse" data-target="#navbar" data-bs-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html">single-cell jamboree</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="index.html">Overview</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/stephenslab/single-cell-jamboree">source</a>
</li>
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<div id="header">



<h1 class="title toc-ignore">NMF analysis of the LPS data set</h1>
<h4 class="author">Peter Carbonetto</h4>

</div>


<p>
<button type="button" class="btn btn-default btn-workflowr btn-workflowr-report" data-toggle="collapse" data-target="#workflowr-report">
<span class="glyphicon glyphicon-list" aria-hidden="true"></span>
workflowr <span class="glyphicon glyphicon-exclamation-sign text-danger"
aria-hidden="true"></span>
</button>
</p>
<div id="workflowr-report" class="collapse">
<ul class="nav nav-tabs">
<li class="active">
<a data-toggle="tab" href="#summary">Summary</a>
</li>
<li>
<a data-toggle="tab" href="#checks"> Checks <span
class="glyphicon glyphicon-exclamation-sign text-danger"
aria-hidden="true"></span> </a>
</li>
<li>
<a data-toggle="tab" href="#versions">Past versions</a>
</li>
</ul>
<div class="tab-content">
<div id="summary" class="tab-pane fade in active">
<p>
<strong>Last updated:</strong> 2025-08-16
</p>
<p>
<strong>Checks:</strong> <span
class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> 6
<span class="glyphicon glyphicon-exclamation-sign text-danger"
aria-hidden="true"></span> 1
</p>
<p>
<strong>Knit directory:</strong>
<code>single-cell-jamboree/analysis/</code> <span
class="glyphicon glyphicon-question-sign" aria-hidden="true"
title="This is the local directory in which the code in this file was executed.">
</span>
</p>
<p>
This reproducible <a href="https://rmarkdown.rstudio.com">R Markdown</a>
analysis was created with <a
  href="https://github.com/workflowr/workflowr">workflowr</a> (version
1.7.1). The <em>Checks</em> tab describes the reproducibility checks
that were applied when the results were created. The <em>Past
versions</em> tab lists the development history.
</p>
<hr>
</div>
<div id="checks" class="tab-pane fade">
<div id="workflowr-checks" class="panel-group">
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongRMarkdownfilestronguptodate">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>R Markdown file:</strong> up-to-date
</a>
</p>
</div>
<div id="strongRMarkdownfilestronguptodate"
class="panel-collapse collapse">
<div class="panel-body">
<p>Great! Since the R Markdown file has been committed to the Git
repository, you know the exact version of the code that produced these
results.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongEnvironmentstrongempty">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Environment:</strong> empty </a>
</p>
</div>
<div id="strongEnvironmentstrongempty" class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! The global environment was empty. Objects defined in the
global environment can affect the analysis in your R Markdown file in
unknown ways. For reproduciblity it’s best to always run the code in an
empty environment.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongSeedstrongcodesetseed1code">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Seed:</strong>
<code>set.seed(1)</code> </a>
</p>
</div>
<div id="strongSeedstrongcodesetseed1code"
class="panel-collapse collapse">
<div class="panel-body">
<p>The command <code>set.seed(1)</code> was run prior to running the
code in the R Markdown file. Setting a seed ensures that any results
that rely on randomness, e.g. subsampling or permutations, are
reproducible.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongSessioninformationstrongrecorded">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Session information:</strong>
recorded </a>
</p>
</div>
<div id="strongSessioninformationstrongrecorded"
class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! Recording the operating system, R version, and package
versions is critical for reproducibility.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongCachestrongdetected">
<span class="glyphicon glyphicon-exclamation-sign text-danger"
aria-hidden="true"></span> <strong>Cache:</strong> detected </a>
</p>
</div>
<div id="strongCachestrongdetected" class="panel-collapse collapse">
<div class="panel-body">
The following chunks had caches available:
<ul>
<li>
fit-topic-model
</li>
<li>
fit-topic-model-k9
</li>
<li>
flashier-nmf-cc
</li>
<li>
flashier-nmf-k15
</li>
<li>
flashier-nmf-k9
</li>
</ul>
<p>To ensure reproducibility of the results, delete the cache directory
<code>lps_cache</code> and re-run the analysis. To have workflowr
automatically delete the cache directory prior to building the file, set
<code>delete_cache = TRUE</code> when running <code>wflow_build()</code>
or <code>wflow_publish()</code>.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongFilepathsstrongrelative">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>File paths:</strong> relative </a>
</p>
</div>
<div id="strongFilepathsstrongrelative" class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! Using relative paths to the files within your workflowr
project makes it easier to run your code on other machines.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongRepositoryversionstrongahrefhttpsgithubcomstephenslabsinglecelljamboreetreea13c906fcc3e7873892211a3d960dad808a1f489targetblanka13c906a">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Repository version:</strong>
<a href="https://github.com/stephenslab/single-cell-jamboree/tree/a13c906fcc3e7873892211a3d960dad808a1f489" target="_blank">a13c906</a>
</a>
</p>
</div>
<div
id="strongRepositoryversionstrongahrefhttpsgithubcomstephenslabsinglecelljamboreetreea13c906fcc3e7873892211a3d960dad808a1f489targetblanka13c906a"
class="panel-collapse collapse">
<div class="panel-body">
<p>
Great! You are using Git for version control. Tracking code development
and connecting the code version to the results is critical for
reproducibility.
</p>
<p>
The results in this page were generated with repository version
<a href="https://github.com/stephenslab/single-cell-jamboree/tree/a13c906fcc3e7873892211a3d960dad808a1f489" target="_blank">a13c906</a>.
See the <em>Past versions</em> tab to see a history of the changes made
to the R Markdown and HTML files.
</p>
<p>
Note that you need to be careful to ensure that all relevant files for
the analysis have been committed to Git prior to generating the results
(you can use <code>wflow_publish</code> or
<code>wflow_git_commit</code>). workflowr only checks the R Markdown
file, but you know if there are other scripts or data files that it
depends on. Below is the status of the Git repository when the results
were generated:
</p>
<pre><code>
Untracked files:
    Untracked:  analysis/lps_cache/
    Untracked:  analysis/lps_fl_nmf_cc_k10.csv
    Untracked:  analysis/lps_gsea_fl_nmf_cc_k10.csv
    Untracked:  analysis/mcf7_cache/
    Untracked:  analysis/pancreas_cytokine_S1_factors_cache/
    Untracked:  analysis/pancreas_cytokine_lsa_clustering_cache/
    Untracked:  data/GSE125162_ALL-fastqTomat0-Counts.tsv.gz
    Untracked:  data/GSE132188_adata.h5ad.h5
    Untracked:  data/GSE156175_RAW/
    Untracked:  data/GSE183010/
    Untracked:  data/Immune_ALL_human.h5ad
    Untracked:  data/pancreas_cytokine.RData
    Untracked:  data/pancreas_cytokine_lsa.RData
    Untracked:  data/pancreas_cytokine_lsa_v2.RData
    Untracked:  data/pancreas_endocrine.RData
    Untracked:  data/pancreas_endocrine_alldays.h5ad
    Untracked:  data/yeast.RData
    Untracked:  output/panc_cyto_lsa_res/

</code></pre>
<p>
Note that any generated files, e.g. HTML, png, CSS, etc., are not
included in this status report because it is ok for generated content to
have uncommitted changes.
</p>
</div>
</div>
</div>
</div>
<hr>
</div>
<div id="versions" class="tab-pane fade">

<p>
These are the previous versions of the repository in which changes were
made to the R Markdown (<code>analysis/lps.Rmd</code>) and HTML
(<code>docs/lps.html</code>) files. If you’ve configured a remote Git
repository (see <code>?wflow_git_remote</code>), click on the hyperlinks
in the table below to view the files as they were in that past version.
</p>
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
File
</th>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
<th>
Message
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/stephenslab/single-cell-jamboree/blob/a13c906fcc3e7873892211a3d960dad808a1f489/analysis/lps.Rmd" target="_blank">a13c906</a>
</td>
<td>
Peter Carbonetto
</td>
<td>
2025-08-16
</td>
<td>
wflow_publish("lps.Rmd", view = F, verbose = T)
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/stephenslab/single-cell-jamboree/blob/62e1130adefc9f572579f252cb2b366d852fc139/analysis/lps.Rmd" target="_blank">62e1130</a>
</td>
<td>
Peter Carbonetto
</td>
<td>
2025-08-16
</td>
<td>
Updated the GSEA to use factor 10 from the fl_nmf_cc model fit in
lps.Rmd.
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/stephenslab/single-cell-jamboree/blob/923976b7bc09d7dd1c29bed632c87497a9879307/analysis/lps.Rmd" target="_blank">923976b</a>
</td>
<td>
Peter Carbonetto
</td>
<td>
2025-08-15
</td>
<td>
In lps.Rmd, added additional MF analysis using selected pn factors; also
updated the umap plots.
</td>
</tr>
<tr>
<td>
html
</td>
<td>
<a href="https://rawcdn.githack.com/stephenslab/single-cell-jamboree/63aa311673ba46aeac2ef2f05e0f6eca2c3ecc02/docs/lps.html" target="_blank">63aa311</a>
</td>
<td>
Peter Carbonetto
</td>
<td>
2025-08-13
</td>
<td>
Ran wflow_publish("lps.Rmd").
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/stephenslab/single-cell-jamboree/blob/daa0c2dcdbe4783da64a0140866dc2d906b8cd6a/analysis/lps.Rmd" target="_blank">daa0c2d</a>
</td>
<td>
Peter Carbonetto
</td>
<td>
2025-08-13
</td>
<td>
wflow_publish("lps.Rmd", verbose = T, view = F)
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/stephenslab/single-cell-jamboree/blob/71595a0a69c3e809f3b0ece32d9c451953597d22/analysis/lps.Rmd" target="_blank">71595a0</a>
</td>
<td>
Peter Carbonetto
</td>
<td>
2025-08-13
</td>
<td>
Working on code for adding ‘cross-cutting’ factors to flashier model in
lps analysis.
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/stephenslab/single-cell-jamboree/blob/46c7c870d802fe84e4170aa4320c209f43233f83/analysis/lps.Rmd" target="_blank">46c7c87</a>
</td>
<td>
Peter Carbonetto
</td>
<td>
2025-08-13
</td>
<td>
Revised a couple of the Structure plots in the lps analysis.
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/stephenslab/single-cell-jamboree/blob/83ce75dc7140ad08cc61f091e6438cde9d3020ed/analysis/lps.Rmd" target="_blank">83ce75d</a>
</td>
<td>
Peter Carbonetto
</td>
<td>
2025-08-12
</td>
<td>
Started initial steps for processing of budding yeast data (GSE125162).
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/stephenslab/single-cell-jamboree/blob/4efb167f0e3587a64a1fab9104784d1e470b1005/analysis/lps.Rmd" target="_blank">4efb167</a>
</td>
<td>
Peter Carbonetto
</td>
<td>
2025-08-04
</td>
<td>
Working on incorporating the new flashier_nmf implementation into the
lps analysis.
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/stephenslab/single-cell-jamboree/blob/4887c84137b0c21ae91433b586e8f67656f41d48/analysis/lps.Rmd" target="_blank">4887c84</a>
</td>
<td>
Peter Carbonetto
</td>
<td>
2025-07-23
</td>
<td>
Added new gsea result from lps analysis (lps_gsea_fl_nmf_k6.csv).
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/stephenslab/single-cell-jamboree/blob/17e427ecb618ee375ea3bd7071498991ca4bff96/analysis/lps.Rmd" target="_blank">17e427e</a>
</td>
<td>
Peter Carbonetto
</td>
<td>
2025-07-23
</td>
<td>
Added some code to save another output from the lps analysis.
</td>
</tr>
<tr>
<td>
html
</td>
<td>
<a href="https://rawcdn.githack.com/stephenslab/single-cell-jamboree/de50dc2b081baf30b7bf8848c4548aaadc9f5027/docs/lps.html" target="_blank">de50dc2</a>
</td>
<td>
Peter Carbonetto
</td>
<td>
2025-07-17
</td>
<td>
Added some histograms showing the enrichment of the IFN-a/b genes in
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/stephenslab/single-cell-jamboree/blob/d11223e5ec4c9727a4531ecb09a8722a04e9239e/analysis/lps.Rmd" target="_blank">d11223e</a>
</td>
<td>
Peter Carbonetto
</td>
<td>
2025-07-17
</td>
<td>
wflow_publish("lps.Rmd", view = F, verbose = T)
</td>
</tr>
<tr>
<td>
html
</td>
<td>
<a href="https://rawcdn.githack.com/stephenslab/single-cell-jamboree/34e8a3a89bd218d00cd6c03c94f68e23da883870/docs/lps.html" target="_blank">34e8a3a</a>
</td>
<td>
Peter Carbonetto
</td>
<td>
2025-07-16
</td>
<td>
Switched to using the flashier_nmf function from the
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/stephenslab/single-cell-jamboree/blob/9682c188775daaff76a25aa18972795f54da9e34/analysis/lps.Rmd" target="_blank">9682c18</a>
</td>
<td>
Peter Carbonetto
</td>
<td>
2025-07-16
</td>
<td>
wflow_publish("lps.Rmd", view = F, verbose = T)
</td>
</tr>
<tr>
<td>
html
</td>
<td>
<a href="https://rawcdn.githack.com/stephenslab/single-cell-jamboree/61bec23064abf79cae4265fe42b9b1bd97d23183/docs/lps.html" target="_blank">61bec23</a>
</td>
<td>
Peter Carbonetto
</td>
<td>
2025-07-11
</td>
<td>
Added some UMAP plots to the lps analysis.
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/stephenslab/single-cell-jamboree/blob/d35b952ad8391b1e61969644b8d6a345ddbf7bbb/analysis/lps.Rmd" target="_blank">d35b952</a>
</td>
<td>
Peter Carbonetto
</td>
<td>
2025-07-11
</td>
<td>
wflow_publish("lps.Rmd", view = F, verbose = T)
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/stephenslab/single-cell-jamboree/blob/9147de4cd2f9cc76a5d0a38bb5d287e6315e6e0e/analysis/lps.Rmd" target="_blank">9147de4</a>
</td>
<td>
Peter Carbonetto
</td>
<td>
2025-07-04
</td>
<td>
Fixed a bug in lps.Rmd.
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/stephenslab/single-cell-jamboree/blob/3d8a2aa81fe4b8589b32b2e4b8e5cffac58586d5/analysis/lps.Rmd" target="_blank">3d8a2aa</a>
</td>
<td>
Peter Carbonetto
</td>
<td>
2025-07-04
</td>
<td>
Added lps_gsea_fl_nmf.csv output.
</td>
</tr>
<tr>
<td>
html
</td>
<td>
<a href="https://rawcdn.githack.com/stephenslab/single-cell-jamboree/e558f1753c6ec9edc1ffa77398ada9acbb5729bc/docs/lps.html" target="_blank">e558f17</a>
</td>
<td>
Peter Carbonetto
</td>
<td>
2025-07-04
</td>
<td>
Ran wflow_publish("lps.Rmd").
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/stephenslab/single-cell-jamboree/blob/3d27ef83d89b76bd33ea47ff8475d243c800cb71/analysis/lps.Rmd" target="_blank">3d27ef8</a>
</td>
<td>
Peter Carbonetto
</td>
<td>
2025-07-04
</td>
<td>
Added GSEA of factor k6 to the lps analysis.
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/stephenslab/single-cell-jamboree/blob/cafd86da943381d31de751b737f00b26cc3b6932/analysis/lps.Rmd" target="_blank">cafd86d</a>
</td>
<td>
Peter Carbonetto
</td>
<td>
2025-07-04
</td>
<td>
Implemented draft pathway analysis in temp4.R; next need to incorporate
into lps.Rmd.
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/stephenslab/single-cell-jamboree/blob/b7aff5986dbe0fa2d48204bd6564821e8665b59a/analysis/lps.Rmd" target="_blank">b7aff59</a>
</td>
<td>
Peter Carbonetto
</td>
<td>
2025-07-01
</td>
<td>
Added some scatterplots to compare topics and factors in the lps
analysis.
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/stephenslab/single-cell-jamboree/blob/31afa30f3fe0346d5bb03f9b1d100223c7c1fee9/analysis/lps.Rmd" target="_blank">31afa30</a>
</td>
<td>
Peter Carbonetto
</td>
<td>
2025-06-30
</td>
<td>
Added some notes/thoughts to lps.Rmd.
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/stephenslab/single-cell-jamboree/blob/23d0a0cf45ea79cf8a1a6cefdf3d2166827c0a64/analysis/lps.Rmd" target="_blank">23d0a0c</a>
</td>
<td>
Peter Carbonetto
</td>
<td>
2025-06-30
</td>
<td>
Fixed one of the structure plots in the lps analysis.
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/stephenslab/single-cell-jamboree/blob/842772ba71bb66d9a97b20db784aa9ce3fbbbb99/analysis/lps.Rmd" target="_blank">842772b</a>
</td>
<td>
Peter Carbonetto
</td>
<td>
2025-06-30
</td>
<td>
Added k=9 fits for the topic model and flashier NMF to the lps analysis.
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/stephenslab/single-cell-jamboree/blob/eb06be51c3f3b3bd8e3392eec83e519c30f6cc4b/analysis/lps.Rmd" target="_blank">eb06be5</a>
</td>
<td>
Peter Carbonetto
</td>
<td>
2025-06-30
</td>
<td>
Added a structure plot for the k=9 topic model to the lps analysis.
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/stephenslab/single-cell-jamboree/blob/3bfc932e41b2e2e3cc75e553a6e80c02ccd18207/analysis/lps.Rmd" target="_blank">3bfc932</a>
</td>
<td>
Peter Carbonetto
</td>
<td>
2025-06-27
</td>
<td>
Added a note to lps.Rmd.
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/stephenslab/single-cell-jamboree/blob/53085f148927b55c18eb8fc58ff7dd0fb66de3a1/analysis/lps.Rmd" target="_blank">53085f1</a>
</td>
<td>
Peter Carbonetto
</td>
<td>
2025-06-26
</td>
<td>
Working on adding a new topic model fit with k=9 topics in the lps
example to illutrate some key ideas.
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/stephenslab/single-cell-jamboree/blob/ce314bb479051afc1cdf9ba8b730f27664b0ab97/analysis/lps.Rmd" target="_blank">ce314bb</a>
</td>
<td>
Peter Carbonetto
</td>
<td>
2025-06-09
</td>
<td>
First try at running fastTopics and flashier on the pancreas_cytokine
data, for mouse = S1 only; from this analysis I learned that I need to
remove the mt and rp genes.
</td>
</tr>
<tr>
<td>
html
</td>
<td>
<a href="https://rawcdn.githack.com/stephenslab/single-cell-jamboree/4abf00ca66081239fb98de9b502ee9ebb25c4a13/docs/lps.html" target="_blank">4abf00c</a>
</td>
<td>
Peter Carbonetto
</td>
<td>
2025-06-06
</td>
<td>
Ran wflow_publish("lps.Rmd").
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/stephenslab/single-cell-jamboree/blob/f38b5868f566a5efd73d58eb9bf1f6f05b52cb8f/analysis/lps.Rmd" target="_blank">f38b586</a>
</td>
<td>
Peter Carbonetto
</td>
<td>
2025-06-06
</td>
<td>
A small fix to the lps analysis.
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/stephenslab/single-cell-jamboree/blob/aae4257ccd0323c365412c4a0ae9f104b30e6387/analysis/lps.Rmd" target="_blank">aae4257</a>
</td>
<td>
Peter Carbonetto
</td>
<td>
2025-06-06
</td>
<td>
A couple fixes to the lps analysis.
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/stephenslab/single-cell-jamboree/blob/6cbad5f1203de7ea08fa9769b49440a2f6138009/analysis/lps.Rmd" target="_blank">6cbad5f</a>
</td>
<td>
Peter Carbonetto
</td>
<td>
2025-06-06
</td>
<td>
Added a structure plot to the lps analysis.
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/stephenslab/single-cell-jamboree/blob/90d6c063f96aa1805b7ab6b456dc05eff2fd9626/analysis/lps.Rmd" target="_blank">90d6c06</a>
</td>
<td>
Peter Carbonetto
</td>
<td>
2025-06-06
</td>
<td>
Improved the structure plots in the lps analysis.
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/stephenslab/single-cell-jamboree/blob/dac95b566c5f66b67e3f46f8e4a1e50f4acf708d/analysis/lps.Rmd" target="_blank">dac95b5</a>
</td>
<td>
Peter Carbonetto
</td>
<td>
2025-06-06
</td>
<td>
Made a few changes to the flashier fit in the lps analysis.
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/stephenslab/single-cell-jamboree/blob/9e1f1279beceee0612c9ca09ce3de52ba22c38f8/analysis/lps.Rmd" target="_blank">9e1f127</a>
</td>
<td>
Peter Carbonetto
</td>
<td>
2025-06-05
</td>
<td>
Added code to pancreas_cytokine analysis to prepare the scrna-seq data
downloaded from geo.
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/stephenslab/single-cell-jamboree/blob/8d945a1ebcb216415939c8bb8811803ae9c79bbe/analysis/lps.Rmd" target="_blank">8d945a1</a>
</td>
<td>
Peter Carbonetto
</td>
<td>
2025-06-04
</td>
<td>
Added flashier fit to lps analysis; need to revise this and the topic
modeling result.
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/stephenslab/single-cell-jamboree/blob/dac6198b50dfb95b95f7ebb1f28227ad3d2f21fa/analysis/lps.Rmd" target="_blank">dac6198</a>
</td>
<td>
Peter Carbonetto
</td>
<td>
2025-06-04
</td>
<td>
Working on topic modeling results for lps data.
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/stephenslab/single-cell-jamboree/blob/8f39607ebba82349cd62736373fc6886c917ec5f/analysis/lps.Rmd" target="_blank">8f39607</a>
</td>
<td>
Peter Carbonetto
</td>
<td>
2025-06-04
</td>
<td>
Added steps to the lps analysis to load and prepare the data.
</td>
</tr>
<tr>
<td>
html
</td>
<td>
<a href="https://rawcdn.githack.com/stephenslab/single-cell-jamboree/2bfef0b5b759c8eacda0d00b99a9193de45ba241/docs/lps.html" target="_blank">2bfef0b</a>
</td>
<td>
Peter Carbonetto
</td>
<td>
2025-06-04
</td>
<td>
First build of the LPS analysis.
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/stephenslab/single-cell-jamboree/blob/85adf3f294176997b859f1e2351d672ce413af11/analysis/lps.Rmd" target="_blank">85adf3f</a>
</td>
<td>
Peter Carbonetto
</td>
<td>
2025-06-04
</td>
<td>
wflow_publish("lps.Rmd")
</td>
</tr>
</tbody>
</table>
</div>
<hr>
</div>
</div>
</div>
<p>Here we will revisit the LPS data set that we analyzed using a topic
model in the <a
href="https://doi.org/10.1038/s41590-023-01722-8">Takahama et al Nat
Immunol paper</a> (LPS = lipopolysaccharide). I believe some interesting
insights can be gained by examining this data set more deeply.</p>
<p>Load packages used to process the data, perform the analyses, and
create the plots.</p>
<pre class="r"><code>library(Matrix)
library(readr)
library(data.table)
library(fastTopics)
library(NNLM)
library(ebnm)
library(flashier)
library(pathways)
library(singlecelljamboreeR)
library(ggplot2)
library(ggrepel)
library(cowplot)
library(rsvd)
library(uwot)</code></pre>
<div id="prepare-the-data-for-analysis-with-fasttopics-and-flashier"
class="section level2">
<h2>Prepare the data for analysis with fastTopics and flashier</h2>
<p>Load the RNA-seq counts:</p>
<pre class="r"><code>read_lps_data &lt;- function (file) {
  counts &lt;- fread(file)
  class(counts) &lt;- &quot;data.frame&quot;
  genes &lt;- counts[,1]
  counts &lt;- t(as.matrix(counts[,-1]))
  colnames(counts) &lt;- genes
  samples &lt;- rownames(counts)
  samples &lt;- strsplit(samples,&quot;_&quot;)
  samples &lt;- data.frame(tissue    = sapply(samples,&quot;[[&quot;,1),
                        timepoint = sapply(samples,&quot;[[&quot;,2),
                        mouse     = sapply(samples,&quot;[[&quot;,3))
  samples &lt;- transform(samples,
                       tissue    = factor(tissue),
                       timepoint = factor(timepoint),
                       mouse     = factor(mouse))
  return(list(samples = samples,counts = counts))
}
out &lt;- read_lps_data(&quot;../data/lps.csv.gz&quot;)
samples &lt;- out$samples
counts  &lt;- out$counts
rm(out)</code></pre>
<p>Remove a sample that appears to be an outlier based on the NMF
analyses:</p>
<pre class="r"><code>i       &lt;- which(rownames(counts) != &quot;iLN_d2_20&quot;)
samples &lt;- samples[i,]
counts  &lt;- counts[i,]</code></pre>
<p>Remove genes that are expressed in fewer than 5 samples:</p>
<pre class="r"><code>j &lt;- which(colSums(counts &gt; 0) &gt; 4)
counts &lt;- counts[,j]</code></pre>
<p>This is the dimension of the data set we will analyze:</p>
<pre class="r"><code>dim(counts)
# [1]   363 33533</code></pre>
<p>For the Gaussian-based analyses, we will need the shifted log
counts:</p>
<pre class="r"><code>a &lt;- 1
s &lt;- rowSums(counts)
s &lt;- s/mean(s)
shifted_log_counts &lt;- log1p(counts/(a*s))</code></pre>
</div>
<div id="topic-model-fasttopics" class="section level2">
<h2>Topic model (fastTopics)</h2>
<p>First let’s fit a topic model with <span class="math inline">\(K =
9\)</span> topics to the counts. This is probably an insufficient number
of topics to fully capture the interesting structure in the data, but
this is done on purpose since I want to illustrate how the topic model
prioritizes the structure.</p>
<pre class="r"><code>set.seed(1)
tm_k9 &lt;- fit_poisson_nmf(counts,k = 9,init.method = &quot;random&quot;,method = &quot;em&quot;,
                         numiter = 20,verbose = &quot;none&quot;,
                         control = list(numiter=4,nc=8,extrapolate=FALSE))
tm_k9 &lt;- fit_poisson_nmf(counts,fit0 = tm_k9,method = &quot;scd&quot;,numiter = 40,
                         control = list(numiter = 4,nc = 8,extrapolate = TRUE),
                         verbose = &quot;none&quot;)</code></pre>
<div class="alert alert-warning">
<p>
<strong>Warning:</strong> The above code chunk cached its results, but
it won’t be re-run if previous chunks it depends on are updated. If you
need to use caching, it is highly recommended to also set
<code>knitr::opts_chunk$set(autodep = TRUE)</code> at the top of the
file (in a chunk that is not cached). Alternatively, you can customize
the option <code>dependson</code> for each individual chunk that is
cached. Using either <code>autodep</code> or <code>dependson</code> will
remove this warning. See the
<a href="https://yihui.name/knitr/options/#cache"
>knitr cache options</a> for more details.
</p>
</div>
<p>Structure plot comparing the topics to the tissue types:</p>
<pre class="r"><code>rows &lt;- order(samples$timepoint)
topic_colors &lt;- c(&quot;magenta&quot;,&quot;darkorange&quot;,&quot;darkblue&quot;,&quot;forestgreen&quot;,
                  &quot;dodgerblue&quot;,&quot;gray&quot;,&quot;red&quot;,&quot;olivedrab&quot;,&quot;darkmagenta&quot;,
                  &quot;sienna&quot;,&quot;limegreen&quot;,&quot;royalblue&quot;,&quot;lightskyblue&quot;,
                  &quot;gold&quot;)
samples &lt;- transform(samples,
  tissue = factor(tissue,c(&quot;PBMC&quot;,&quot;BM&quot;,&quot;CO&quot;,&quot;SI&quot;,&quot;iLN&quot;,&quot;SP&quot;,
                           &quot;TH&quot;,&quot;SK&quot;,&quot;BR&quot;,&quot;LI&quot;,&quot;HE&quot;,&quot;KI&quot;,&quot;LU&quot;)))
structure_plot(tm_k9,grouping = samples$tissue,gap = 4,
               topics = 1:9,colors = topic_colors,
               loadings_order = rows) +
  labs(fill = &quot;&quot;) +
  theme(axis.text.x = element_text(angle = 0,hjust = 0.5))</code></pre>
<p><img src="figure/lps.Rmd/structure-plot-topic-model-k9-1.png" width="840" style="display: block; margin: auto;" /></p>
<p>
<button type="button" class="btn btn-default btn-xs btn-workflowr btn-workflowr-fig" data-toggle="collapse" data-target="#fig-structure-plot-topic-model-k9-1">
Past versions of structure-plot-topic-model-k9-1.png
</button>
</p>
<div id="fig-structure-plot-topic-model-k9-1" class="collapse">
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
<a href="https://github.com/stephenslab/single-cell-jamboree/blob/e558f1753c6ec9edc1ffa77398ada9acbb5729bc/docs/figure/lps.Rmd/structure-plot-topic-model-k9-1.png" target="_blank">e558f17</a>
</td>
<td>
Peter Carbonetto
</td>
<td>
2025-07-04
</td>
</tr>
</tbody>
</table>
</div>
</div>
<p>Abbreviations used: BM = bone marrow; BR = brain; CO = colon; HE =
heart; iLN = inguinal lymph node; KI = kidney; LI = liver; LU = lung; SI
= small intestine; SK = skin; SP = spleen; TH = thymus.</p>
<p>The topics largely correspond to the different tissues, although
because there are more tissues than topics, some tissues that are more
similar to each other shared the same topic. It is also interesting
that, for the most part, none of the topics are capturing changes
downstream of the LPS treatment. So presumably these expression changes
are more subtle than the differences in expression among the
tissues.</p>
<p>Fit a topic model with <span class="math inline">\(K = 14\)</span>
topics to the counts:</p>
<pre class="r"><code>set.seed(1)
tm &lt;- fit_poisson_nmf(counts,k = 14,init.method = &quot;random&quot;,method = &quot;em&quot;,
                      numiter = 20,verbose = &quot;none&quot;,
                      control = list(numiter = 4,nc = 8,extrapolate = FALSE))
tm &lt;- fit_poisson_nmf(counts,fit0 = tm,method = &quot;scd&quot;,numiter = 40,
                      control = list(numiter = 4,nc = 8,extrapolate = TRUE),
                      verbose = &quot;none&quot;)</code></pre>
<div class="alert alert-warning">
<p>
<strong>Warning:</strong> The above code chunk cached its results, but
it won’t be re-run if previous chunks it depends on are updated. If you
need to use caching, it is highly recommended to also set
<code>knitr::opts_chunk$set(autodep = TRUE)</code> at the top of the
file (in a chunk that is not cached). Alternatively, you can customize
the option <code>dependson</code> for each individual chunk that is
cached. Using either <code>autodep</code> or <code>dependson</code> will
remove this warning. See the
<a href="https://yihui.name/knitr/options/#cache"
>knitr cache options</a> for more details.
</p>
</div>
<p>Structure plot comparing the topics to the tissue types:</p>
<pre class="r"><code>structure_plot(tm,grouping = samples$tissue,gap = 4,
               topics = 1:14,colors = topic_colors,
               loadings_order = rows) +
  labs(fill = &quot;&quot;) +
  theme(axis.text.x = element_text(angle = 0,hjust = 0.5),
        legend.key.height = unit(0.01,&quot;cm&quot;),
        legend.key.width = unit(0.2,&quot;cm&quot;),
        legend.text = element_text(size = 6))</code></pre>
<p><img src="figure/lps.Rmd/structure-plot-topic-model-1.png" width="840" style="display: block; margin: auto;" /></p>
<p>
<button type="button" class="btn btn-default btn-xs btn-workflowr btn-workflowr-fig" data-toggle="collapse" data-target="#fig-structure-plot-topic-model-1">
Past versions of structure-plot-topic-model-1.png
</button>
</p>
<div id="fig-structure-plot-topic-model-1" class="collapse">
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
<a href="https://github.com/stephenslab/single-cell-jamboree/blob/e558f1753c6ec9edc1ffa77398ada9acbb5729bc/docs/figure/lps.Rmd/structure-plot-topic-model-1.png" target="_blank">e558f17</a>
</td>
<td>
Peter Carbonetto
</td>
<td>
2025-07-04
</td>
</tr>
<tr>
<td>
<a href="https://github.com/stephenslab/single-cell-jamboree/blob/4abf00ca66081239fb98de9b502ee9ebb25c4a13/docs/figure/lps.Rmd/structure-plot-topic-model-1.png" target="_blank">4abf00c</a>
</td>
<td>
Peter Carbonetto
</td>
<td>
2025-06-06
</td>
</tr>
</tbody>
</table>
</div>
</div>
<p>This next structure plot better highlights the topics that are
capturing expression changes over time, some being presumably driven by
the LPS-induced sepsis:</p>
<pre class="r"><code>topic_colors &lt;- c(&quot;magenta&quot;,&quot;gray50&quot;,&quot;gray65&quot;,&quot;gray40&quot;,
                  &quot;gray85&quot;,&quot;gray75&quot;,&quot;red&quot;,&quot;gray80&quot;,&quot;gray90&quot;,
                  &quot;gray60&quot;,&quot;limegreen&quot;,&quot;gray70&quot;,&quot;gray55&quot;,
                  &quot;gold&quot;)
structure_plot(tm,grouping = samples$tissue,gap = 4,
               topics = 1:14,colors = topic_colors,
               loadings_order = rows) +
  labs(fill = &quot;&quot;) +
  theme(axis.text.x = element_text(angle = 0,hjust = 0.5),
        legend.key.height = unit(0.01,&quot;cm&quot;),
        legend.key.width = unit(0.2,&quot;cm&quot;),
        legend.text = element_text(size = 6))</code></pre>
<p><img src="figure/lps.Rmd/structure-plot-topic-model-2-1.png" width="840" style="display: block; margin: auto;" /></p>
<p>
<button type="button" class="btn btn-default btn-xs btn-workflowr btn-workflowr-fig" data-toggle="collapse" data-target="#fig-structure-plot-topic-model-2-1">
Past versions of structure-plot-topic-model-2-1.png
</button>
</p>
<div id="fig-structure-plot-topic-model-2-1" class="collapse">
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
<a href="https://github.com/stephenslab/single-cell-jamboree/blob/e558f1753c6ec9edc1ffa77398ada9acbb5729bc/docs/figure/lps.Rmd/structure-plot-topic-model-2-1.png" target="_blank">e558f17</a>
</td>
<td>
Peter Carbonetto
</td>
<td>
2025-07-04
</td>
</tr>
<tr>
<td>
<a href="https://github.com/stephenslab/single-cell-jamboree/blob/4abf00ca66081239fb98de9b502ee9ebb25c4a13/docs/figure/lps.Rmd/structure-plot-topic-model-2-1.png" target="_blank">4abf00c</a>
</td>
<td>
Peter Carbonetto
</td>
<td>
2025-06-06
</td>
</tr>
</tbody>
</table>
</div>
</div>
</div>
<div id="ebnmf-flashier" class="section level2">
<h2>EBNMF (flashier)</h2>
<p>Similar to the topic modeling analysis, let’s start by fitting an
EBNMF to the shifted log counts using flashier, first with <span
class="math inline">\(K = 9\)</span>. Since the greedy initialization
does not seem to work well in this example, I’ll use a different
initialization strategy: obtain a “good” initialization using the NNLM
package, then use this initialization to fit a NMF using flashier. This
approach is implemented in the following function:</p>
<p>Now fit an NMF to the shifted log counts, with <span
class="math inline">\(K = 9\)</span>:</p>
<pre class="r"><code>set.seed(1)
n  &lt;- nrow(shifted_log_counts)
x  &lt;- rpois(1e7,1/n)
s1 &lt;- sd(log(x + 1))
set.seed(1)
fl_nmf_k9 &lt;- flashier_nmf(shifted_log_counts,k = 9,greedy_init = TRUE,
                          var_type = 2,S = s1,verbose = 0)</code></pre>
<div class="alert alert-warning">
<p>
<strong>Warning:</strong> The above code chunk cached its results, but
it won’t be re-run if previous chunks it depends on are updated. If you
need to use caching, it is highly recommended to also set
<code>knitr::opts_chunk$set(autodep = TRUE)</code> at the top of the
file (in a chunk that is not cached). Alternatively, you can customize
the option <code>dependson</code> for each individual chunk that is
cached. Using either <code>autodep</code> or <code>dependson</code> will
remove this warning. See the
<a href="https://yihui.name/knitr/options/#cache"
>knitr cache options</a> for more details.
</p>
</div>
<p>Structure plot comparing the topics to the tissue types:</p>
<pre class="r"><code>rows &lt;- order(samples$timepoint)
topic_colors &lt;- c(&quot;powderblue&quot;,&quot;darkorange&quot;,&quot;olivedrab&quot;,&quot;limegreen&quot;,
                  &quot;forestgreen&quot;,&quot;cyan&quot;,&quot;darkmagenta&quot;,&quot;gray&quot;,&quot;dodgerblue&quot;,
                  &quot;red&quot;,&quot;gold&quot;,&quot;darkblue&quot;,&quot;lightskyblue&quot;,
                  &quot;royalblue&quot;,&quot;sienna&quot;)
L &lt;- ldf(fl_nmf_k9,type = &quot;i&quot;)$L
structure_plot(L,grouping = samples$tissue,gap = 4,
               topics = 1:9,colors = topic_colors,
               loadings_order = rows) +
  labs(fill = &quot;&quot;,y = &quot;membership&quot;)</code></pre>
<p><img src="figure/lps.Rmd/structure-plot-flashier-nmf-k9-1.png" width="840" style="display: block; margin: auto;" /></p>
<p>
<button type="button" class="btn btn-default btn-xs btn-workflowr btn-workflowr-fig" data-toggle="collapse" data-target="#fig-structure-plot-flashier-nmf-k9-1">
Past versions of structure-plot-flashier-nmf-k9-1.png
</button>
</p>
<div id="fig-structure-plot-flashier-nmf-k9-1" class="collapse">
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
<a href="https://github.com/stephenslab/single-cell-jamboree/blob/63aa311673ba46aeac2ef2f05e0f6eca2c3ecc02/docs/figure/lps.Rmd/structure-plot-flashier-nmf-k9-1.png" target="_blank">63aa311</a>
</td>
<td>
Peter Carbonetto
</td>
<td>
2025-08-13
</td>
</tr>
<tr>
<td>
<a href="https://github.com/stephenslab/single-cell-jamboree/blob/e558f1753c6ec9edc1ffa77398ada9acbb5729bc/docs/figure/lps.Rmd/structure-plot-flashier-nmf-k9-1.png" target="_blank">e558f17</a>
</td>
<td>
Peter Carbonetto
</td>
<td>
2025-07-04
</td>
</tr>
</tbody>
</table>
</div>
</div>
<p>Like the topic model, the EBNMF model with <span
class="math inline">\(K = 9\)</span> does not capture any changes
downstream of the LPS-induced sepsis.</p>
<p>Next fit an NMF to the shifted log counts using flashier, with <span
class="math inline">\(K =
15\)</span>:</p>
<pre class="r"><code>set.seed(1)
fl_nmf &lt;- flashier_nmf(shifted_log_counts,k = 15,greedy_init = TRUE,
                       var_type = 2,S = s1,verbose = 0)</code></pre>
<div class="alert alert-warning">
<p>
<strong>Warning:</strong> The above code chunk cached its results, but
it won’t be re-run if previous chunks it depends on are updated. If you
need to use caching, it is highly recommended to also set
<code>knitr::opts_chunk$set(autodep = TRUE)</code> at the top of the
file (in a chunk that is not cached). Alternatively, you can customize
the option <code>dependson</code> for each individual chunk that is
cached. Using either <code>autodep</code> or <code>dependson</code> will
remove this warning. See the
<a href="https://yihui.name/knitr/options/#cache"
>knitr cache options</a> for more details.
</p>
</div>
<p>Structure plot comparing the factors to the tissue types:</p>
<pre class="r"><code>L &lt;- ldf(fl_nmf,type = &quot;i&quot;)$L
structure_plot(L,grouping = samples$tissue,gap = 4,
               topics = 1:15,colors = topic_colors,
               loadings_order = rows) +
  labs(fill = &quot;&quot;,y = &quot;membership&quot;) +
  theme(axis.text.x = element_text(angle = 0,hjust = 0.5),
        legend.key.height = unit(0.01,&quot;cm&quot;),
        legend.key.width = unit(0.25,&quot;cm&quot;),
        legend.text = element_text(size = 7))</code></pre>
<p><img src="figure/lps.Rmd/structure-plot-flashier-nmf-1.png" width="840" style="display: block; margin: auto;" /></p>
<p>
<button type="button" class="btn btn-default btn-xs btn-workflowr btn-workflowr-fig" data-toggle="collapse" data-target="#fig-structure-plot-flashier-nmf-1">
Past versions of structure-plot-flashier-nmf-1.png
</button>
</p>
<div id="fig-structure-plot-flashier-nmf-1" class="collapse">
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
<a href="https://github.com/stephenslab/single-cell-jamboree/blob/63aa311673ba46aeac2ef2f05e0f6eca2c3ecc02/docs/figure/lps.Rmd/structure-plot-flashier-nmf-1.png" target="_blank">63aa311</a>
</td>
<td>
Peter Carbonetto
</td>
<td>
2025-08-13
</td>
</tr>
<tr>
<td>
<a href="https://github.com/stephenslab/single-cell-jamboree/blob/34e8a3a89bd218d00cd6c03c94f68e23da883870/docs/figure/lps.Rmd/structure-plot-flashier-nmf-1.png" target="_blank">34e8a3a</a>
</td>
<td>
Peter Carbonetto
</td>
<td>
2025-07-16
</td>
</tr>
<tr>
<td>
<a href="https://github.com/stephenslab/single-cell-jamboree/blob/e558f1753c6ec9edc1ffa77398ada9acbb5729bc/docs/figure/lps.Rmd/structure-plot-flashier-nmf-1.png" target="_blank">e558f17</a>
</td>
<td>
Peter Carbonetto
</td>
<td>
2025-07-04
</td>
</tr>
<tr>
<td>
<a href="https://github.com/stephenslab/single-cell-jamboree/blob/4abf00ca66081239fb98de9b502ee9ebb25c4a13/docs/figure/lps.Rmd/structure-plot-flashier-nmf-1.png" target="_blank">4abf00c</a>
</td>
<td>
Peter Carbonetto
</td>
<td>
2025-06-06
</td>
</tr>
</tbody>
</table>
</div>
</div>
<p>This next structure plot better highlights the topics that capture
the processes that are driven or may be driven by LPS-induced
sepsis:</p>
<pre class="r"><code>rows &lt;- order(samples$timepoint)
topic_colors &lt;- c(&quot;gray95&quot;,&quot;gray70&quot;,&quot;gray80&quot;,&quot;gray50&quot;,
                  &quot;gray60&quot;,&quot;gray90&quot;,&quot;gray75&quot;,&quot;gray&quot;,&quot;gray85&quot;,
                  &quot;red&quot;,&quot;gold&quot;,&quot;gray65&quot;,&quot;gray45&quot;,
                  &quot;gray35&quot;,&quot;gray75&quot;)
L &lt;- ldf(fl_nmf,type = &quot;i&quot;)$L
structure_plot(L,grouping = samples$tissue,gap = 4,
               topics = 1:15,colors = topic_colors,
               loadings_order = rows) +
  labs(fill = &quot;&quot;,y = &quot;membership&quot;) +
  theme(axis.text.x = element_text(angle = 0,hjust = 0.5),
        legend.key.height = unit(0.01,&quot;cm&quot;),
        legend.key.width = unit(0.25,&quot;cm&quot;),
        legend.text = element_text(size = 7))</code></pre>
<p><img src="figure/lps.Rmd/structure-plot-flashier-nmf-2-1.png" width="840" style="display: block; margin: auto;" /></p>
<p>
<button type="button" class="btn btn-default btn-xs btn-workflowr btn-workflowr-fig" data-toggle="collapse" data-target="#fig-structure-plot-flashier-nmf-2-1">
Past versions of structure-plot-flashier-nmf-2-1.png
</button>
</p>
<div id="fig-structure-plot-flashier-nmf-2-1" class="collapse">
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
<a href="https://github.com/stephenslab/single-cell-jamboree/blob/63aa311673ba46aeac2ef2f05e0f6eca2c3ecc02/docs/figure/lps.Rmd/structure-plot-flashier-nmf-2-1.png" target="_blank">63aa311</a>
</td>
<td>
Peter Carbonetto
</td>
<td>
2025-08-13
</td>
</tr>
<tr>
<td>
<a href="https://github.com/stephenslab/single-cell-jamboree/blob/34e8a3a89bd218d00cd6c03c94f68e23da883870/docs/figure/lps.Rmd/structure-plot-flashier-nmf-2-1.png" target="_blank">34e8a3a</a>
</td>
<td>
Peter Carbonetto
</td>
<td>
2025-07-16
</td>
</tr>
<tr>
<td>
<a href="https://github.com/stephenslab/single-cell-jamboree/blob/e558f1753c6ec9edc1ffa77398ada9acbb5729bc/docs/figure/lps.Rmd/structure-plot-flashier-nmf-2-1.png" target="_blank">e558f17</a>
</td>
<td>
Peter Carbonetto
</td>
<td>
2025-07-04
</td>
</tr>
<tr>
<td>
<a href="https://github.com/stephenslab/single-cell-jamboree/blob/4abf00ca66081239fb98de9b502ee9ebb25c4a13/docs/figure/lps.Rmd/structure-plot-flashier-nmf-2-1.png" target="_blank">4abf00c</a>
</td>
<td>
Peter Carbonetto
</td>
<td>
2025-06-06
</td>
</tr>
</tbody>
</table>
</div>
</div>
</div>
<div id="ebnmf-with-cross-cutting-factors" class="section level2">
<h2>EBNMF with “cross-cutting” factor(s)</h2>
<p>One question I have is whether it is sufficient to model increases in
expression for these “cross-cutting” factors, or whether it is better to
allow for both decreases in expression (negative on log-scale) and
increases (positive on log-scale). Let’s test this.</p>
<pre class="r"><code>cc_factors &lt;- 10:11
nn_factors &lt;- c(1:9,12:15)
fl_nmf_cc &lt;- convert_factors_nn_to_pn(fl_nmf,cc_factors,shifted_log_counts,
                                      S = s1,var_type = 2)
fl_nmf_cc &lt;- flash_backfit(fl_nmf_cc,extrapolate = FALSE,maxiter = 100,
                           verbose = 2)
fl_nmf_cc &lt;- flash_backfit(fl_nmf_cc,extrapolate = TRUE,maxiter = 100,
                           verbose = 2)
# Backfitting 15 factors (tolerance: 1.81e-01)...
#   Difference between iterations is within 1.0e+03...
#   Difference between iterations is within 1.0e+02...
#   --Maximum number of iterations reached!
# Wrapping up...
# Done.
# Backfitting 15 factors (tolerance: 1.81e-01)...
#   Difference between iterations is within 1.0e+01...
#   --Maximum number of iterations reached!
#   Backfit complete. Objective: -3286910.603 
# Wrapping up...
# Done.</code></pre>
<div class="alert alert-warning">
<p>
<strong>Warning:</strong> The above code chunk cached its results, but
it won’t be re-run if previous chunks it depends on are updated. If you
need to use caching, it is highly recommended to also set
<code>knitr::opts_chunk$set(autodep = TRUE)</code> at the top of the
file (in a chunk that is not cached). Alternatively, you can customize
the option <code>dependson</code> for each individual chunk that is
cached. Using either <code>autodep</code> or <code>dependson</code> will
remove this warning. See the
<a href="https://yihui.name/knitr/options/#cache"
>knitr cache options</a> for more details.
</p>
</div>
<p>For the two cross-cutting factors, modeling both decreases and
increases does seem to produce a number of decreases in expression, but
overall the LFCs remain very similar:</p>
<pre class="r"><code>pdat &lt;- data.frame(nmf    = as.vector(ldf(fl_nmf,type=&quot;i&quot;)$F[,cc_factors]),
                   nmf_cc = as.vector(ldf(fl_nmf_cc,type=&quot;i&quot;)$F[,cc_factors]))
ggplot(pdat,aes(x = nmf,y = nmf_cc)) +
  geom_point() +
  geom_abline(intercept = 0,slope = 1,color = &quot;magenta&quot;,linetype = &quot;dotted&quot;) +
  theme_cowplot(font_size = 10)</code></pre>
<p><img src="figure/lps.Rmd/compare-nmf-vs-cc-1-1.png" width="300" style="display: block; margin: auto;" /></p>
<p>This suggests that while modeling both positive and negative changes
is useful, it is perhaps not essential.</p>
<p>The following three plots compare: (i) F for the non-negative factors
only; (ii) L for the non-negative factors only; (iii) L for the
(positive and negative) cross-cutting factors. Overall, these parameter
estimates did not change much.</p>
<pre class="r"><code>pdat1 &lt;- data.frame(nmf    = as.vector(ldf(fl_nmf,type=&quot;i&quot;)$F[,nn_factors]),
                    nmf_cc = as.vector(ldf(fl_nmf_cc,type=&quot;i&quot;)$F[,nn_factors]))
pdat1 &lt;- subset(pdat1,abs(nmf) &gt; 0.01 &amp; abs(nmf_cc) &gt; 0.01)
p1 &lt;- ggplot(pdat1,aes(x = nmf,y = nmf_cc)) +
  geom_point() +
  geom_abline(intercept = 0,slope = 1,color = &quot;magenta&quot;,linetype = &quot;dotted&quot;) +
  theme_cowplot(font_size = 10)
pdat2 &lt;- data.frame(nmf    = as.vector(ldf(fl_nmf,type=&quot;i&quot;)$L[,nn_factors]),
                    nmf_cc = as.vector(ldf(fl_nmf_cc,type=&quot;i&quot;)$L[,nn_factors]))
pdat2 &lt;- subset(pdat2,abs(nmf) &gt; 0.01 &amp; abs(nmf_cc) &gt; 0.01)
p2 &lt;- ggplot(pdat2,aes(x = nmf,y = nmf_cc)) +
  geom_point() +
  geom_abline(intercept = 0,slope = 1,color = &quot;magenta&quot;,linetype = &quot;dotted&quot;) +
  theme_cowplot(font_size = 10)
pdat3 &lt;- data.frame(nmf    = as.vector(ldf(fl_nmf,type=&quot;i&quot;)$L[,cc_factors]),
                    nmf_cc = as.vector(ldf(fl_nmf_cc,type=&quot;i&quot;)$L[,cc_factors]))
p3 &lt;- ggplot(pdat3,aes(x = nmf,y = nmf_cc)) +
  geom_point() +
  geom_abline(intercept = 0,slope = 1,color = &quot;magenta&quot;,linetype = &quot;dotted&quot;) +
  theme_cowplot(font_size = 10)
plot_grid(p1,p2,p3,nrow = 1,ncol = 3)</code></pre>
<p><img src="figure/lps.Rmd/compare-nmf-vs-cc-2-1.png" width="720" style="display: block; margin: auto;" /></p>
<p>Here is the updated Structure plot, highlighting again the
sepsis-related (“cross-cutting”) factors:</p>
<pre class="r"><code>rows &lt;- order(samples$timepoint)
L &lt;- ldf(fl_nmf_cc,type = &quot;i&quot;)$L
structure_plot(L,grouping = samples$tissue,gap = 4,
                    topics = 1:15,colors = topic_colors,
                    loadings_order = rows) +
  labs(fill = &quot;&quot;,y = &quot;membership&quot;) +
  theme(axis.text.x = element_text(angle = 0,hjust = 0.5),
        legend.key.height = unit(0.01,&quot;cm&quot;),
        legend.key.width = unit(0.25,&quot;cm&quot;),
        legend.text = element_text(size = 7))</code></pre>
<p><img src="figure/lps.Rmd/structure-plot-flashier-nmf-cc-1.png" width="840" style="display: block; margin: auto;" /></p>
</div>
<div id="comparison-with-umap" class="section level2">
<h2>Comparison with UMAP</h2>
<p>Here I project the samples onto a 2-d embedding using UMAP to show
that this LPS-related substructure is not obvious from a UMAP plot.</p>
<pre class="r"><code>set.seed(1)
U &lt;- rsvd(shifted_log_counts,k = 40)$u
Y &lt;- umap(U,n_neighbors = 20,metric = &quot;cosine&quot;,min_dist = 0.3,
          n_threads = 8,verbose = FALSE)
x &lt;- Y[,1]
y &lt;- Y[,2]
samples$umap1 &lt;- x
samples$umap2 &lt;- y</code></pre>
<p>I color the samples in the UMAP plot by tissue (top-left) and by
membership to factor 10 (top-right, bottom-left):</p>
<pre class="r"><code>tissue_colors &lt;- c(&quot;magenta&quot;,&quot;darkorange&quot;,&quot;darkblue&quot;,&quot;forestgreen&quot;,
                   &quot;dodgerblue&quot;,&quot;red&quot;,&quot;olivedrab&quot;,&quot;darkmagenta&quot;,
                   &quot;sienna&quot;,&quot;limegreen&quot;,&quot;royalblue&quot;,&quot;lightskyblue&quot;,
                   &quot;gold&quot;)
L            &lt;- ldf(fl_nmf,type = &quot;i&quot;)$L
colnames(L)  &lt;- paste0(&quot;k&quot;,1:15)
pdat         &lt;- samples
pdat$factor10 &lt;- L[,&quot;k10&quot;]
pdat$topic7  &lt;- poisson2multinom(tm)$L[,&quot;k7&quot;]
p1 &lt;- ggplot(pdat,aes(x = umap1,y = umap2,color = tissue)) +
  geom_point(size = 1) +
  scale_color_manual(values = tissue_colors) +
  theme_cowplot(font_size = 10)
p2 &lt;- ggplot(pdat,aes(x = umap1,y = umap2,color = factor10)) +
  geom_point(size = 1) +
  scale_color_gradient2(low = &quot;deepskyblue&quot;,mid = &quot;gold&quot;,high = &quot;tomato&quot;,
                        midpoint = 0.66) +
  theme_cowplot(font_size = 10)
p3 &lt;- ggplot(pdat,aes(x = umap1,y = umap2,color = topic7)) +
  geom_point(size = 1) +
  scale_color_gradient2(low = &quot;deepskyblue&quot;,mid = &quot;gold&quot;,high = &quot;tomato&quot;,
                        midpoint = 0.15) +
  theme_cowplot(font_size = 10)
plot_grid(p1,p2,p3,nrow = 2,ncol = 2)</code></pre>
<p><img src="figure/lps.Rmd/umap-plots-1.png" width="780" style="display: block; margin: auto;" /></p>
<p>
<button type="button" class="btn btn-default btn-xs btn-workflowr btn-workflowr-fig" data-toggle="collapse" data-target="#fig-umap-plots-1">
Past versions of umap-plots-1.png
</button>
</p>
<div id="fig-umap-plots-1" class="collapse">
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
<a href="https://github.com/stephenslab/single-cell-jamboree/blob/63aa311673ba46aeac2ef2f05e0f6eca2c3ecc02/docs/figure/lps.Rmd/umap-plots-1.png" target="_blank">63aa311</a>
</td>
<td>
Peter Carbonetto
</td>
<td>
2025-08-13
</td>
</tr>
<tr>
<td>
<a href="https://github.com/stephenslab/single-cell-jamboree/blob/34e8a3a89bd218d00cd6c03c94f68e23da883870/docs/figure/lps.Rmd/umap-plots-1.png" target="_blank">34e8a3a</a>
</td>
<td>
Peter Carbonetto
</td>
<td>
2025-07-16
</td>
</tr>
<tr>
<td>
<a href="https://github.com/stephenslab/single-cell-jamboree/blob/61bec23064abf79cae4265fe42b9b1bd97d23183/docs/figure/lps.Rmd/umap-plots-1.png" target="_blank">61bec23</a>
</td>
<td>
Peter Carbonetto
</td>
<td>
2025-07-11
</td>
</tr>
</tbody>
</table>
</div>
</div>
<p>As expected, the predominant structure is due to the different
tissues, with more similar tissues clustering together. There is also
some more subtle structure within each cluster that appears to
correspond well with factor 10 (and, correspondingly, topic 7). So
althought the UMAP does seem to reveal the sepsis-related structure, it
cannot isolate the sepsis-related gene expression signature.</p>
</div>
<div id="factors-isolating-responses-to-lps-induced-sepsis"
class="section level2">
<h2>Factors isolating responses to LPS-induced sepsis</h2>
<p>From the Structure plots, it appears that topic 7, and possibly topic
1, are capturing processes activated by LPS. However, I conjecture that
it is difficult to determine which genes should members of topic 1 and
which are members of the colon and small intensine topics. Indeed, topic
1 shares many genes with these two topics:</p>
<pre class="r"><code>pdat &lt;- cbind(data.frame(gene = colnames(counts)),
              poisson2multinom(tm)$F)
rows &lt;- which(pdat$k1 &lt; 0.01)
pdat[rows,&quot;gene&quot;] &lt;- &quot;&quot;
p1 &lt;- ggplot(pdat,aes(x = k1,y = k4,label = gene)) +
  geom_point() +
  geom_text_repel(color = &quot;gray&quot;,size = 2.5,max.overlaps = Inf) +
  scale_x_continuous(trans = &quot;sqrt&quot;) +
  scale_y_continuous(trans = &quot;sqrt&quot;) +
  theme_cowplot(font_size = 10)
p2 &lt;- ggplot(pdat,aes(x = k1,y = k13,label = gene)) +
  geom_point() +
  geom_text_repel(color = &quot;gray&quot;,size = 2.5,max.overlaps = Inf) +
  scale_x_continuous(trans = &quot;sqrt&quot;) +
  scale_y_continuous(trans = &quot;sqrt&quot;) +
  theme_cowplot(font_size = 10)
plot_grid(p1,p2,nrow = 1,ncol = 2)</code></pre>
<p><img src="figure/lps.Rmd/compare-topics-1-1.png" width="600" style="display: block; margin: auto;" /></p>
<p>
<button type="button" class="btn btn-default btn-xs btn-workflowr btn-workflowr-fig" data-toggle="collapse" data-target="#fig-compare-topics-1-1">
Past versions of compare-topics-1-1.png
</button>
</p>
<div id="fig-compare-topics-1-1" class="collapse">
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
<a href="https://github.com/stephenslab/single-cell-jamboree/blob/63aa311673ba46aeac2ef2f05e0f6eca2c3ecc02/docs/figure/lps.Rmd/compare-topics-1-1.png" target="_blank">63aa311</a>
</td>
<td>
Peter Carbonetto
</td>
<td>
2025-08-13
</td>
</tr>
<tr>
<td>
<a href="https://github.com/stephenslab/single-cell-jamboree/blob/34e8a3a89bd218d00cd6c03c94f68e23da883870/docs/figure/lps.Rmd/compare-topics-1-1.png" target="_blank">34e8a3a</a>
</td>
<td>
Peter Carbonetto
</td>
<td>
2025-07-16
</td>
</tr>
<tr>
<td>
<a href="https://github.com/stephenslab/single-cell-jamboree/blob/61bec23064abf79cae4265fe42b9b1bd97d23183/docs/figure/lps.Rmd/compare-topics-1-1.png" target="_blank">61bec23</a>
</td>
<td>
Peter Carbonetto
</td>
<td>
2025-07-11
</td>
</tr>
<tr>
<td>
<a href="https://github.com/stephenslab/single-cell-jamboree/blob/e558f1753c6ec9edc1ffa77398ada9acbb5729bc/docs/figure/lps.Rmd/compare-topics-1-1.png" target="_blank">e558f17</a>
</td>
<td>
Peter Carbonetto
</td>
<td>
2025-07-04
</td>
</tr>
</tbody>
</table>
</div>
</div>
<p>Still, it is interesting that three genes, <em>Reg3b</em>,
<em>Reg3g</em> and <em>Lypd8</em>, stand out in topic 1 as distinct from
the colon and SI topics. Let’s now contrast this to the situation for
topic 7:</p>
<pre class="r"><code>pdat &lt;- cbind(data.frame(gene = colnames(counts)),
              poisson2multinom(tm)$F)
rows &lt;- which(pdat$k7 &lt; 0.008)
pdat[rows,&quot;gene&quot;] &lt;- &quot;&quot;
p1 &lt;- ggplot(pdat,aes(x = k7,y = k6,label = gene)) +
  geom_point() +
  geom_text_repel(color = &quot;gray&quot;,size = 2.5,max.overlaps = Inf) +
  scale_x_continuous(trans = &quot;sqrt&quot;) +
  scale_y_continuous(trans = &quot;sqrt&quot;) +
  theme_cowplot(font_size = 10)
p2 &lt;- ggplot(pdat,aes(x = k7,y = k8,label = gene)) +
  geom_point() +
  geom_text_repel(color = &quot;gray&quot;,size = 2.5,max.overlaps = Inf) +
  scale_x_continuous(trans = &quot;sqrt&quot;) +
  scale_y_continuous(trans = &quot;sqrt&quot;) +
  theme_cowplot(font_size = 10)
plot_grid(p1,p2,nrow = 1,ncol = 2)</code></pre>
<p><img src="figure/lps.Rmd/compare-topics-2-1.png" width="600" style="display: block; margin: auto;" /></p>
<p>
<button type="button" class="btn btn-default btn-xs btn-workflowr btn-workflowr-fig" data-toggle="collapse" data-target="#fig-compare-topics-2-1">
Past versions of compare-topics-2-1.png
</button>
</p>
<div id="fig-compare-topics-2-1" class="collapse">
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
<a href="https://github.com/stephenslab/single-cell-jamboree/blob/63aa311673ba46aeac2ef2f05e0f6eca2c3ecc02/docs/figure/lps.Rmd/compare-topics-2-1.png" target="_blank">63aa311</a>
</td>
<td>
Peter Carbonetto
</td>
<td>
2025-08-13
</td>
</tr>
<tr>
<td>
<a href="https://github.com/stephenslab/single-cell-jamboree/blob/de50dc2b081baf30b7bf8848c4548aaadc9f5027/docs/figure/lps.Rmd/compare-topics-2-1.png" target="_blank">de50dc2</a>
</td>
<td>
Peter Carbonetto
</td>
<td>
2025-07-17
</td>
</tr>
<tr>
<td>
<a href="https://github.com/stephenslab/single-cell-jamboree/blob/34e8a3a89bd218d00cd6c03c94f68e23da883870/docs/figure/lps.Rmd/compare-topics-2-1.png" target="_blank">34e8a3a</a>
</td>
<td>
Peter Carbonetto
</td>
<td>
2025-07-16
</td>
</tr>
<tr>
<td>
<a href="https://github.com/stephenslab/single-cell-jamboree/blob/61bec23064abf79cae4265fe42b9b1bd97d23183/docs/figure/lps.Rmd/compare-topics-2-1.png" target="_blank">61bec23</a>
</td>
<td>
Peter Carbonetto
</td>
<td>
2025-07-11
</td>
</tr>
<tr>
<td>
<a href="https://github.com/stephenslab/single-cell-jamboree/blob/e558f1753c6ec9edc1ffa77398ada9acbb5729bc/docs/figure/lps.Rmd/compare-topics-2-1.png" target="_blank">e558f17</a>
</td>
<td>
Peter Carbonetto
</td>
<td>
2025-07-04
</td>
</tr>
</tbody>
</table>
</div>
</div>
<p>For illustration, I compared topic 7 to the kidney and lung topics.
The key point here is that the topic model has selected genes for topic
7 that are very independent of the tissue topics. So this looks quite
promising. Let’s now see if the result is similar for the EBNMF model
fitted to the shifted log counts:</p>
<pre class="r"><code>F &lt;- ldf(fl_nmf_cc,type = &quot;i&quot;)$F
colnames(F) &lt;- paste0(&quot;k&quot;,1:15)
pdat &lt;- data.frame(tm     = poisson2multinom(tm)$F[,&quot;k7&quot;],
                   nmf_cc = F[,&quot;k10&quot;],
                   gene   = rownames(F))
rows &lt;- which(with(pdat,tm &lt; 0.005 &amp; nmf_cc &lt; 0.8))
pdat[rows,&quot;gene&quot;] &lt;- &quot;&quot;
ggplot(pdat,aes(x = (tm)^(1/3),y = nmf_cc,label = gene)) +
  geom_point() +
  geom_text_repel(color = &quot;gray&quot;,size = 2.5,max.overlaps = Inf) +
  labs(x = &quot;topic 7&quot;,y = &quot;factor 10&quot;) +
  theme_cowplot(font_size = 12)</code></pre>
<p><img src="figure/lps.Rmd/sepsis-tm-vs-nmf-cc-1.png" width="480" style="display: block; margin: auto;" /></p>
<p>Indeed, factor 10 and topic 7 are cpaturing very similar expression
patterns.</p>
<p>Next I ran GSEA on the on factor 10. (Running GSEA on topic 7 is
complicated by the fact that it would be better to “shrink” the
estimates before running GSEA, whereas this was automatically done for
EBNMF result.)</p>
<pre class="r"><code>data(gene_sets_mouse)
gene_sets     &lt;- gene_sets_mouse$gene_sets
gene_info     &lt;- gene_sets_mouse$gene_info
gene_set_info &lt;- gene_sets_mouse$gene_set_info
j &lt;- which(with(gene_sets_mouse$gene_set_info,
                (database == &quot;MSigDB-C2&quot; &amp;
                 grepl(&quot;CP&quot;,sub_category_code,fixed = TRUE)) |
                (database == &quot;MSigDB-C5&quot;) &amp;
                 grepl(&quot;GO&quot;,sub_category_code,fixed = TRUE)))
genes &lt;- sort(intersect(rownames(F),gene_info$Symbol))
i     &lt;- which(is.element(gene_info$Symbol,genes))
gene_info     &lt;- gene_info[i,]
gene_set_info &lt;- gene_set_info[j,]
gene_sets     &lt;- gene_sets[i,j]
rownames(gene_sets) &lt;- gene_info$Symbol
rownames(gene_set_info) &lt;- gene_set_info$id
gene_set_info &lt;- gene_set_info[,-2]
F &lt;- ldf(fl_nmf_cc,type = &quot;i&quot;)$F
colnames(F) &lt;- paste0(&quot;k&quot;,1:15)
gsea_fl_nmf_cc &lt;- singlecelljamboreeR::perform_gsea(F[,&quot;k10&quot;],gene_sets,
                                                    gene_set_info,L = 15,
                                                    verbose = FALSE)
write.csv(data.frame(gene = rownames(F),signal = round(F[,&quot;k10&quot;],digits = 6)),
          &quot;lps_fl_nmf_cc_k10.csv&quot;,row.names = FALSE,quote = FALSE)
out &lt;- gsea_fl_nmf_cc$selected_gene_sets
out$top_genes &lt;- sapply(out$top_genes,function (x) paste(x,collapse = &quot; &quot;))
out$lbf  &lt;- round(out$lbf,digits = 6)
out$pip  &lt;- round(out$pip,digits = 6)
out$coef &lt;- round(out$coef,digits = 6)
write_csv(out,&quot;lps_gsea_fl_nmf_cc_k10.csv&quot;,quote = &quot;none&quot;)</code></pre>
<p>The top gene set is the IFN-<span
class="math inline">\(\alpha/\beta\)</span> signaling pathway, but other
gene sets clearly relate to inflammation and immune system function:</p>
<pre class="r"><code>print(gsea_fl_nmf_cc$selected_gene_sets[c(2:7,9)],n = Inf)
# # A tibble: 25 × 7
#    CS    gene_set   lbf    pip    coef genes name                               
#    &lt;fct&gt; &lt;chr&gt;    &lt;dbl&gt;  &lt;dbl&gt;   &lt;dbl&gt; &lt;dbl&gt; &lt;chr&gt;                              
#  1 L1    M973     327.  1       0.272     49 REACTOME_INTERFERON_ALPHA_BETA_SIG…
#  2 L2    M27605   171.  1       0.211     44 REACTOME_INTERLEUKIN_10_SIGNALING  
#  3 L3    M14405   113.  0.999   0.0777   222 GO_DEFENSE_RESPONSE_TO_VIRUS       
#  4 L4    M15265    79.7 1.000   0.0757   171 GO_RESPONSE_TO_INTERFERON_GAMMA    
#  5 L7    M18914    58.5 1       0.223     15 GO_CXCR_CHEMOKINE_RECEPTOR_BINDING 
#  6 L5    M13436    53.0 1.000   0.0609   186 GO_RESPONSE_TO_INTERLEUKIN_1       
#  7 L10   M17358    31.6 1.000   0.0539   157 GO_SPECIFIC_GRANULE                
#  8 L6    M39785    29.3 1.000  -0.146     20 WP_OVERVIEW_OF_INTERFERONSMEDIATED…
#  9 L9    M26883    28.5 1.000   0.168     15 GO_PROTEIN_ADP_RIBOSYLASE_ACTIVITY 
# 10 L12   M10974    27.7 0.997   0.0674    91 GO_NEGATIVE_REGULATION_OF_VIRAL_PR…
# 11 L14   M12167    22.6 0.994   0.0432   192 GO_PATTERN_RECOGNITION_RECEPTOR_SI…
# 12 L11   M39818    22.0 1.000   0.0366   264 WP_IL18_SIGNALING_PATHWAY          
# 13 L8    M15261    21.2 0.652   0.0316   345 GO_REGULATION_OF_INFLAMMATORY_RESP…
# 14 L8    M15877    21.2 0.346   0.0486   138 GO_POSITIVE_REGULATION_OF_INFLAMMA…
# 15 L13   M39618    19.5 0.998   0.109     27 WP_PHOTODYNAMIC_THERAPYINDUCED_UNF…
# 16 L15   M255      15.6 0.343   0.0644    65 PID_HIF1_TFPATHWAY                 
# 17 L15   M39678    15.6 0.249  -0.0559    86 WP_RETINOBLASTOMA_GENE_IN_CANCER   
# 18 L15   M12630    15.6 0.0935  0.0463   122 GO_TRANSITION_METAL_ION_HOMEOSTASIS
# 19 L15   M15481    15.6 0.0790  0.0508   100 GO_CELLULAR_TRANSITION_METAL_ION_H…
# 20 L15   M4217     15.6 0.0741 -0.0368   195 REACTOME_MITOTIC_PROMETAPHASE      
# 21 L15   M11809    15.6 0.0355  0.0916    28 GO_RESPONSE_TO_INTERFERON_BETA     
# 22 L15   M39363    15.6 0.0281  0.0843    33 WP_TYPE_II_INTERFERON_SIGNALING_IF…
# 23 L15   M11125    15.6 0.0248  0.0666    54 GO_NEGATIVE_REGULATION_OF_INNATE_I…
# 24 L15   M23341    15.6 0.0223  0.139     11 GO_CELLULAR_RESPONSE_TO_INTERFERON…
# 25 L15   M17276    15.6 0.0132  0.0565    74 GO_PRERIBOSOME</code></pre>
<p>This is the same scatterplot as the one just above, but with the
genes in the IFN-<span class="math inline">\(\alpha/\beta\)</span>
signaling pathway highlighted:</p>
<pre class="r"><code>pdat$pathway &lt;- FALSE
pathway_genes &lt;- names(which(gene_sets[,&quot;M973&quot;] &gt; 0))
pdat[pathway_genes,&quot;pathway&quot;] &lt;- TRUE
pdat &lt;- pdat[order(pdat$pathway),]
ggplot(pdat,aes(x = (tm)^(1/3),y = nmf_cc,label = gene,color = pathway)) +
  geom_point() +
  geom_text_repel(color = &quot;gray&quot;,size = 2.5,max.overlaps = Inf) +
  scale_color_manual(values = c(&quot;darkblue&quot;,&quot;orangered&quot;)) +
  labs(x = &quot;topic 7&quot;,y = &quot;factor 10&quot;) +
  theme_cowplot(font_size = 12)</code></pre>
<p><img src="figure/lps.Rmd/sepsis-tm-vs-nmf-cc-with-gsea-1.png" width="600" style="display: block; margin: auto;" /></p>
<p>Here’s another view of the enrichment of the genes in the IFN-<span
class="math inline">\(\alpha/\beta\)</span> pathway:</p>
<pre class="r"><code>p1 &lt;- ggplot(subset(pdat,pathway),aes(x = nmf_cc)) +
  geom_histogram(color = &quot;white&quot;,fill = &quot;black&quot;,bins = 16) +
  labs(x = &quot;factor 10&quot;,y = &quot;IFN-a/b genes&quot;) +
  theme_cowplot(font_size = 10)
p2 &lt;- ggplot(subset(pdat,!pathway),aes(x = nmf_cc)) +
  geom_histogram(color = &quot;white&quot;,fill = &quot;black&quot;,bins = 24) +
  labs(x = &quot;factor 10&quot;,y = &quot;other genes&quot;) +
  theme_cowplot(font_size = 10)
p3 &lt;- ggplot(subset(pdat,pathway),aes(x = log10(tm + 1e-6))) +
  geom_histogram(color = &quot;white&quot;,fill = &quot;black&quot;,bins = 16) +
  labs(x = &quot;topic 7&quot;,y = &quot;IFN-a/b genes&quot;) +
  theme_cowplot(font_size = 10)
p4 &lt;- ggplot(subset(pdat,!pathway),aes(x = log10(tm + 1e-6))) +
  geom_histogram(color = &quot;white&quot;,fill = &quot;black&quot;,bins = 24) +
  labs(x = &quot;topic 7&quot;,y = &quot;other genes&quot;) +
  theme_cowplot(font_size = 10)
plot_grid(p1,p2,p3,p4,nrow = 2,ncol = 2,rel_widths = c(2,3))</code></pre>
<p><img src="figure/lps.Rmd/sepsis-tm-vs-nmf-cc-with-gsea-2-1.png" width="480" style="display: block; margin: auto;" /></p>
<br>
<p>
<button type="button" class="btn btn-default btn-workflowr btn-workflowr-sessioninfo" data-toggle="collapse" data-target="#workflowr-sessioninfo" style="display: block;">
<span class="glyphicon glyphicon-wrench" aria-hidden="true"></span>
Session information
</button>
</p>
<div id="workflowr-sessioninfo" class="collapse">
<pre class="r"><code>sessionInfo()
# R version 4.3.3 (2024-02-29)
# Platform: aarch64-apple-darwin20 (64-bit)
# Running under: macOS 15.5
# 
# Matrix products: default
# BLAS:   /Library/Frameworks/R.framework/Versions/4.3-arm64/Resources/lib/libRblas.0.dylib 
# LAPACK: /Library/Frameworks/R.framework/Versions/4.3-arm64/Resources/lib/libRlapack.dylib;  LAPACK version 3.11.0
# 
# locale:
# [1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8
# 
# time zone: America/Chicago
# tzcode source: internal
# 
# attached base packages:
# [1] stats     graphics  grDevices utils     datasets  methods   base     
# 
# other attached packages:
#  [1] uwot_0.2.3                 rsvd_1.0.5                
#  [3] cowplot_1.1.3              ggrepel_0.9.6             
#  [5] ggplot2_3.5.2              singlecelljamboreeR_0.1-39
#  [7] pathways_0.1-20            flashier_1.0.56           
#  [9] ebnm_1.1-34                NNLM_0.4.4                
# [11] fastTopics_0.7-25          data.table_1.17.6         
# [13] readr_2.1.5                Matrix_1.6-5              
# 
# loaded via a namespace (and not attached):
#   [1] pbapply_1.7-2        rlang_1.1.6          magrittr_2.0.3      
#   [4] git2r_0.33.0         RcppAnnoy_0.0.22     horseshoe_0.2.0     
#   [7] matrixStats_1.2.0    susieR_0.14.2        compiler_4.3.3      
#  [10] vctrs_0.6.5          reshape2_1.4.4       RcppZiggurat_0.1.6  
#  [13] quadprog_1.5-8       stringr_1.5.1        pkgconfig_2.0.3     
#  [16] crayon_1.5.3         fastmap_1.2.0        labeling_0.4.3      
#  [19] utf8_1.2.6           promises_1.3.3       rmarkdown_2.29      
#  [22] tzdb_0.4.0           bit_4.0.5            purrr_1.0.4         
#  [25] Rfast_2.1.0          xfun_0.52            cachem_1.1.0        
#  [28] trust_0.1-8          jsonlite_2.0.0       progress_1.2.3      
#  [31] later_1.4.2          reshape_0.8.9        BiocParallel_1.36.0 
#  [34] irlba_2.3.5.1        parallel_4.3.3       prettyunits_1.2.0   
#  [37] R6_2.6.1             bslib_0.9.0          stringi_1.8.7       
#  [40] RColorBrewer_1.1-3   SQUAREM_2021.1       jquerylib_0.1.4     
#  [43] Rcpp_1.1.0           knitr_1.50           R.utils_2.12.3      
#  [46] httpuv_1.6.14        splines_4.3.3        tidyselect_1.2.1    
#  [49] dichromat_2.0-0.1    yaml_2.3.10          codetools_0.2-19    
#  [52] lattice_0.22-5       tibble_3.3.0         plyr_1.8.9          
#  [55] withr_3.0.2          evaluate_1.0.4       Rtsne_0.17          
#  [58] RcppParallel_5.1.10  pillar_1.11.0        whisker_0.4.1       
#  [61] plotly_4.11.0        softImpute_1.4-3     generics_0.1.4      
#  [64] vroom_1.6.5          rprojroot_2.0.4      invgamma_1.2        
#  [67] truncnorm_1.0-9      hms_1.1.3            scales_1.4.0        
#  [70] ashr_2.2-66          gtools_3.9.5         RhpcBLASctl_0.23-42 
#  [73] glue_1.8.0           scatterplot3d_0.3-44 lazyeval_0.2.2      
#  [76] tools_4.3.3          fgsea_1.35.4         fs_1.6.6            
#  [79] fastmatch_1.1-6      grid_4.3.3           tidyr_1.3.1         
#  [82] colorspace_2.1-0     deconvolveR_1.2-1    cli_3.6.5           
#  [85] Polychrome_1.5.1     workflowr_1.7.1      mixsqp_0.3-54       
#  [88] viridisLite_0.4.2    dplyr_1.1.4          gtable_0.3.6        
#  [91] R.methodsS3_1.8.2    sass_0.4.10          digest_0.6.37       
#  [94] htmlwidgets_1.6.4    farver_2.1.2         R.oo_1.26.0         
#  [97] htmltools_0.5.8.1    lifecycle_1.0.4      httr_1.4.7          
# [100] bit64_4.0.5</code></pre>
</div>
</div>


<!-- Adjust MathJax settings so that all math formulae are shown using
TeX fonts only; see
https://docs.mathjax.org/en/latest/web/configuration.html. This will make
the presentation more consistent at the cost of the webpage sometimes
taking slightly longer to load. Note that this only works because the
footer is added to webpages before the MathJax javascript. -->
<script type="text/x-mathjax-config">
  MathJax.Hub.Config({
    "HTML-CSS": { availableFonts: ["TeX"] }
  });
</script>





</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.odd').parent('tbody').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- tabsets -->

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open');
  });
});
</script>

<!-- code folding -->


<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
