<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />


<meta name="author" content="Peter Carbonetto" />


<title>NMF analysis of the LPS data set</title>

<script src="site_libs/header-attrs-2.26/header-attrs.js"></script>
<script src="site_libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/readable.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<style>h1 {font-size: 34px;}
       h1.title {font-size: 38px;}
       h2 {font-size: 30px;}
       h3 {font-size: 24px;}
       h4 {font-size: 18px;}
       h5 {font-size: 16px;}
       h6 {font-size: 12px;}
       code {color: inherit; background-color: rgba(0, 0, 0, 0.04);}
       pre:not([class]) { background-color: white }</style>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<link href="site_libs/highlightjs-9.12.0/textmate.css" rel="stylesheet" />
<script src="site_libs/highlightjs-9.12.0/highlight.js"></script>

<link rel="icon" href="https://github.com/workflowr/workflowr-assets/raw/main/img/reproducible.png">
<!-- Add a small amount of space between sections. -->
<style type="text/css">
div.section {
  padding-top: 12px;
}
</style>



<style type="text/css">
  code{white-space: pre-wrap;}
  span.smallcaps{font-variant: small-caps;}
  span.underline{text-decoration: underline;}
  div.column{display: inline-block; vertical-align: top; width: 50%;}
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  ul.task-list{list-style: none;}
    </style>

<style type="text/css">code{white-space: pre;}</style>
<script type="text/javascript">
if (window.hljs) {
  hljs.configure({languages: []});
  hljs.initHighlightingOnLoad();
  if (document.readyState && document.readyState === "complete") {
    window.setTimeout(function() { hljs.initHighlighting(); }, 0);
  }
}
</script>









<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
img {
  max-width:100%;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
details > summary > p:only-child {
  display: inline;
}
pre code {
  padding: 0;
}
</style>


<style type="text/css">
.dropdown-submenu {
  position: relative;
}
.dropdown-submenu>.dropdown-menu {
  top: 0;
  left: 100%;
  margin-top: -6px;
  margin-left: -1px;
  border-radius: 0 6px 6px 6px;
}
.dropdown-submenu:hover>.dropdown-menu {
  display: block;
}
.dropdown-submenu>a:after {
  display: block;
  content: " ";
  float: right;
  width: 0;
  height: 0;
  border-color: transparent;
  border-style: solid;
  border-width: 5px 0 5px 5px;
  border-left-color: #cccccc;
  margin-top: 5px;
  margin-right: -10px;
}
.dropdown-submenu:hover>a:after {
  border-left-color: #adb5bd;
}
.dropdown-submenu.pull-left {
  float: none;
}
.dropdown-submenu.pull-left>.dropdown-menu {
  left: -100%;
  margin-left: 10px;
  border-radius: 6px 0 6px 6px;
}
</style>

<script type="text/javascript">
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark the anchor link active (and if it's in a dropdown, also mark that active)
  var dropdown = menuAnchor.closest('li.dropdown');
  if (window.bootstrap) { // Bootstrap 4+
    menuAnchor.addClass('active');
    dropdown.find('> .dropdown-toggle').addClass('active');
  } else { // Bootstrap 3
    menuAnchor.parent().addClass('active');
    dropdown.addClass('active');
  }

  // Navbar adjustments
  var navHeight = $(".navbar").first().height() + 15;
  var style = document.createElement('style');
  var pt = "padding-top: " + navHeight + "px; ";
  var mt = "margin-top: -" + navHeight + "px; ";
  var css = "";
  // offset scroll position for anchor links (for fixed navbar)
  for (var i = 1; i <= 6; i++) {
    css += ".section h" + i + "{ " + pt + mt + "}\n";
  }
  style.innerHTML = "body {" + pt + "padding-bottom: 40px; }\n" + css;
  document.head.appendChild(style);
});
</script>

<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before, .tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "\e259";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "\e258";
  font-family: 'Glyphicons Halflings';
  border: none;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
  background-color: transparent;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<!-- code folding -->




</head>

<body>


<div class="container-fluid main-container">




<div class="navbar navbar-default  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-bs-toggle="collapse" data-target="#navbar" data-bs-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html">single-cell jamboree</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="index.html">Overview</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/stephenslab/single-cell-jamboree">source</a>
</li>
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<div id="header">



<h1 class="title toc-ignore">NMF analysis of the LPS data set</h1>
<h4 class="author">Peter Carbonetto</h4>

</div>


<p>
<button type="button" class="btn btn-default btn-workflowr btn-workflowr-report" data-toggle="collapse" data-target="#workflowr-report">
<span class="glyphicon glyphicon-list" aria-hidden="true"></span>
workflowr <span class="glyphicon glyphicon-exclamation-sign text-danger"
aria-hidden="true"></span>
</button>
</p>
<div id="workflowr-report" class="collapse">
<ul class="nav nav-tabs">
<li class="active">
<a data-toggle="tab" href="#summary">Summary</a>
</li>
<li>
<a data-toggle="tab" href="#checks"> Checks <span
class="glyphicon glyphicon-exclamation-sign text-danger"
aria-hidden="true"></span> </a>
</li>
<li>
<a data-toggle="tab" href="#versions">Past versions</a>
</li>
</ul>
<div class="tab-content">
<div id="summary" class="tab-pane fade in active">
<p>
<strong>Last updated:</strong> 2025-06-06
</p>
<p>
<strong>Checks:</strong> <span
class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> 6
<span class="glyphicon glyphicon-exclamation-sign text-danger"
aria-hidden="true"></span> 1
</p>
<p>
<strong>Knit directory:</strong>
<code>single-cell-jamboree/analysis/</code> <span
class="glyphicon glyphicon-question-sign" aria-hidden="true"
title="This is the local directory in which the code in this file was executed.">
</span>
</p>
<p>
This reproducible <a href="https://rmarkdown.rstudio.com">R Markdown</a>
analysis was created with <a
  href="https://github.com/workflowr/workflowr">workflowr</a> (version
1.7.1). The <em>Checks</em> tab describes the reproducibility checks
that were applied when the results were created. The <em>Past
versions</em> tab lists the development history.
</p>
<hr>
</div>
<div id="checks" class="tab-pane fade">
<div id="workflowr-checks" class="panel-group">
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongRMarkdownfilestronguptodate">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>R Markdown file:</strong> up-to-date
</a>
</p>
</div>
<div id="strongRMarkdownfilestronguptodate"
class="panel-collapse collapse">
<div class="panel-body">
<p>Great! Since the R Markdown file has been committed to the Git
repository, you know the exact version of the code that produced these
results.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongEnvironmentstrongempty">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Environment:</strong> empty </a>
</p>
</div>
<div id="strongEnvironmentstrongempty" class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! The global environment was empty. Objects defined in the
global environment can affect the analysis in your R Markdown file in
unknown ways. For reproduciblity it’s best to always run the code in an
empty environment.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongSeedstrongcodesetseed1code">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Seed:</strong>
<code>set.seed(1)</code> </a>
</p>
</div>
<div id="strongSeedstrongcodesetseed1code"
class="panel-collapse collapse">
<div class="panel-body">
<p>The command <code>set.seed(1)</code> was run prior to running the
code in the R Markdown file. Setting a seed ensures that any results
that rely on randomness, e.g. subsampling or permutations, are
reproducible.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongSessioninformationstrongrecorded">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Session information:</strong>
recorded </a>
</p>
</div>
<div id="strongSessioninformationstrongrecorded"
class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! Recording the operating system, R version, and package
versions is critical for reproducibility.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongCachestrongdetected">
<span class="glyphicon glyphicon-exclamation-sign text-danger"
aria-hidden="true"></span> <strong>Cache:</strong> detected </a>
</p>
</div>
<div id="strongCachestrongdetected" class="panel-collapse collapse">
<div class="panel-body">
The following chunks had caches available:
<ul>
<li>
fit-topic-model
</li>
<li>
flashier-nmf
</li>
</ul>
<p>To ensure reproducibility of the results, delete the cache directory
<code>lps_cache</code> and re-run the analysis. To have workflowr
automatically delete the cache directory prior to building the file, set
<code>delete_cache = TRUE</code> when running <code>wflow_build()</code>
or <code>wflow_publish()</code>.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongFilepathsstrongrelative">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>File paths:</strong> relative </a>
</p>
</div>
<div id="strongFilepathsstrongrelative" class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! Using relative paths to the files within your workflowr
project makes it easier to run your code on other machines.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongRepositoryversionstrongahrefhttpsgithubcomstephenslabsinglecelljamboreetreef38b5868f566a5efd73d58eb9bf1f6f05b52cb8ftargetblankf38b586a">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Repository version:</strong>
<a href="https://github.com/stephenslab/single-cell-jamboree/tree/f38b5868f566a5efd73d58eb9bf1f6f05b52cb8f" target="_blank">f38b586</a>
</a>
</p>
</div>
<div
id="strongRepositoryversionstrongahrefhttpsgithubcomstephenslabsinglecelljamboreetreef38b5868f566a5efd73d58eb9bf1f6f05b52cb8ftargetblankf38b586a"
class="panel-collapse collapse">
<div class="panel-body">
<p>
Great! You are using Git for version control. Tracking code development
and connecting the code version to the results is critical for
reproducibility.
</p>
<p>
The results in this page were generated with repository version
<a href="https://github.com/stephenslab/single-cell-jamboree/tree/f38b5868f566a5efd73d58eb9bf1f6f05b52cb8f" target="_blank">f38b586</a>.
See the <em>Past versions</em> tab to see a history of the changes made
to the R Markdown and HTML files.
</p>
<p>
Note that you need to be careful to ensure that all relevant files for
the analysis have been committed to Git prior to generating the results
(you can use <code>wflow_publish</code> or
<code>wflow_git_commit</code>). workflowr only checks the R Markdown
file, but you know if there are other scripts or data files that it
depends on. Below is the status of the Git repository when the results
were generated:
</p>
<pre><code>
Untracked files:
    Untracked:  analysis/lps_cache/
    Untracked:  analysis/mcf7_cache/
    Untracked:  analysis/temp2.R
    Untracked:  data/GSE132188_adata.h5ad.h5
    Untracked:  data/GSE183010/
    Untracked:  data/Immune_ALL_human.h5ad
    Untracked:  data/pancreas_cytokine.RData
    Untracked:  data/pancreas_endocrine.RData
    Untracked:  data/pancreas_endocrine_alldays.h5ad

</code></pre>
<p>
Note that any generated files, e.g. HTML, png, CSS, etc., are not
included in this status report because it is ok for generated content to
have uncommitted changes.
</p>
</div>
</div>
</div>
</div>
<hr>
</div>
<div id="versions" class="tab-pane fade">

<p>
These are the previous versions of the repository in which changes were
made to the R Markdown (<code>analysis/lps.Rmd</code>) and HTML
(<code>docs/lps.html</code>) files. If you’ve configured a remote Git
repository (see <code>?wflow_git_remote</code>), click on the hyperlinks
in the table below to view the files as they were in that past version.
</p>
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
File
</th>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
<th>
Message
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/stephenslab/single-cell-jamboree/blob/f38b5868f566a5efd73d58eb9bf1f6f05b52cb8f/analysis/lps.Rmd" target="_blank">f38b586</a>
</td>
<td>
Peter Carbonetto
</td>
<td>
2025-06-06
</td>
<td>
A small fix to the lps analysis.
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/stephenslab/single-cell-jamboree/blob/aae4257ccd0323c365412c4a0ae9f104b30e6387/analysis/lps.Rmd" target="_blank">aae4257</a>
</td>
<td>
Peter Carbonetto
</td>
<td>
2025-06-06
</td>
<td>
A couple fixes to the lps analysis.
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/stephenslab/single-cell-jamboree/blob/6cbad5f1203de7ea08fa9769b49440a2f6138009/analysis/lps.Rmd" target="_blank">6cbad5f</a>
</td>
<td>
Peter Carbonetto
</td>
<td>
2025-06-06
</td>
<td>
Added a structure plot to the lps analysis.
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/stephenslab/single-cell-jamboree/blob/90d6c063f96aa1805b7ab6b456dc05eff2fd9626/analysis/lps.Rmd" target="_blank">90d6c06</a>
</td>
<td>
Peter Carbonetto
</td>
<td>
2025-06-06
</td>
<td>
Improved the structure plots in the lps analysis.
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/stephenslab/single-cell-jamboree/blob/dac95b566c5f66b67e3f46f8e4a1e50f4acf708d/analysis/lps.Rmd" target="_blank">dac95b5</a>
</td>
<td>
Peter Carbonetto
</td>
<td>
2025-06-06
</td>
<td>
Made a few changes to the flashier fit in the lps analysis.
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/stephenslab/single-cell-jamboree/blob/9e1f1279beceee0612c9ca09ce3de52ba22c38f8/analysis/lps.Rmd" target="_blank">9e1f127</a>
</td>
<td>
Peter Carbonetto
</td>
<td>
2025-06-05
</td>
<td>
Added code to pancreas_cytokine analysis to prepare the scrna-seq data
downloaded from geo.
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/stephenslab/single-cell-jamboree/blob/8d945a1ebcb216415939c8bb8811803ae9c79bbe/analysis/lps.Rmd" target="_blank">8d945a1</a>
</td>
<td>
Peter Carbonetto
</td>
<td>
2025-06-04
</td>
<td>
Added flashier fit to lps analysis; need to revise this and the topic
modeling result.
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/stephenslab/single-cell-jamboree/blob/dac6198b50dfb95b95f7ebb1f28227ad3d2f21fa/analysis/lps.Rmd" target="_blank">dac6198</a>
</td>
<td>
Peter Carbonetto
</td>
<td>
2025-06-04
</td>
<td>
Working on topic modeling results for lps data.
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/stephenslab/single-cell-jamboree/blob/8f39607ebba82349cd62736373fc6886c917ec5f/analysis/lps.Rmd" target="_blank">8f39607</a>
</td>
<td>
Peter Carbonetto
</td>
<td>
2025-06-04
</td>
<td>
Added steps to the lps analysis to load and prepare the data.
</td>
</tr>
<tr>
<td>
html
</td>
<td>
<a href="https://rawcdn.githack.com/stephenslab/single-cell-jamboree/2bfef0b5b759c8eacda0d00b99a9193de45ba241/docs/lps.html" target="_blank">2bfef0b</a>
</td>
<td>
Peter Carbonetto
</td>
<td>
2025-06-04
</td>
<td>
First build of the LPS analysis.
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/stephenslab/single-cell-jamboree/blob/85adf3f294176997b859f1e2351d672ce413af11/analysis/lps.Rmd" target="_blank">85adf3f</a>
</td>
<td>
Peter Carbonetto
</td>
<td>
2025-06-04
</td>
<td>
wflow_publish("lps.Rmd")
</td>
</tr>
</tbody>
</table>
</div>
<hr>
</div>
</div>
</div>
<p>Here we will revisit the LPS data set that we analyzed using a topic
model in the <a
href="https://doi.org/10.1038/s41590-023-01722-8">Takahama et al Nat
Immunol paper</a>. (LPS = lipopolysaccharide).</p>
<p>Load packages used to process the data, perform the analyses, and
create the plots.</p>
<pre class="r"><code>library(data.table)
library(fastTopics)
library(NNLM)
library(ebnm)
library(flashier)
library(singlecelljamboreeR)
library(ggplot2)
library(cowplot)</code></pre>
<p>Set the seed for reproducibility:</p>
<pre class="r"><code>set.seed(1)</code></pre>
<div id="prepare-the-data-for-analysis-with-fasttopics-and-flashier"
class="section level2">
<h2>Prepare the data for analysis with fastTopics and flashier</h2>
<p>Load the RNA-seq counts:</p>
<pre class="r"><code>read_lps_data &lt;- function (file) {
  counts &lt;- fread(file)
  class(counts) &lt;- &quot;data.frame&quot;
  genes &lt;- counts[,1]
  counts &lt;- t(as.matrix(counts[,-1]))
  colnames(counts) &lt;- genes
  samples &lt;- rownames(counts)
  samples &lt;- strsplit(samples,&quot;_&quot;)
  samples &lt;- data.frame(tissue    = sapply(samples,&quot;[[&quot;,1),
                        timepoint = sapply(samples,&quot;[[&quot;,2),
                        mouse     = sapply(samples,&quot;[[&quot;,3))
  samples &lt;- transform(samples,
                       tissue    = factor(tissue),
                       timepoint = factor(timepoint),
                       mouse     = factor(mouse))
  return(list(samples = samples,counts = counts))
}
out &lt;- read_lps_data(&quot;../data/lps.csv.gz&quot;)
samples &lt;- out$samples
counts  &lt;- out$counts
rm(out)</code></pre>
<p>Remove a sample that appears to be an outlier based on the NMF
analyses:</p>
<pre class="r"><code>i       &lt;- which(rownames(counts) != &quot;iLN_d2_20&quot;)
samples &lt;- samples[i,]
counts  &lt;- counts[i,]</code></pre>
<p>Remove genes that are expressed in fewer than 5 samples:</p>
<pre class="r"><code>j &lt;- which(colSums(counts &gt; 0) &gt; 4)
counts &lt;- counts[,j]</code></pre>
<p>This is the dimension of the data set we will analyze:</p>
<pre class="r"><code>dim(counts)
# [1]   363 33533</code></pre>
<p>For the Gaussian-based analyses, we will need the shifted log
counts:</p>
<pre class="r"><code>a &lt;- 1
s &lt;- rowSums(counts)
s &lt;- s/mean(s)
shifted_log_counts &lt;- log1p(counts/(a*s))</code></pre>
</div>
<div id="topic-model-fasttopics" class="section level2">
<h2>Topic model (fastTopics)</h2>
<p>Fit a topic model with <span class="math inline">\(K = 14\)</span>
topics to the counts:</p>
<pre class="r"><code>tm &lt;- fit_poisson_nmf(counts,k = 14,init.method = &quot;random&quot;,method = &quot;em&quot;,
                      numiter = 20,verbose = &quot;none&quot;,
                      control = list(numiter = 4,nc = 8,extrapolate = FALSE))
tm &lt;- fit_poisson_nmf(counts,fit0 = tm,method = &quot;scd&quot;,numiter = 40,
                      control = list(numiter = 4,nc = 8,extrapolate = TRUE),
                      verbose = &quot;none&quot;)</code></pre>
<div class="alert alert-warning">
<p>
<strong>Warning:</strong> The above code chunk cached its results, but
it won’t be re-run if previous chunks it depends on are updated. If you
need to use caching, it is highly recommended to also set
<code>knitr::opts_chunk$set(autodep = TRUE)</code> at the top of the
file (in a chunk that is not cached). Alternatively, you can customize
the option <code>dependson</code> for each individual chunk that is
cached. Using either <code>autodep</code> or <code>dependson</code> will
remove this warning. See the
<a href="https://yihui.name/knitr/options/#cache"
>knitr cache options</a> for more details.
</p>
</div>
<p>Structure plot comparing the topics to the organ types:</p>
<pre class="r"><code>rows &lt;- order(samples$timepoint)
topic_colors &lt;- c(&quot;magenta&quot;,&quot;darkorange&quot;,&quot;darkblue&quot;,&quot;forestgreen&quot;,
                  &quot;dodgerblue&quot;,&quot;gray&quot;,&quot;red&quot;,&quot;olivedrab&quot;,&quot;darkmagenta&quot;,
                  &quot;sienna&quot;,&quot;limegreen&quot;,&quot;royalblue&quot;,&quot;lightskyblue&quot;,
                  &quot;gold&quot;)
samples &lt;- transform(samples,
  tissue = factor(tissue,c(&quot;PBMC&quot;,&quot;BM&quot;,&quot;LU&quot;,&quot;CO&quot;,&quot;SI&quot;,&quot;iLN&quot;,&quot;SP&quot;,
                           &quot;TH&quot;,&quot;SK&quot;,&quot;KI&quot;,&quot;LI&quot;,&quot;BR&quot;,&quot;HE&quot;)))
structure_plot(tm,grouping = samples$tissue,gap = 4,
               topics = 1:14,colors = topic_colors,
               loadings_order = rows) +
  labs(fill = &quot;&quot;) +
  theme(axis.text.x = element_text(angle = 0,hjust = 0.5),
        legend.key.height = unit(0.01,&quot;cm&quot;),
        legend.key.width = unit(0.2,&quot;cm&quot;),
        legend.text = element_text(size = 6))</code></pre>
<p><img src="figure/lps.Rmd/structure-plot-topic-model-1.png" width="840" style="display: block; margin: auto;" /></p>
<p>Abbreviations used: BM = bone marrow; BR = brain; CO = colon; HE =
heart; iLN = inguinal lymph node; KI = kidney; LI = liver; LU = lung; SI
= small intestine; SK = skin; SP = spleen; TH = thymus.</p>
<p>This next structure plot better highlights the topics that capture
the processes driven by LPS-induced sepsis:</p>
<pre class="r"><code>topic_colors &lt;- c(&quot;magenta&quot;,&quot;gray50&quot;,&quot;gray65&quot;,&quot;gray40&quot;,
                  &quot;gray85&quot;,&quot;gray75&quot;,&quot;red&quot;,&quot;gray80&quot;,&quot;gray90&quot;,
                  &quot;gray60&quot;,&quot;limegreen&quot;,&quot;gray70&quot;,&quot;gray55&quot;,
                  &quot;gold&quot;)
structure_plot(tm,grouping = samples$tissue,gap = 4,
               topics = 1:14,colors = topic_colors,
               loadings_order = rows) +
  labs(fill = &quot;&quot;) +
  theme(axis.text.x = element_text(angle = 0,hjust = 0.5),
        legend.key.height = unit(0.01,&quot;cm&quot;),
        legend.key.width = unit(0.2,&quot;cm&quot;),
        legend.text = element_text(size = 6))</code></pre>
<p><img src="figure/lps.Rmd/structure-plot-topic-model-2-1.png" width="840" style="display: block; margin: auto;" /></p>
</div>
<div id="ebnmf-flashier" class="section level2">
<h2>EBNMF (flashier)</h2>
<p>Next fit an NMF to the shifted log counts using flashier, with <span
class="math inline">\(K =
15\)</span>:</p>
<pre class="r"><code>k &lt;- 15
n &lt;- nrow(shifted_log_counts)
m &lt;- ncol(shifted_log_counts)
nmf0 &lt;- nnmf(shifted_log_counts,k = 1,loss = &quot;mse&quot;,method = &quot;scd&quot;,
            max.iter = 10,verbose = 0,n.threads = 4)
W0 &lt;- nmf0$W
H0 &lt;- nmf0$H
W0 &lt;- cbind(W0,matrix(runif(n*(k-1)),n,k-1))
H0 &lt;- rbind(H0,matrix(runif(m*(k-1)),k-1,m))
nmf &lt;- nnmf(shifted_log_counts,k,init = list(W = W0,H = H0),
            loss = &quot;mse&quot;,method = &quot;scd&quot;,max.iter = 10,
            verbose = 0,n.threads = 8)
x  &lt;- rpois(1e7,1/n)
s1 &lt;- sd(log(x + 1))
sparse_prior &lt;- ebnm_point_exponential(x = c(rep(1,100)))
sparse_prior$fitted_g$pi &lt;- c(0.99,0.01)
ebnm_sparse_prior &lt;- flash_ebnm(prior_family = &quot;point_exponential&quot;,
                                fix_g = TRUE,g_init = sparse_prior)
fl_nmf &lt;- flash_init(shifted_log_counts,var_type = 2,S = s1)
fl_nmf &lt;- flash_factors_init(fl_nmf,list(nmf$W,t(nmf$H)),
                             c(ebnm_sparse_prior,ebnm_point_exponential))
fl_nmf &lt;- flash_backfit(fl_nmf,extrapolate = FALSE,maxiter = 100,verbose = 0)
fl_nmf &lt;- flash_backfit(fl_nmf,extrapolate = TRUE,maxiter = 100,verbose = 0)</code></pre>
<div class="alert alert-warning">
<p>
<strong>Warning:</strong> The above code chunk cached its results, but
it won’t be re-run if previous chunks it depends on are updated. If you
need to use caching, it is highly recommended to also set
<code>knitr::opts_chunk$set(autodep = TRUE)</code> at the top of the
file (in a chunk that is not cached). Alternatively, you can customize
the option <code>dependson</code> for each individual chunk that is
cached. Using either <code>autodep</code> or <code>dependson</code> will
remove this warning. See the
<a href="https://yihui.name/knitr/options/#cache"
>knitr cache options</a> for more details.
</p>
</div>
<p>Structure plot comparing the factors to the organ types:</p>
<pre class="r"><code>rows &lt;- order(samples$timepoint)
topic_colors &lt;- c(&quot;powderblue&quot;,&quot;dodgerblue&quot;,&quot;olivedrab&quot;,&quot;limegreen&quot;,
                  &quot;forestgreen&quot;,&quot;red&quot;,&quot;darkmagenta&quot;,&quot;gray&quot;,&quot;darkorange&quot;,
                  &quot;cyan&quot;,&quot;royalblue&quot;,&quot;darkblue&quot;,&quot;lightskyblue&quot;,
                  &quot;gold&quot;,&quot;sienna&quot;)
L &lt;- ldf(fl_nmf,type = &quot;i&quot;)$L
structure_plot(L,grouping = samples$tissue,gap = 4,
                    topics = 1:15,colors = topic_colors,
                    loadings_order = rows) +
  labs(fill = &quot;&quot;,y = &quot;membership&quot;) +
  theme(axis.text.x = element_text(angle = 0,hjust = 0.5),
        legend.key.height = unit(0.01,&quot;cm&quot;),
        legend.key.width = unit(0.25,&quot;cm&quot;),
        legend.text = element_text(size = 7))</code></pre>
<p><img src="figure/lps.Rmd/structure-plot-flashier-nmf-1.png" width="840" style="display: block; margin: auto;" /></p>
<p>This next structure plot better highlights the topics that capture
the processes driven by LPS-induced sepsis:</p>
<pre class="r"><code>rows &lt;- order(samples$timepoint)
topic_colors &lt;- c(&quot;gray95&quot;,&quot;gray70&quot;,&quot;gray80&quot;,&quot;limegreen&quot;,
                  &quot;gray60&quot;,&quot;red&quot;,&quot;gray75&quot;,&quot;gray&quot;,&quot;gray85&quot;,
                  &quot;gray90&quot;,&quot;gray65&quot;,&quot;gray50&quot;,&quot;gray45&quot;,
                  &quot;gray35&quot;,&quot;gray75&quot;)
L &lt;- ldf(fl_nmf,type = &quot;i&quot;)$L
structure_plot(L,grouping = samples$tissue,gap = 4,
                    topics = 1:15,colors = topic_colors,
                    loadings_order = rows) +
  labs(fill = &quot;&quot;,y = &quot;membership&quot;) +
  theme(axis.text.x = element_text(angle = 0,hjust = 0.5),
        legend.key.height = unit(0.01,&quot;cm&quot;),
        legend.key.width = unit(0.25,&quot;cm&quot;),
        legend.text = element_text(size = 7))</code></pre>
<p><img src="figure/lps.Rmd/structure-plot-flashier-nmf-2-1.png" width="840" style="display: block; margin: auto;" /></p>
<br>
<p>
<button type="button" class="btn btn-default btn-workflowr btn-workflowr-sessioninfo" data-toggle="collapse" data-target="#workflowr-sessioninfo" style="display: block;">
<span class="glyphicon glyphicon-wrench" aria-hidden="true"></span>
Session information
</button>
</p>
<div id="workflowr-sessioninfo" class="collapse">
<pre class="r"><code>sessionInfo()
# R version 4.3.3 (2024-02-29)
# Platform: aarch64-apple-darwin20 (64-bit)
# Running under: macOS 15.4.1
# 
# Matrix products: default
# BLAS:   /Library/Frameworks/R.framework/Versions/4.3-arm64/Resources/lib/libRblas.0.dylib 
# LAPACK: /Library/Frameworks/R.framework/Versions/4.3-arm64/Resources/lib/libRlapack.dylib;  LAPACK version 3.11.0
# 
# locale:
# [1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8
# 
# time zone: America/Chicago
# tzcode source: internal
# 
# attached base packages:
# [1] stats     graphics  grDevices utils     datasets  methods   base     
# 
# other attached packages:
# [1] workflowr_1.7.1           cowplot_1.1.3            
# [3] ggplot2_3.5.0             singlecelljamboreeR_0.1-3
# [5] flashier_1.0.55           ebnm_1.1-34              
# [7] NNLM_0.4.4                fastTopics_0.7-25        
# [9] data.table_1.17.4        
# 
# loaded via a namespace (and not attached):
#  [1] pbapply_1.7-2        rlang_1.1.5          magrittr_2.0.3      
#  [4] git2r_0.33.0         horseshoe_0.2.0      compiler_4.3.3      
#  [7] getPass_0.2-4        callr_3.7.5          vctrs_0.6.5         
# [10] reshape2_1.4.4       quadprog_1.5-8       stringr_1.5.1       
# [13] pkgconfig_2.0.3      crayon_1.5.2         fastmap_1.1.1       
# [16] labeling_0.4.3       utf8_1.2.4           promises_1.2.1      
# [19] rmarkdown_2.26       ps_1.7.6             purrr_1.0.2         
# [22] xfun_0.42            cachem_1.0.8         trust_0.1-8         
# [25] jsonlite_1.8.8       progress_1.2.3       highr_0.10          
# [28] later_1.3.2          irlba_2.3.5.1        parallel_4.3.3      
# [31] prettyunits_1.2.0    R6_2.5.1             bslib_0.6.1         
# [34] stringi_1.8.3        RColorBrewer_1.1-3   SQUAREM_2021.1      
# [37] jquerylib_0.1.4      Rcpp_1.0.12          knitr_1.45          
# [40] R.utils_2.12.3       httpuv_1.6.14        Matrix_1.6-5        
# [43] splines_4.3.3        tidyselect_1.2.1     rstudioapi_0.15.0   
# [46] yaml_2.3.8           codetools_0.2-19     processx_3.8.3      
# [49] lattice_0.22-5       tibble_3.2.1         plyr_1.8.9          
# [52] withr_3.0.2          evaluate_1.0.3       Rtsne_0.17          
# [55] RcppParallel_5.1.10  pillar_1.9.0         whisker_0.4.1       
# [58] plotly_4.10.4        softImpute_1.4-1     generics_0.1.3      
# [61] rprojroot_2.0.4      invgamma_1.1         truncnorm_1.0-9     
# [64] hms_1.1.3            munsell_0.5.0        scales_1.3.0        
# [67] ashr_2.2-66          gtools_3.9.5         RhpcBLASctl_0.23-42 
# [70] glue_1.8.0           scatterplot3d_0.3-44 lazyeval_0.2.2      
# [73] tools_4.3.3          fs_1.6.5             grid_4.3.3          
# [76] tidyr_1.3.1          colorspace_2.1-0     deconvolveR_1.2-1   
# [79] cli_3.6.4            Polychrome_1.5.1     fansi_1.0.6         
# [82] mixsqp_0.3-54        viridisLite_0.4.2    dplyr_1.1.4         
# [85] uwot_0.2.3           gtable_0.3.4         R.methodsS3_1.8.2   
# [88] sass_0.4.9           digest_0.6.34        ggrepel_0.9.5       
# [91] htmlwidgets_1.6.4    farver_2.1.1         htmltools_0.5.8.1   
# [94] R.oo_1.26.0          lifecycle_1.0.4      httr_1.4.7</code></pre>
</div>
</div>


<!-- Adjust MathJax settings so that all math formulae are shown using
TeX fonts only; see
https://docs.mathjax.org/en/latest/web/configuration.html. This will make
the presentation more consistent at the cost of the webpage sometimes
taking slightly longer to load. Note that this only works because the
footer is added to webpages before the MathJax javascript. -->
<script type="text/x-mathjax-config">
  MathJax.Hub.Config({
    "HTML-CSS": { availableFonts: ["TeX"] }
  });
</script>





</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.odd').parent('tbody').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- tabsets -->

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open');
  });
});
</script>

<!-- code folding -->


<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
