<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />


<meta name="author" content="Peter Carbonetto" />


<title>NMF analysis of the MCF-7 data set</title>

<script src="site_libs/header-attrs-2.29/header-attrs.js"></script>
<script src="site_libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/readable.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<style>h1 {font-size: 34px;}
       h1.title {font-size: 38px;}
       h2 {font-size: 30px;}
       h3 {font-size: 24px;}
       h4 {font-size: 18px;}
       h5 {font-size: 16px;}
       h6 {font-size: 12px;}
       code {color: inherit; background-color: rgba(0, 0, 0, 0.04);}
       pre:not([class]) { background-color: white }</style>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<link href="site_libs/highlightjs-9.12.0/textmate.css" rel="stylesheet" />
<script src="site_libs/highlightjs-9.12.0/highlight.js"></script>

<link rel="icon" href="https://github.com/workflowr/workflowr-assets/raw/main/img/reproducible.png">
<!-- Add a small amount of space between sections. -->
<style type="text/css">
div.section {
  padding-top: 12px;
}
</style>



<style type="text/css">
  code{white-space: pre-wrap;}
  span.smallcaps{font-variant: small-caps;}
  span.underline{text-decoration: underline;}
  div.column{display: inline-block; vertical-align: top; width: 50%;}
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  ul.task-list{list-style: none;}
    </style>

<style type="text/css">code{white-space: pre;}</style>
<script type="text/javascript">
if (window.hljs) {
  hljs.configure({languages: []});
  hljs.initHighlightingOnLoad();
  if (document.readyState && document.readyState === "complete") {
    window.setTimeout(function() { hljs.initHighlighting(); }, 0);
  }
}
</script>









<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
img {
  max-width:100%;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
details > summary > p:only-child {
  display: inline;
}
pre code {
  padding: 0;
}
</style>


<style type="text/css">
.dropdown-submenu {
  position: relative;
}
.dropdown-submenu>.dropdown-menu {
  top: 0;
  left: 100%;
  margin-top: -6px;
  margin-left: -1px;
  border-radius: 0 6px 6px 6px;
}
.dropdown-submenu:hover>.dropdown-menu {
  display: block;
}
.dropdown-submenu>a:after {
  display: block;
  content: " ";
  float: right;
  width: 0;
  height: 0;
  border-color: transparent;
  border-style: solid;
  border-width: 5px 0 5px 5px;
  border-left-color: #cccccc;
  margin-top: 5px;
  margin-right: -10px;
}
.dropdown-submenu:hover>a:after {
  border-left-color: #adb5bd;
}
.dropdown-submenu.pull-left {
  float: none;
}
.dropdown-submenu.pull-left>.dropdown-menu {
  left: -100%;
  margin-left: 10px;
  border-radius: 6px 0 6px 6px;
}
</style>

<script type="text/javascript">
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark the anchor link active (and if it's in a dropdown, also mark that active)
  var dropdown = menuAnchor.closest('li.dropdown');
  if (window.bootstrap) { // Bootstrap 4+
    menuAnchor.addClass('active');
    dropdown.find('> .dropdown-toggle').addClass('active');
  } else { // Bootstrap 3
    menuAnchor.parent().addClass('active');
    dropdown.addClass('active');
  }

  // Navbar adjustments
  var navHeight = $(".navbar").first().height() + 15;
  var style = document.createElement('style');
  var pt = "padding-top: " + navHeight + "px; ";
  var mt = "margin-top: -" + navHeight + "px; ";
  var css = "";
  // offset scroll position for anchor links (for fixed navbar)
  for (var i = 1; i <= 6; i++) {
    css += ".section h" + i + "{ " + pt + mt + "}\n";
  }
  style.innerHTML = "body {" + pt + "padding-bottom: 40px; }\n" + css;
  document.head.appendChild(style);
});
</script>

<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before, .tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "\e259";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "\e258";
  font-family: 'Glyphicons Halflings';
  border: none;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
  background-color: transparent;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<!-- code folding -->




</head>

<body>


<div class="container-fluid main-container">




<div class="navbar navbar-default  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-bs-toggle="collapse" data-target="#navbar" data-bs-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html">single-cell jamboree</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="index.html">Overview</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/stephenslab/single-cell-jamboree">source</a>
</li>
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<div id="header">



<h1 class="title toc-ignore">NMF analysis of the MCF-7 data set</h1>
<h4 class="author">Peter Carbonetto</h4>

</div>


<p>
<button type="button" class="btn btn-default btn-workflowr btn-workflowr-report" data-toggle="collapse" data-target="#workflowr-report">
<span class="glyphicon glyphicon-list" aria-hidden="true"></span>
workflowr <span class="glyphicon glyphicon-exclamation-sign text-danger"
aria-hidden="true"></span>
</button>
</p>
<div id="workflowr-report" class="collapse">
<ul class="nav nav-tabs">
<li class="active">
<a data-toggle="tab" href="#summary">Summary</a>
</li>
<li>
<a data-toggle="tab" href="#checks"> Checks <span
class="glyphicon glyphicon-exclamation-sign text-danger"
aria-hidden="true"></span> </a>
</li>
<li>
<a data-toggle="tab" href="#versions">Past versions</a>
</li>
</ul>
<div class="tab-content">
<div id="summary" class="tab-pane fade in active">
<p>
<strong>Last updated:</strong> 2025-08-12
</p>
<p>
<strong>Checks:</strong> <span
class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> 6
<span class="glyphicon glyphicon-exclamation-sign text-danger"
aria-hidden="true"></span> 1
</p>
<p>
<strong>Knit directory:</strong>
<code>single-cell-jamboree/analysis/</code> <span
class="glyphicon glyphicon-question-sign" aria-hidden="true"
title="This is the local directory in which the code in this file was executed.">
</span>
</p>
<p>
This reproducible <a href="https://rmarkdown.rstudio.com">R Markdown</a>
analysis was created with <a
  href="https://github.com/workflowr/workflowr">workflowr</a> (version
1.7.1). The <em>Checks</em> tab describes the reproducibility checks
that were applied when the results were created. The <em>Past
versions</em> tab lists the development history.
</p>
<hr>
</div>
<div id="checks" class="tab-pane fade">
<div id="workflowr-checks" class="panel-group">
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongRMarkdownfilestronguptodate">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>R Markdown file:</strong> up-to-date
</a>
</p>
</div>
<div id="strongRMarkdownfilestronguptodate"
class="panel-collapse collapse">
<div class="panel-body">
<p>Great! Since the R Markdown file has been committed to the Git
repository, you know the exact version of the code that produced these
results.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongEnvironmentstrongempty">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Environment:</strong> empty </a>
</p>
</div>
<div id="strongEnvironmentstrongempty" class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! The global environment was empty. Objects defined in the
global environment can affect the analysis in your R Markdown file in
unknown ways. For reproduciblity it’s best to always run the code in an
empty environment.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongSeedstrongcodesetseed1code">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Seed:</strong>
<code>set.seed(1)</code> </a>
</p>
</div>
<div id="strongSeedstrongcodesetseed1code"
class="panel-collapse collapse">
<div class="panel-body">
<p>The command <code>set.seed(1)</code> was run prior to running the
code in the R Markdown file. Setting a seed ensures that any results
that rely on randomness, e.g. subsampling or permutations, are
reproducible.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongSessioninformationstrongrecorded">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Session information:</strong>
recorded </a>
</p>
</div>
<div id="strongSessioninformationstrongrecorded"
class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! Recording the operating system, R version, and package
versions is critical for reproducibility.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongCachestrongdetected">
<span class="glyphicon glyphicon-exclamation-sign text-danger"
aria-hidden="true"></span> <strong>Cache:</strong> detected </a>
</p>
</div>
<div id="strongCachestrongdetected" class="panel-collapse collapse">
<div class="panel-body">
The following chunks had caches available:
<ul>
<li>
flashier-nmf
</li>
<li>
flashier-nmf-k3
</li>
</ul>
<p>To ensure reproducibility of the results, delete the cache directory
<code>mcf7_cache</code> and re-run the analysis. To have workflowr
automatically delete the cache directory prior to building the file, set
<code>delete_cache = TRUE</code> when running <code>wflow_build()</code>
or <code>wflow_publish()</code>.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongFilepathsstrongrelative">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>File paths:</strong> relative </a>
</p>
</div>
<div id="strongFilepathsstrongrelative" class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! Using relative paths to the files within your workflowr
project makes it easier to run your code on other machines.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongRepositoryversionstrongahrefhttpsgithubcomstephenslabsinglecelljamboreetreef5e855ca943b6cddf3caa925ad1a4a5cff115695targetblankf5e855ca">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Repository version:</strong>
<a href="https://github.com/stephenslab/single-cell-jamboree/tree/f5e855ca943b6cddf3caa925ad1a4a5cff115695" target="_blank">f5e855c</a>
</a>
</p>
</div>
<div
id="strongRepositoryversionstrongahrefhttpsgithubcomstephenslabsinglecelljamboreetreef5e855ca943b6cddf3caa925ad1a4a5cff115695targetblankf5e855ca"
class="panel-collapse collapse">
<div class="panel-body">
<p>
Great! You are using Git for version control. Tracking code development
and connecting the code version to the results is critical for
reproducibility.
</p>
<p>
The results in this page were generated with repository version
<a href="https://github.com/stephenslab/single-cell-jamboree/tree/f5e855ca943b6cddf3caa925ad1a4a5cff115695" target="_blank">f5e855c</a>.
See the <em>Past versions</em> tab to see a history of the changes made
to the R Markdown and HTML files.
</p>
<p>
Note that you need to be careful to ensure that all relevant files for
the analysis have been committed to Git prior to generating the results
(you can use <code>wflow_publish</code> or
<code>wflow_git_commit</code>). workflowr only checks the R Markdown
file, but you know if there are other scripts or data files that it
depends on. Below is the status of the Git repository when the results
were generated:
</p>
<pre><code>
Untracked files:
    Untracked:  analysis/lps_cache/
    Untracked:  analysis/mcf7_cache/
    Untracked:  analysis/pancreas_cytokine_S1_factors_cache/
    Untracked:  analysis/pancreas_cytokine_lsa_clustering_cache/
    Untracked:  data/GSE132188_adata.h5ad.h5
    Untracked:  data/GSE156175_RAW/
    Untracked:  data/GSE183010/
    Untracked:  data/Immune_ALL_human.h5ad
    Untracked:  data/pancreas_cytokine.RData
    Untracked:  data/pancreas_cytokine_lsa.RData
    Untracked:  data/pancreas_cytokine_lsa_v2.RData
    Untracked:  data/pancreas_endocrine.RData
    Untracked:  data/pancreas_endocrine_alldays.h5ad
    Untracked:  output/panc_cyto_lsa_res/

Unstaged changes:
    Modified:   analysis/temp4.R

</code></pre>
<p>
Note that any generated files, e.g. HTML, png, CSS, etc., are not
included in this status report because it is ok for generated content to
have uncommitted changes.
</p>
</div>
</div>
</div>
</div>
<hr>
</div>
<div id="versions" class="tab-pane fade">

<p>
These are the previous versions of the repository in which changes were
made to the R Markdown (<code>analysis/mcf7.Rmd</code>) and HTML
(<code>docs/mcf7.html</code>) files. If you’ve configured a remote Git
repository (see <code>?wflow_git_remote</code>), click on the hyperlinks
in the table below to view the files as they were in that past version.
</p>
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
File
</th>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
<th>
Message
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/stephenslab/single-cell-jamboree/blob/f5e855ca943b6cddf3caa925ad1a4a5cff115695/analysis/mcf7.Rmd" target="_blank">f5e855c</a>
</td>
<td>
Peter Carbonetto
</td>
<td>
2025-08-12
</td>
<td>
wflow_publish("mcf7.Rmd", verbose = T, view = F)
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/stephenslab/single-cell-jamboree/blob/cdeb683679b4513b98444dc6fe2e9b745ad02bd9/analysis/mcf7.Rmd" target="_blank">cdeb683</a>
</td>
<td>
Peter Carbonetto
</td>
<td>
2025-08-12
</td>
<td>
Added cosine similarity scores to the scatterplots in the mcf7 analysis.
</td>
</tr>
<tr>
<td>
html
</td>
<td>
<a href="https://rawcdn.githack.com/stephenslab/single-cell-jamboree/41566f1f063a1bdf9ebfff5d84ab306388f57e4b/docs/mcf7.html" target="_blank">41566f1</a>
</td>
<td>
Peter Carbonetto
</td>
<td>
2025-07-05
</td>
<td>
Added some ‘other results’ to the mcf7 analysis.
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/stephenslab/single-cell-jamboree/blob/c3e4bdce9de0edebe4622d594936d314c92a1153/analysis/mcf7.Rmd" target="_blank">c3e4bdc</a>
</td>
<td>
Peter Carbonetto
</td>
<td>
2025-07-05
</td>
<td>
wflow_publish("mcf7.Rmd", view = F, verbose = T)
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/stephenslab/single-cell-jamboree/blob/88a7010be1b56e657c31c54f66169488e35effb9/analysis/mcf7.Rmd" target="_blank">88a7010</a>
</td>
<td>
Peter Carbonetto
</td>
<td>
2025-07-05
</td>
<td>
Added k=3 EBNMF fit to the mcf7 example.
</td>
</tr>
<tr>
<td>
html
</td>
<td>
<a href="https://rawcdn.githack.com/stephenslab/single-cell-jamboree/3d8a2aa81fe4b8589b32b2e4b8e5cffac58586d5/docs/mcf7.html" target="_blank">3d8a2aa</a>
</td>
<td>
Peter Carbonetto
</td>
<td>
2025-07-04
</td>
<td>
Added lps_gsea_fl_nmf.csv output.
</td>
</tr>
<tr>
<td>
html
</td>
<td>
<a href="https://rawcdn.githack.com/stephenslab/single-cell-jamboree/3c1002719b626045bfa65ca157f839fd9500617b/docs/mcf7.html" target="_blank">3c10027</a>
</td>
<td>
Peter Carbonetto
</td>
<td>
2025-07-01
</td>
<td>
Added scatterplots to the mcf7 analysis comparing the factors to the
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/stephenslab/single-cell-jamboree/blob/5f49f795c5120b6824eb761782b0eb93640c32ea/analysis/mcf7.Rmd" target="_blank">5f49f79</a>
</td>
<td>
Peter Carbonetto
</td>
<td>
2025-07-01
</td>
<td>
wflow_publish("mcf7.Rmd", verbose = T, view = F)
</td>
</tr>
<tr>
<td>
html
</td>
<td>
<a href="https://rawcdn.githack.com/stephenslab/single-cell-jamboree/ac8a7b86ddfd2a752df3c161abe18a62b9d536c9/docs/mcf7.html" target="_blank">ac8a7b8</a>
</td>
<td>
Peter Carbonetto
</td>
<td>
2025-06-30
</td>
<td>
Ran wflow_publish("mcf7.Rmd").
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/stephenslab/single-cell-jamboree/blob/7c52050d0a55790668ca50e7b615e6f835acf019/analysis/mcf7.Rmd" target="_blank">7c52050</a>
</td>
<td>
Peter Carbonetto
</td>
<td>
2025-06-30
</td>
<td>
Added some scatterplots to the mcf7 analysis to visualize the F matrix.
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/stephenslab/single-cell-jamboree/blob/22b502cd35372799375a6663135dedeee1756c6e/analysis/mcf7.Rmd" target="_blank">22b502c</a>
</td>
<td>
Peter Carbonetto
</td>
<td>
2025-06-28
</td>
<td>
Added a couple more notes to mcf7.Rmd.
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/stephenslab/single-cell-jamboree/blob/4559fc0d6016e471fc0f9adc90528f13702bbe37/analysis/mcf7.Rmd" target="_blank">4559fc0</a>
</td>
<td>
Peter Carbonetto
</td>
<td>
2025-06-28
</td>
<td>
Added some notes to the mcf7 analysis.
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/stephenslab/single-cell-jamboree/blob/0cb8ea79df6ccb8d523dc1c175a7c1a22356ce05/analysis/mcf7.Rmd" target="_blank">0cb8ea7</a>
</td>
<td>
Peter Carbonetto
</td>
<td>
2025-06-27
</td>
<td>
Added scatterplots to compare the gene expression estimates (F) from the
topic model.
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/stephenslab/single-cell-jamboree/blob/1860816e06ad321cdba453f1730e00473c73b7b5/analysis/mcf7.Rmd" target="_blank">1860816</a>
</td>
<td>
Peter Carbonetto
</td>
<td>
2025-06-27
</td>
<td>
Working on some improvements to the mcf-7 data analysis.
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/stephenslab/single-cell-jamboree/blob/8f39607ebba82349cd62736373fc6886c917ec5f/analysis/mcf7.Rmd" target="_blank">8f39607</a>
</td>
<td>
Peter Carbonetto
</td>
<td>
2025-06-04
</td>
<td>
Added steps to the lps analysis to load and prepare the data.
</td>
</tr>
<tr>
<td>
html
</td>
<td>
<a href="https://rawcdn.githack.com/stephenslab/single-cell-jamboree/de904d4aee4e79694d1beb3248e1ecdb43dccc50/docs/mcf7.html" target="_blank">de904d4</a>
</td>
<td>
Peter Carbonetto
</td>
<td>
2025-06-03
</td>
<td>
Added flashier fit to the mcf7 analysis.
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/stephenslab/single-cell-jamboree/blob/1ac5a526fb7f8ae61144f12d54f3aba0e6c6f259/analysis/mcf7.Rmd" target="_blank">1ac5a52</a>
</td>
<td>
Peter Carbonetto
</td>
<td>
2025-06-03
</td>
<td>
wflow_publish("mcf7.Rmd", verbose = TRUE, view = FALSE)
</td>
</tr>
<tr>
<td>
html
</td>
<td>
<a href="https://rawcdn.githack.com/stephenslab/single-cell-jamboree/5dceacd10d9338a079e584a022bc9cf3db82c76a/docs/mcf7.html" target="_blank">5dceacd</a>
</td>
<td>
Peter Carbonetto
</td>
<td>
2025-06-03
</td>
<td>
A few small adjustments to the topic modeling results in the mcf7
analysis.
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/stephenslab/single-cell-jamboree/blob/4499b61e965920ec63d0eb72952425037f9b4238/analysis/mcf7.Rmd" target="_blank">4499b61</a>
</td>
<td>
Peter Carbonetto
</td>
<td>
2025-06-03
</td>
<td>
wflow_publish("mcf7.Rmd", verbose = TRUE, view = FALSE)
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/stephenslab/single-cell-jamboree/blob/d3cb20f57c0f473983153de8b75e835ecc0f8a1b/analysis/mcf7.Rmd" target="_blank">d3cb20f</a>
</td>
<td>
Peter Carbonetto
</td>
<td>
2025-06-03
</td>
<td>
Added topic model fit and structure plot to mcf7 analysis.
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/stephenslab/single-cell-jamboree/blob/b191ab5428a7beba44f1310d878acddb1bef629d/analysis/mcf7.Rmd" target="_blank">b191ab5</a>
</td>
<td>
Peter Carbonetto
</td>
<td>
2025-06-03
</td>
<td>
Added PCA and GLM-PCA results to mcf7 analysis.
</td>
</tr>
<tr>
<td>
html
</td>
<td>
<a href="https://rawcdn.githack.com/stephenslab/single-cell-jamboree/a0b1bbe080d046440adbbba34c1d4b06882281ad/docs/mcf7.html" target="_blank">a0b1bbe</a>
</td>
<td>
Peter Carbonetto
</td>
<td>
2025-06-03
</td>
<td>
First build of the mcf7 analysis.
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/stephenslab/single-cell-jamboree/blob/a3d82cef442efa6b577db11d1c576343538b436f/analysis/mcf7.Rmd" target="_blank">a3d82ce</a>
</td>
<td>
Peter Carbonetto
</td>
<td>
2025-06-03
</td>
<td>
wflow_publish("mcf7.Rmd", verbose = TRUE, view = FALSE)
</td>
</tr>
</tbody>
</table>
</div>
<hr>
</div>
</div>
</div>
<p>The MCF-7 data set from <a
href="https://doi.org/10.7554/eLife.59388">Sanford et al 2020 eLife
paper</a> is helpful for illustrating some of the basic ideas behind NMF
analysis of single-cell data, although interpretation of the factors as
gene programs is less clear.</p>
<p>This paper analyzed the transcriptional response of human MCF-7 cells
to retinoic acid and TGF-<span class="math inline">\(\beta\)</span>,
applied individually and in combination. These data were downloaded from
<a
href="https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE152749">GEO
accession GSE152749</a>.</p>
<p>The following abbreviations are used in analysis: EtOH = ethanol; RA
= retinoic acid; TGFb = TGF-<span
class="math inline">\(\beta\)</span>.</p>
<p>Load packages used to process the data, perform the analyses, and
create the plots.</p>
<pre class="r"><code>library(rsvd)
library(lsa)
library(fastglmpca)
library(fastTopics)
library(flashier)
library(data.table)
library(GEOquery)
library(singlecelljamboreeR)
library(ggplot2)
library(cowplot)</code></pre>
<p>Set the seed for reproducibility:</p>
<pre class="r"><code>set.seed(1)</code></pre>
<div id="prepare-the-data-for-analysis-with-fasttopics-and-flashier"
class="section level2">
<h2>Prepare the data for analysis with fastTopics and flashier</h2>
<p>Load the RNA-seq counts:</p>
<pre class="r"><code>counts &lt;- fread(&quot;../data/GSE152749_raw_counts_GRCh38.p13_NCBI.tsv.gz&quot;,
                sep = &quot;\t&quot;,header = TRUE,stringsAsFactors = FALSE)
class(counts) &lt;- &quot;data.frame&quot;
rownames(counts) &lt;- counts$GeneID
counts &lt;- counts[,-1]
counts &lt;- as.matrix(counts)
storage.mode(counts) &lt;- &quot;double&quot;
counts &lt;- t(counts)
ids &lt;- rownames(counts)</code></pre>
<p>Load the gene information:</p>
<pre class="r"><code>genes &lt;- fread(&quot;../data/Human.GRCh38.p13.annot.tsv.gz&quot;,sep = &quot;\t&quot;,
               header = TRUE,stringsAsFactors = FALSE)
class(genes) &lt;- &quot;data.frame&quot;
genes &lt;- genes[1:10]
genes &lt;- transform(genes,
                   GeneType = factor(GeneType),
                   Status   = factor(Status))</code></pre>
<p>Load the sample information:</p>
<pre class="r"><code>geo &lt;- getGEO(filename = &quot;../data/GSE152749_family.soft.gz&quot;)
samples &lt;- data.frame(id = names(GSMList(geo)),
                      treatment = sapply(GSMList(geo),
                                         function (x) Meta(x)$title))
samples &lt;- samples[ids,]
rownames(samples) &lt;- NULL
samples &lt;- transform(samples,
                     EtOH = grepl(&quot;EtOH&quot;,treatment,fixed = TRUE),
                     RA   = grepl(&quot;RA&quot;,treatment,fixed = TRUE),
                     TGFb = grepl(&quot;TGFb&quot;,treatment,fixed = TRUE))
samples$label                 &lt;- &quot;EtOH&quot;
samples[samples$RA,&quot;label&quot;]   &lt;- &quot;RA&quot;
samples[samples$TGFb,&quot;label&quot;] &lt;- &quot;TGFb&quot;
samples[with(samples,RA &amp; TGFb),&quot;label&quot;] &lt;- &quot;RA+TGFb&quot;
samples &lt;- transform(samples,
                     label = factor(label,c(&quot;EtOH&quot;,&quot;RA&quot;,&quot;TGFb&quot;,&quot;RA+TGFb&quot;)))</code></pre>
<p>Remove the non-protein-coding genes, and the genes that are expressed
in fewer than 4 samples:</p>
<pre class="r"><code>x &lt;- colSums(counts &gt; 0)
i &lt;- which(x &gt; 3 &amp;
           genes$GeneType == &quot;protein-coding&quot; &amp;
           genes$Status == &quot;active&quot;)
genes  &lt;- genes[i,]
counts &lt;- counts[,i]</code></pre>
<p><strong>Some background:</strong> it turns out that there is some
structure in the non-coding RNA genes that is is picked up by the
Poisson-based methods but not by the Gaussian-based methods, so to avoid
this complication in the comparisons I have removed these genes from the
data, which aren’t of interest anyhow.</p>
<p>This is the dimension of the data set we will analyze:</p>
<pre class="r"><code>dim(counts)
# [1]    41 16773</code></pre>
<p>For the Gaussian-based analyses, we will need the shifted log
counts:</p>
<pre class="r"><code>a &lt;- 1
s &lt;- rowSums(counts)
s &lt;- s/mean(s)
shifted_log_counts &lt;- log1p(counts/(a*s))</code></pre>
</div>
<div id="pca-and-glm-pca" class="section level2">
<h2>PCA and GLM-PCA</h2>
<p>Let’s see what happens when we apply PCA to the shifted log counts
and GLM-PCA to the counts. As we will see, PCA and GLM-PCA both show a
clear clustering of the data that corresponds to the different
treatments. However, the PCs are not necessarily individually
interpretable.</p>
<p>Run PCA on the shifted log counts, and plot the first 2 PCs:</p>
<pre class="r"><code>pca &lt;- rpca(shifted_log_counts,k = 2,center = TRUE,scale = FALSE)
colnames(pca$x) &lt;- c(&quot;PC1&quot;,&quot;PC2&quot;)
pdat &lt;- data.frame(samples,pca$x)
ggplot(pdat,aes(x = PC1,y = PC2,color = label)) +
  geom_point() +
  scale_color_manual(values = c(&quot;dodgerblue&quot;,&quot;tomato&quot;,&quot;darkblue&quot;,
                                &quot;limegreen&quot;)) + 
  theme_cowplot(font_size = 10)</code></pre>
<p><img src="figure/mcf7.Rmd/pca-1.png" width="360" style="display: block; margin: auto;" /></p>
<p>
<button type="button" class="btn btn-default btn-xs btn-workflowr btn-workflowr-fig" data-toggle="collapse" data-target="#fig-pca-1">
Past versions of pca-1.png
</button>
</p>
<div id="fig-pca-1" class="collapse">
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
<a href="https://github.com/stephenslab/single-cell-jamboree/blob/5dceacd10d9338a079e584a022bc9cf3db82c76a/docs/figure/mcf7.Rmd/pca-1.png" target="_blank">5dceacd</a>
</td>
<td>
Peter Carbonetto
</td>
<td>
2025-06-03
</td>
</tr>
</tbody>
</table>
</div>
</div>
<p>GLM-PCA applied to the counts essentially produces the same result
(aside from an arbitrary rotation):</p>
<pre class="r"><code>fit_glmpca &lt;- init_glmpca_pois(t(counts),K = 2)
fit_glmpca &lt;- fit_glmpca_pois(t(counts),fit0 = fit_glmpca,verbose = FALSE,
                              control = list(maxiter = 50))
colnames(fit_glmpca$V) &lt;- c(&quot;k1&quot;,&quot;k2&quot;)
pdat &lt;- data.frame(samples,fit_glmpca$V)
ggplot(pdat,aes(x = k1,y = k2,color = label)) +
  geom_point() +
  scale_color_manual(values = c(&quot;dodgerblue&quot;,&quot;tomato&quot;,&quot;darkblue&quot;,
                                &quot;limegreen&quot;)) + 
  theme_cowplot(font_size = 10)</code></pre>
<p><img src="figure/mcf7.Rmd/fastglmpca-1.png" width="360" style="display: block; margin: auto;" /></p>
<p>
<button type="button" class="btn btn-default btn-xs btn-workflowr btn-workflowr-fig" data-toggle="collapse" data-target="#fig-fastglmpca-1">
Past versions of fastglmpca-1.png
</button>
</p>
<div id="fig-fastglmpca-1" class="collapse">
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
<a href="https://github.com/stephenslab/single-cell-jamboree/blob/5dceacd10d9338a079e584a022bc9cf3db82c76a/docs/figure/mcf7.Rmd/fastglmpca-1.png" target="_blank">5dceacd</a>
</td>
<td>
Peter Carbonetto
</td>
<td>
2025-06-03
</td>
</tr>
</tbody>
</table>
</div>
</div>
</div>
<div id="topic-model-fasttopics" class="section level2">
<h2>Topic model (fastTopics)</h2>
<p>Fit a topic model with <span class="math inline">\(K = 3\)</span>
topics to the counts:</p>
<pre class="r"><code>tm0 &lt;- fit_poisson_nmf(counts,k = 3,init.method = &quot;random&quot;,
                       numiter = 50,verbose = &quot;none&quot;,
                       control = list(nc = 4,extrapolate = FALSE))
tm &lt;- fit_poisson_nmf(counts,fit0 = tm0,numiter = 50,verbose = &quot;none&quot;,
                      control = list(nc = 4,extrapolate = TRUE))</code></pre>
<p>Structure plot comparing the topics to the clusters or treatment
conditions:</p>
<pre class="r"><code>topic_colors &lt;- c(&quot;tomato&quot;,&quot;darkblue&quot;,&quot;dodgerblue&quot;)
n &lt;- nrow(counts)
L &lt;- poisson2multinom(tm)$L
structure_plot(L,grouping = samples$label,topics = 1:3,
               loadings_order = 1:n,colors = topic_colors) +
  theme(axis.text.x = element_text(angle = 0,hjust = 0.5))</code></pre>
<p><img src="figure/mcf7.Rmd/structure-plot-topic-model-1.png" width="600" style="display: block; margin: auto;" /></p>
<p>
<button type="button" class="btn btn-default btn-xs btn-workflowr btn-workflowr-fig" data-toggle="collapse" data-target="#fig-structure-plot-topic-model-1">
Past versions of structure-plot-topic-model-1.png
</button>
</p>
<div id="fig-structure-plot-topic-model-1" class="collapse">
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
<a href="https://github.com/stephenslab/single-cell-jamboree/blob/ac8a7b86ddfd2a752df3c161abe18a62b9d536c9/docs/figure/mcf7.Rmd/structure-plot-topic-model-1.png" target="_blank">ac8a7b8</a>
</td>
<td>
Peter Carbonetto
</td>
<td>
2025-06-30
</td>
</tr>
<tr>
<td>
<a href="https://github.com/stephenslab/single-cell-jamboree/blob/5dceacd10d9338a079e584a022bc9cf3db82c76a/docs/figure/mcf7.Rmd/structure-plot-topic-model-1.png" target="_blank">5dceacd</a>
</td>
<td>
Peter Carbonetto
</td>
<td>
2025-06-03
</td>
</tr>
</tbody>
</table>
</div>
</div>
<p>It is also helpful to visualize the gene matrix, which we can roughly
think about as capturing levels of gene expression. These next three
scatterplots compare the estimated expression levels among the three
topics (on the logarithmic scale).</p>
<p>TO DO: Add cosine similarity to the scatterplots.</p>
<pre class="r"><code>gene_matrix_scatterplot &lt;-
  function (gene_matrix, k1, k2, log = &quot;none&quot;, cosine = TRUE) {
  gene_matrix &lt;- as.data.frame(gene_matrix)
  if (cosine)
    s &lt;- cosine(gene_matrix[,k1],gene_matrix[,k2])
  if (log == &quot;x&quot;)
    gene_matrix[,k1] &lt;- log10(gene_matrix[,k1] + 1e-6)
  else if (log == &quot;y&quot;)
    gene_matrix[,k2] &lt;- log10(gene_matrix[,k2] + 1e-6)
  else if (log == &quot;both&quot;)
    gene_matrix[,c(k1,k2)] &lt;- log10(gene_matrix[,c(k1,k2)] + 1e-6)
  out &lt;- ggplot(gene_matrix,aes_string(x = k1,y = k2)) +
      geom_point(color = &quot;black&quot;) +
      geom_abline(intercept = 0,slope = 1,color = &quot;orangered&quot;,
                  linetype = &quot;dashed&quot;) +
      
      theme_cowplot(font_size = 10) +
      theme(plot.title = element_text(face = &quot;plain&quot;,size = 10))
  if (cosine) {
    out &lt;- out + ggtitle(sprintf(&quot;cosine = %0.3f&quot;,s))
  }
  return(out)
}
F &lt;- poisson2multinom(tm)$F
colnames(F) &lt;- c(&quot;TGFb&quot;,&quot;RA&quot;,&quot;EtOH&quot;)
p1 &lt;- gene_matrix_scatterplot(F,&quot;EtOH&quot;,&quot;RA&quot;,log = &quot;both&quot;,cosine = TRUE)
p2 &lt;- gene_matrix_scatterplot(F,&quot;EtOH&quot;,&quot;TGFb&quot;,log = &quot;both&quot;,cosine = TRUE)
p3 &lt;- gene_matrix_scatterplot(F,&quot;RA&quot;,&quot;TGFb&quot;,log = &quot;both&quot;,cosine = TRUE)
plot_grid(p1,p2,p3,nrow = 1,ncol = 3)</code></pre>
<p><img src="figure/mcf7.Rmd/gene-matrix-scatterplots-topic-model-1.png" width="720" style="display: block; margin: auto;" /></p>
<p>
<button type="button" class="btn btn-default btn-xs btn-workflowr btn-workflowr-fig" data-toggle="collapse" data-target="#fig-gene-matrix-scatterplots-topic-model-1">
Past versions of gene-matrix-scatterplots-topic-model-1.png
</button>
</p>
<div id="fig-gene-matrix-scatterplots-topic-model-1" class="collapse">
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
<a href="https://github.com/stephenslab/single-cell-jamboree/blob/ac8a7b86ddfd2a752df3c161abe18a62b9d536c9/docs/figure/mcf7.Rmd/gene-matrix-scatterplots-topic-model-1.png" target="_blank">ac8a7b8</a>
</td>
<td>
Peter Carbonetto
</td>
<td>
2025-06-30
</td>
</tr>
</tbody>
</table>
</div>
</div>
<p>(Note: the gene matrix is being shown on the logarithmic scale in
these plots.)</p>
<p>What we see here is that most of the gene expression does not change
much, except for a few genes far from the diagonal. These are the
“distinctive” or “characteristic” genes for these factors. Looking at
the genes with the top expression values (the largest elements of the
<span class="math inline">\({\bf F}\)</span> matrix) will not isolate
these distinctive genes. Intuitively, this is what we are often trying
to do when performing differential expression analyses that compare
different cell types: we are most interested in identifying the genes
that show large differences in gene expression in the given cell type
vs. other celltypes.</p>
<p>Summary: although the topic model clearly is not just doing a simple
clustering, the topics are often “cluster-like”, and therefore it is
natural to think about analyses that are performed on clusters (e.g.,
differential expression analysis) as also applying to this setting.</p>
</div>
<div id="ebnmf-flashier" class="section level2">
<h2>EBNMF (flashier)</h2>
<p>Next fit an NMF to the shifted log counts using flashier, with <span
class="math inline">\(K = 4\)</span>:</p>
<pre class="r"><code>n  &lt;- nrow(counts)
x  &lt;- rpois(1e7,1/n)
s1 &lt;- sd(log(x + 1))
fl_nmf &lt;- flash(shifted_log_counts,ebnm_fn = ebnm_point_exponential,
                var_type = 2, greedy_Kmax = 4,S = s1,backfit = TRUE,
                verbose = 0)</code></pre>
<div class="alert alert-warning">
<p>
<strong>Warning:</strong> The above code chunk cached its results, but
it won’t be re-run if previous chunks it depends on are updated. If you
need to use caching, it is highly recommended to also set
<code>knitr::opts_chunk$set(autodep = TRUE)</code> at the top of the
file (in a chunk that is not cached). Alternatively, you can customize
the option <code>dependson</code> for each individual chunk that is
cached. Using either <code>autodep</code> or <code>dependson</code> will
remove this warning. See the
<a href="https://yihui.name/knitr/options/#cache"
>knitr cache options</a> for more details.
</p>
</div>
<p>EBNMF applied to the shifted log counts decomposes the samples in a
similar way as the topic model:</p>
<pre class="r"><code>topic_colors &lt;- c(&quot;olivedrab&quot;,&quot;dodgerblue&quot;,&quot;darkblue&quot;,&quot;tomato&quot;)
L &lt;- ldf(fl_nmf,type = &quot;i&quot;)$L
structure_plot(L,grouping = samples$label,topics = 4:1,
               loadings_order = 1:n,colors = topic_colors) +
  labs(y = &quot;membership&quot;) +
  theme(axis.text.x = element_text(angle = 0,hjust = 0.5))</code></pre>
<p><img src="figure/mcf7.Rmd/structure-plot-nmf-1.png" width="600" style="display: block; margin: auto;" /></p>
<p>
<button type="button" class="btn btn-default btn-xs btn-workflowr btn-workflowr-fig" data-toggle="collapse" data-target="#fig-structure-plot-nmf-1">
Past versions of structure-plot-nmf-1.png
</button>
</p>
<div id="fig-structure-plot-nmf-1" class="collapse">
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
<a href="https://github.com/stephenslab/single-cell-jamboree/blob/ac8a7b86ddfd2a752df3c161abe18a62b9d536c9/docs/figure/mcf7.Rmd/structure-plot-nmf-1.png" target="_blank">ac8a7b8</a>
</td>
<td>
Peter Carbonetto
</td>
<td>
2025-06-30
</td>
</tr>
<tr>
<td>
<a href="https://github.com/stephenslab/single-cell-jamboree/blob/de904d4aee4e79694d1beb3248e1ecdb43dccc50/docs/figure/mcf7.Rmd/structure-plot-nmf-1.png" target="_blank">de904d4</a>
</td>
<td>
Peter Carbonetto
</td>
<td>
2025-06-03
</td>
</tr>
</tbody>
</table>
</div>
</div>
<p>These next three scatterplots compare the factor estimates
(approximately, LFCs) among the three factors other than the
baseline:</p>
<pre class="r"><code>F &lt;- ldf(fl_nmf,type = &quot;i&quot;)$F
colnames(F) &lt;- c(&quot;baseline&quot;,&quot;EtOH&quot;,&quot;RA&quot;,&quot;TGFb&quot;)
pdat &lt;- cbind(genes[c(&quot;GeneID&quot;,&quot;Symbol&quot;)],F)
p1 &lt;- gene_matrix_scatterplot(F,&quot;EtOH&quot;,&quot;RA&quot;,log = &quot;none&quot;,cosine = TRUE)
p2 &lt;- gene_matrix_scatterplot(F,&quot;EtOH&quot;,&quot;TGFb&quot;,log = &quot;none&quot;,cosine = TRUE)
p3 &lt;- gene_matrix_scatterplot(F,&quot;RA&quot;,&quot;TGFb&quot;,log = &quot;none&quot;,cosine = TRUE)
print(plot_grid(p1,p2,p3,nrow = 1,ncol = 3))</code></pre>
<p><img src="figure/mcf7.Rmd/gene-matrix-scatterplot-1.png" width="720" style="display: block; margin: auto;" /></p>
<p>
<button type="button" class="btn btn-default btn-xs btn-workflowr btn-workflowr-fig" data-toggle="collapse" data-target="#fig-gene-matrix-scatterplot-1">
Past versions of gene-matrix-scatterplot-1.png
</button>
</p>
<div id="fig-gene-matrix-scatterplot-1" class="collapse">
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
<a href="https://github.com/stephenslab/single-cell-jamboree/blob/ac8a7b86ddfd2a752df3c161abe18a62b9d536c9/docs/figure/mcf7.Rmd/gene-matrix-scatterplot-1.png" target="_blank">ac8a7b8</a>
</td>
<td>
Peter Carbonetto
</td>
<td>
2025-06-30
</td>
</tr>
</tbody>
</table>
</div>
</div>
<p>More than the topic model, the NMF factors better isolate more
“distinctive” genes; we see that the largest values of the gene matrix
are often also a gene that is zero or small in the other factors,
whereas this was clearly not true in the topic model.</p>
<p>To reinforce this point, this last set of scatterplots compares the
topics to the corresponding factors:</p>
<pre class="r"><code>F1 &lt;- poisson2multinom(tm)$F
F2 &lt;- ldf(fl_nmf,type = &quot;i&quot;)$F
colnames(F1) &lt;- c(&quot;TGFb_tm&quot;,&quot;RA_tm&quot;,&quot;EtOH_tm&quot;)
colnames(F2) &lt;- c(&quot;baseline&quot;,&quot;EtOH_nmf&quot;,&quot;RA_nmf&quot;,&quot;TGFb_nmf&quot;)
F &lt;- cbind(F1,F2)
p1 &lt;- gene_matrix_scatterplot(F,&quot;EtOH_tm&quot;,&quot;EtOH_nmf&quot;,log = &quot;x&quot;,cosine = FALSE)
p2 &lt;- gene_matrix_scatterplot(F,&quot;RA_tm&quot;,&quot;RA_nmf&quot;,log = &quot;x&quot;,cosine = FALSE)
p3 &lt;- gene_matrix_scatterplot(F,&quot;TGFb_tm&quot;,&quot;TGFb_nmf&quot;,log = &quot;x&quot;,cosine = FALSE)
print(plot_grid(p1,p2,p3,nrow = 1,ncol = 3))</code></pre>
<p><img src="figure/mcf7.Rmd/nmf-vs-tm-1.png" width="720" style="display: block; margin: auto;" /></p>
<p>
<button type="button" class="btn btn-default btn-xs btn-workflowr btn-workflowr-fig" data-toggle="collapse" data-target="#fig-nmf-vs-tm-1">
Past versions of nmf-vs-tm-1.png
</button>
</p>
<div id="fig-nmf-vs-tm-1" class="collapse">
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
<a href="https://github.com/stephenslab/single-cell-jamboree/blob/3c1002719b626045bfa65ca157f839fd9500617b/docs/figure/mcf7.Rmd/nmf-vs-tm-1.png" target="_blank">3c10027</a>
</td>
<td>
Peter Carbonetto
</td>
<td>
2025-07-01
</td>
</tr>
</tbody>
</table>
</div>
</div>
<p>Note that the topics (“tm”) are plotted along the x-axis on the log
scale.</p>
</div>
<div id="other-results" class="section level2">
<h2>Other results</h2>
<p>Note that the EBNMF result with <span class="math inline">\(K =
3\)</span> is less satisfying—it does not identify a factor specifically
for the RA exposure—which is why I have focussed above on the <span
class="math inline">\(K = 4\)</span> result.</p>
<pre class="r"><code>fl_nmf_k3 &lt;- flash(shifted_log_counts,ebnm_fn = ebnm_point_exponential,
                   var_type = 2, greedy_Kmax = 3,S = s1,backfit = TRUE,
                   verbose = 0)</code></pre>
<div class="alert alert-warning">
<p>
<strong>Warning:</strong> The above code chunk cached its results, but
it won’t be re-run if previous chunks it depends on are updated. If you
need to use caching, it is highly recommended to also set
<code>knitr::opts_chunk$set(autodep = TRUE)</code> at the top of the
file (in a chunk that is not cached). Alternatively, you can customize
the option <code>dependson</code> for each individual chunk that is
cached. Using either <code>autodep</code> or <code>dependson</code> will
remove this warning. See the
<a href="https://yihui.name/knitr/options/#cache"
>knitr cache options</a> for more details.
</p>
</div>
<p>Here’s the Structure plot for this EBNMF model with <span
class="math inline">\(K = 3\)</span>:</p>
<pre class="r"><code>topic_colors &lt;- c(&quot;dodgerblue&quot;,&quot;darkblue&quot;,&quot;tomato&quot;)
L &lt;- ldf(fl_nmf_k3,type = &quot;i&quot;)$L
structure_plot(L,grouping = samples$label,topics = 3:1,
               loadings_order = 1:n,colors = topic_colors) +
  labs(y = &quot;membership&quot;) +
  theme(axis.text.x = element_text(angle = 0,hjust = 0.5))</code></pre>
<p><img src="figure/mcf7.Rmd/structure-plot-nmf-k3-1.png" width="600" style="display: block; margin: auto;" /></p>
<p>
<button type="button" class="btn btn-default btn-xs btn-workflowr btn-workflowr-fig" data-toggle="collapse" data-target="#fig-structure-plot-nmf-k3-1">
Past versions of structure-plot-nmf-k3-1.png
</button>
</p>
<div id="fig-structure-plot-nmf-k3-1" class="collapse">
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
<a href="https://github.com/stephenslab/single-cell-jamboree/blob/41566f1f063a1bdf9ebfff5d84ab306388f57e4b/docs/figure/mcf7.Rmd/structure-plot-nmf-k3-1.png" target="_blank">41566f1</a>
</td>
<td>
Peter Carbonetto
</td>
<td>
2025-07-05
</td>
</tr>
</tbody>
</table>
</div>
</div>
<p>Also, with <span class="math inline">\(K = 4\)</span>, the topic
model doesn’t quite identify a “baseline”, but instead identifies two
“control condition” (ethanol) topics:</p>
<pre class="r"><code>topic_colors &lt;- c(&quot;olivedrab&quot;,&quot;dodgerblue&quot;,&quot;darkblue&quot;,&quot;tomato&quot;)
set.seed(1)
tm_k4 &lt;- fit_poisson_nmf(counts,k = 4,init.method = &quot;random&quot;,
                         numiter = 100,verbose = &quot;none&quot;,
                         control = list(nc = 4,extrapolate = FALSE))
tm_k4 &lt;- fit_poisson_nmf(counts,fit0 = tm_k4,numiter = 200,verbose = &quot;none&quot;,
                         control = list(nc = 4,extrapolate = TRUE))
L &lt;- poisson2multinom(tm_k4)$L
structure_plot(L,grouping = samples$label,topics = 1:4,
               loadings_order = 1:n,colors = topic_colors) +
  theme(axis.text.x = element_text(angle = 0,hjust = 0.5))</code></pre>
<p><img src="figure/mcf7.Rmd/fit-topic-model-k4-1.png" width="600" style="display: block; margin: auto;" /></p>
<p>
<button type="button" class="btn btn-default btn-xs btn-workflowr btn-workflowr-fig" data-toggle="collapse" data-target="#fig-fit-topic-model-k4-1">
Past versions of fit-topic-model-k4-1.png
</button>
</p>
<div id="fig-fit-topic-model-k4-1" class="collapse">
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
<a href="https://github.com/stephenslab/single-cell-jamboree/blob/41566f1f063a1bdf9ebfff5d84ab306388f57e4b/docs/figure/mcf7.Rmd/fit-topic-model-k4-1.png" target="_blank">41566f1</a>
</td>
<td>
Peter Carbonetto
</td>
<td>
2025-07-05
</td>
</tr>
</tbody>
</table>
</div>
</div>
<p>As before, the expression levels remain very strongly correlated
amongst the topics, e.g.,</p>
<pre class="r"><code>F &lt;- poisson2multinom(tm_k4)$F
p1 &lt;- gene_matrix_scatterplot(F,&quot;k2&quot;,&quot;k3&quot;,log = &quot;both&quot;,cosine = TRUE)
p2 &lt;- gene_matrix_scatterplot(F,&quot;k2&quot;,&quot;k4&quot;,log = &quot;both&quot;,cosine = TRUE)
p3 &lt;- gene_matrix_scatterplot(F,&quot;k3&quot;,&quot;k4&quot;,log = &quot;both&quot;,cosine = TRUE)
plot_grid(p1,p2,p3,nrow = 1,ncol = 3)</code></pre>
<p><img src="figure/mcf7.Rmd/gene-matrix-scatterplot-topic-model-k4-1.png" width="720" style="display: block; margin: auto;" /></p>
<p>
<button type="button" class="btn btn-default btn-xs btn-workflowr btn-workflowr-fig" data-toggle="collapse" data-target="#fig-gene-matrix-scatterplot-topic-model-k4-1">
Past versions of gene-matrix-scatterplot-topic-model-k4-1.png
</button>
</p>
<div id="fig-gene-matrix-scatterplot-topic-model-k4-1" class="collapse">
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
<a href="https://github.com/stephenslab/single-cell-jamboree/blob/41566f1f063a1bdf9ebfff5d84ab306388f57e4b/docs/figure/mcf7.Rmd/gene-matrix-scatterplot-topic-model-k4-1.png" target="_blank">41566f1</a>
</td>
<td>
Peter Carbonetto
</td>
<td>
2025-07-05
</td>
</tr>
</tbody>
</table>
</div>
</div>
<p>reinforcing the general point that the topic model has a tendency
toward “cluster-like” structure.</p>
<br>
<p>
<button type="button" class="btn btn-default btn-workflowr btn-workflowr-sessioninfo" data-toggle="collapse" data-target="#workflowr-sessioninfo" style="display: block;">
<span class="glyphicon glyphicon-wrench" aria-hidden="true"></span>
Session information
</button>
</p>
<div id="workflowr-sessioninfo" class="collapse">
<pre class="r"><code>sessionInfo()
# R version 4.3.3 (2024-02-29)
# Platform: aarch64-apple-darwin20 (64-bit)
# Running under: macOS 15.5
# 
# Matrix products: default
# BLAS:   /Library/Frameworks/R.framework/Versions/4.3-arm64/Resources/lib/libRblas.0.dylib 
# LAPACK: /Library/Frameworks/R.framework/Versions/4.3-arm64/Resources/lib/libRlapack.dylib;  LAPACK version 3.11.0
# 
# locale:
# [1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8
# 
# time zone: America/Chicago
# tzcode source: internal
# 
# attached base packages:
# [1] stats     graphics  grDevices utils     datasets  methods   base     
# 
# other attached packages:
#  [1] cowplot_1.1.3              ggplot2_3.5.2             
#  [3] singlecelljamboreeR_0.1-36 GEOquery_2.70.0           
#  [5] Biobase_2.62.0             BiocGenerics_0.48.1       
#  [7] data.table_1.17.6          flashier_1.0.56           
#  [9] ebnm_1.1-34                fastTopics_0.7-25         
# [11] fastglmpca_0.1-108         lsa_0.73.3                
# [13] SnowballC_0.7.1            rsvd_1.0.5                
# 
# loaded via a namespace (and not attached):
#   [1] distr_2.9.3          pbapply_1.7-2        rlang_1.1.6         
#   [4] magrittr_2.0.3       git2r_0.33.0         horseshoe_0.2.0     
#   [7] matrixStats_1.2.0    susieR_0.14.6        compiler_4.3.3      
#  [10] vctrs_0.6.5          reshape2_1.4.4       daarem_0.7          
#  [13] quadprog_1.5-8       stringr_1.5.1        pkgconfig_2.0.3     
#  [16] crayon_1.5.3         fastmap_1.2.0        labeling_0.4.3      
#  [19] promises_1.3.3       rmarkdown_2.29       tzdb_0.4.0          
#  [22] purrr_1.0.4          xfun_0.52            cachem_1.1.0        
#  [25] trust_0.1-8          jsonlite_2.0.0       progress_1.2.3      
#  [28] later_1.4.2          reshape_0.8.9        irlba_2.3.5.1       
#  [31] parallel_4.3.3       prettyunits_1.2.0    R6_2.6.1            
#  [34] bslib_0.9.0          stringi_1.8.7        RColorBrewer_1.1-3  
#  [37] SQUAREM_2021.1       limma_3.58.1         jquerylib_0.1.4     
#  [40] Rcpp_1.1.0           knitr_1.50           R.utils_2.12.3      
#  [43] readr_2.1.5          httpuv_1.6.14        Matrix_1.6-5        
#  [46] splines_4.3.3        tidyselect_1.2.1     dichromat_2.0-0.1   
#  [49] yaml_2.3.10          lattice_0.22-5       tibble_3.3.0        
#  [52] plyr_1.8.9           withr_3.0.2          evaluate_1.0.4      
#  [55] Rtsne_0.17           RcppParallel_5.1.10  startupmsg_0.9.6.1  
#  [58] xml2_1.3.6           pillar_1.11.0        whisker_0.4.1       
#  [61] plotly_4.11.0        softImpute_1.4-3     generics_0.1.4      
#  [64] rprojroot_2.0.4      invgamma_1.2         truncnorm_1.0-9     
#  [67] hms_1.1.3            scales_1.4.0         ashr_2.2-66         
#  [70] gtools_3.9.5         RhpcBLASctl_0.23-42  glue_1.8.0          
#  [73] scatterplot3d_0.3-44 lazyeval_0.2.2       tools_4.3.3         
#  [76] fs_1.6.6             grid_4.3.3           tidyr_1.3.1         
#  [79] colorspace_2.1-0     sfsmisc_1.1-18       deconvolveR_1.2-1   
#  [82] cli_3.6.5            Polychrome_1.5.1     workflowr_1.7.1     
#  [85] mixsqp_0.3-54        viridisLite_0.4.2    dplyr_1.1.4         
#  [88] uwot_0.2.3           gtable_0.3.6         R.methodsS3_1.8.2   
#  [91] sass_0.4.10          digest_0.6.37        ggrepel_0.9.6       
#  [94] htmlwidgets_1.6.4    farver_2.1.2         R.oo_1.26.0         
#  [97] htmltools_0.5.8.1    lifecycle_1.0.4      httr_1.4.7          
# [100] statmod_1.5.0        MASS_7.3-60.0.1</code></pre>
</div>
</div>


<!-- Adjust MathJax settings so that all math formulae are shown using
TeX fonts only; see
https://docs.mathjax.org/en/latest/web/configuration.html. This will make
the presentation more consistent at the cost of the webpage sometimes
taking slightly longer to load. Note that this only works because the
footer is added to webpages before the MathJax javascript. -->
<script type="text/x-mathjax-config">
  MathJax.Hub.Config({
    "HTML-CSS": { availableFonts: ["TeX"] }
  });
</script>





</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.odd').parent('tbody').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- tabsets -->

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open');
  });
});
</script>

<!-- code folding -->


<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
