<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />


<meta name="author" content="Matthew Stephens" />

<meta name="date" content="2025-07-02" />

<title>pancreas_celseq2_gbcd</title>

<script src="site_libs/header-attrs-2.29/header-attrs.js"></script>
<script src="site_libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/readable.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<style>h1 {font-size: 34px;}
       h1.title {font-size: 38px;}
       h2 {font-size: 30px;}
       h3 {font-size: 24px;}
       h4 {font-size: 18px;}
       h5 {font-size: 16px;}
       h6 {font-size: 12px;}
       code {color: inherit; background-color: rgba(0, 0, 0, 0.04);}
       pre:not([class]) { background-color: white }</style>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<link href="site_libs/highlightjs-9.12.0/textmate.css" rel="stylesheet" />
<script src="site_libs/highlightjs-9.12.0/highlight.js"></script>

<link rel="icon" href="https://github.com/workflowr/workflowr-assets/raw/main/img/reproducible.png">
<!-- Add a small amount of space between sections. -->
<style type="text/css">
div.section {
  padding-top: 12px;
}
</style>



<style type="text/css">
  code{white-space: pre-wrap;}
  span.smallcaps{font-variant: small-caps;}
  span.underline{text-decoration: underline;}
  div.column{display: inline-block; vertical-align: top; width: 50%;}
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  ul.task-list{list-style: none;}
    </style>

<style type="text/css">code{white-space: pre;}</style>
<script type="text/javascript">
if (window.hljs) {
  hljs.configure({languages: []});
  hljs.initHighlightingOnLoad();
  if (document.readyState && document.readyState === "complete") {
    window.setTimeout(function() { hljs.initHighlighting(); }, 0);
  }
}
</script>









<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
img {
  max-width:100%;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
details > summary > p:only-child {
  display: inline;
}
pre code {
  padding: 0;
}
</style>


<style type="text/css">
.dropdown-submenu {
  position: relative;
}
.dropdown-submenu>.dropdown-menu {
  top: 0;
  left: 100%;
  margin-top: -6px;
  margin-left: -1px;
  border-radius: 0 6px 6px 6px;
}
.dropdown-submenu:hover>.dropdown-menu {
  display: block;
}
.dropdown-submenu>a:after {
  display: block;
  content: " ";
  float: right;
  width: 0;
  height: 0;
  border-color: transparent;
  border-style: solid;
  border-width: 5px 0 5px 5px;
  border-left-color: #cccccc;
  margin-top: 5px;
  margin-right: -10px;
}
.dropdown-submenu:hover>a:after {
  border-left-color: #adb5bd;
}
.dropdown-submenu.pull-left {
  float: none;
}
.dropdown-submenu.pull-left>.dropdown-menu {
  left: -100%;
  margin-left: 10px;
  border-radius: 6px 0 6px 6px;
}
</style>

<script type="text/javascript">
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark the anchor link active (and if it's in a dropdown, also mark that active)
  var dropdown = menuAnchor.closest('li.dropdown');
  if (window.bootstrap) { // Bootstrap 4+
    menuAnchor.addClass('active');
    dropdown.find('> .dropdown-toggle').addClass('active');
  } else { // Bootstrap 3
    menuAnchor.parent().addClass('active');
    dropdown.addClass('active');
  }

  // Navbar adjustments
  var navHeight = $(".navbar").first().height() + 15;
  var style = document.createElement('style');
  var pt = "padding-top: " + navHeight + "px; ";
  var mt = "margin-top: -" + navHeight + "px; ";
  var css = "";
  // offset scroll position for anchor links (for fixed navbar)
  for (var i = 1; i <= 6; i++) {
    css += ".section h" + i + "{ " + pt + mt + "}\n";
  }
  style.innerHTML = "body {" + pt + "padding-bottom: 40px; }\n" + css;
  document.head.appendChild(style);
});
</script>

<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before, .tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "\e259";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "\e258";
  font-family: 'Glyphicons Halflings';
  border: none;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
  background-color: transparent;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<!-- code folding -->




</head>

<body>


<div class="container-fluid main-container">




<div class="navbar navbar-default  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-bs-toggle="collapse" data-target="#navbar" data-bs-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html">single-cell jamboree</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="index.html">Overview</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/stephenslab/single-cell-jamboree">source</a>
</li>
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<div id="header">



<h1 class="title toc-ignore">pancreas_celseq2_gbcd</h1>
<h4 class="author">Matthew Stephens</h4>
<h4 class="date">2025-07-02</h4>

</div>


<p>
<button type="button" class="btn btn-default btn-workflowr btn-workflowr-report" data-toggle="collapse" data-target="#workflowr-report">
<span class="glyphicon glyphicon-list" aria-hidden="true"></span>
workflowr <span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span>
</button>
</p>
<div id="workflowr-report" class="collapse">
<ul class="nav nav-tabs">
<li class="active">
<a data-toggle="tab" href="#summary">Summary</a>
</li>
<li>
<a data-toggle="tab" href="#checks"> Checks <span
class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span>
</a>
</li>
<li>
<a data-toggle="tab" href="#versions">Past versions</a>
</li>
</ul>
<div class="tab-content">
<div id="summary" class="tab-pane fade in active">
<p>
<strong>Last updated:</strong> 2025-07-03
</p>
<p>
<strong>Checks:</strong> <span
class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> 7
<span class="glyphicon glyphicon-exclamation-sign text-danger"
aria-hidden="true"></span> 0
</p>
<p>
<strong>Knit directory:</strong>
<code>single-cell-jamboree/analysis/</code> <span
class="glyphicon glyphicon-question-sign" aria-hidden="true"
title="This is the local directory in which the code in this file was executed.">
</span>
</p>
<p>
This reproducible <a href="https://rmarkdown.rstudio.com">R Markdown</a>
analysis was created with <a
  href="https://github.com/workflowr/workflowr">workflowr</a> (version
1.7.1). The <em>Checks</em> tab describes the reproducibility checks
that were applied when the results were created. The <em>Past
versions</em> tab lists the development history.
</p>
<hr>
</div>
<div id="checks" class="tab-pane fade">
<div id="workflowr-checks" class="panel-group">
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongRMarkdownfilestronguptodate">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>R Markdown file:</strong> up-to-date
</a>
</p>
</div>
<div id="strongRMarkdownfilestronguptodate"
class="panel-collapse collapse">
<div class="panel-body">
<p>Great! Since the R Markdown file has been committed to the Git
repository, you know the exact version of the code that produced these
results.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongEnvironmentstrongempty">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Environment:</strong> empty </a>
</p>
</div>
<div id="strongEnvironmentstrongempty" class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! The global environment was empty. Objects defined in the
global environment can affect the analysis in your R Markdown file in
unknown ways. For reproduciblity itâ€™s best to always run the code in an
empty environment.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongSeedstrongcodesetseed1code">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Seed:</strong>
<code>set.seed(1)</code> </a>
</p>
</div>
<div id="strongSeedstrongcodesetseed1code"
class="panel-collapse collapse">
<div class="panel-body">
<p>The command <code>set.seed(1)</code> was run prior to running the
code in the R Markdown file. Setting a seed ensures that any results
that rely on randomness, e.g. subsampling or permutations, are
reproducible.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongSessioninformationstrongrecorded">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Session information:</strong>
recorded </a>
</p>
</div>
<div id="strongSessioninformationstrongrecorded"
class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! Recording the operating system, R version, and package
versions is critical for reproducibility.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongCachestrongnone">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Cache:</strong> none </a>
</p>
</div>
<div id="strongCachestrongnone" class="panel-collapse collapse">
<div class="panel-body">
<p>Nice! There were no cached chunks for this analysis, so you can be
confident that you successfully produced the results during this
run.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongFilepathsstrongrelative">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>File paths:</strong> relative </a>
</p>
</div>
<div id="strongFilepathsstrongrelative" class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! Using relative paths to the files within your workflowr
project makes it easier to run your code on other machines.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongRepositoryversionstrongahrefhttpsgithubcomstephenslabsinglecelljamboreetree4d8c6f69ad3602c1ce6cdb49f8737df6c66733fftargetblank4d8c6f6a">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Repository version:</strong>
<a href="https://github.com/stephenslab/single-cell-jamboree/tree/4d8c6f69ad3602c1ce6cdb49f8737df6c66733ff" target="_blank">4d8c6f6</a>
</a>
</p>
</div>
<div
id="strongRepositoryversionstrongahrefhttpsgithubcomstephenslabsinglecelljamboreetree4d8c6f69ad3602c1ce6cdb49f8737df6c66733fftargetblank4d8c6f6a"
class="panel-collapse collapse">
<div class="panel-body">
<p>
Great! You are using Git for version control. Tracking code development
and connecting the code version to the results is critical for
reproducibility.
</p>
<p>
The results in this page were generated with repository version
<a href="https://github.com/stephenslab/single-cell-jamboree/tree/4d8c6f69ad3602c1ce6cdb49f8737df6c66733ff" target="_blank">4d8c6f6</a>.
See the <em>Past versions</em> tab to see a history of the changes made
to the R Markdown and HTML files.
</p>
<p>
Note that you need to be careful to ensure that all relevant files for
the analysis have been committed to Git prior to generating the results
(you can use <code>wflow_publish</code> or
<code>wflow_git_commit</code>). workflowr only checks the R Markdown
file, but you know if there are other scripts or data files that it
depends on. Below is the status of the Git repository when the results
were generated:
</p>
<pre><code>
Ignored files:
    Ignored:    .Rhistory
    Ignored:    .Rproj.user/

Untracked files:
    Untracked:  analysis/Rplot.pdf
    Untracked:  analysis/fit_pancreas_celseq2_snmf_k100.R
    Untracked:  analysis/pancreas_celseq2_snmf_k100.RData
    Untracked:  analysis/pancreas_celseq2_snmf_ms.Rmd
    Untracked:  output/pancreas_celseq2_snmf_k100.RData

Unstaged changes:
    Modified:   single-cell-jamboree.Rproj

</code></pre>
<p>
Note that any generated files, e.g.Â HTML, png, CSS, etc., are not
included in this status report because it is ok for generated content to
have uncommitted changes.
</p>
</div>
</div>
</div>
</div>
<hr>
</div>
<div id="versions" class="tab-pane fade">
<p>
These are the previous versions of the repository in which changes were
made to the R Markdown (<code>analysis/pancreas_celseq2_gbcd.Rmd</code>)
and HTML (<code>docs/pancreas_celseq2_gbcd.html</code>) files. If youâ€™ve
configured a remote Git repository (see <code>?wflow_git_remote</code>),
click on the hyperlinks in the table below to view the files as they
were in that past version.
</p>
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
File
</th>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
<th>
Message
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/stephenslab/single-cell-jamboree/blob/4d8c6f69ad3602c1ce6cdb49f8737df6c66733ff/analysis/pancreas_celseq2_gbcd.Rmd" target="_blank">4d8c6f6</a>
</td>
<td>
Matthew Stephens
</td>
<td>
2025-07-03
</td>
<td>
workflowr::wflow_publish("analysis/pancreas_celseq2_gbcd.Rmd")
</td>
</tr>
</tbody>
</table>
</div>
<hr>
</div>
</div>
</div>
<pre class="r"><code>library(gbcd)</code></pre>
<pre><code>Loading required package: ebnm</code></pre>
<pre class="r"><code>library(flashier)
library(fastTopics)
library(Matrix)
library(ggplot2)
library(cowplot)
library(ggrepel)
library(RcppML)</code></pre>
<pre><code>RcppML v0.5.5 using &#39;options(RcppML.threads = 0)&#39; (all available threads), &#39;options(RcppML.verbose = FALSE)&#39;</code></pre>
<div id="introduction" class="section level2">
<h2>Introduction</h2>
<p>I want to try fitting gbcd to the Pancreas celseq2 data.</p>
<p>First I load the data and select the CEL-seq2 data, which should
select 2,285 cells. Then I remove genes that are expressed in fewer than
10 cells, and compute the shifted log counts.</p>
<pre class="r"><code>load(&quot;../data/pancreas.RData&quot;)
set.seed(1)

# Select the CEL-seq2 data (Muraro et al, 2016).
# This should select 2,285 cells.
i           &lt;- which(sample_info$tech == &quot;celseq2&quot;)
sample_info &lt;- sample_info[i,]
counts      &lt;- counts[i,]

# Remove genes that are expressed in fewer than 10 cells.
x      &lt;- colSums(counts &gt; 0)
j      &lt;- which(x &gt; 9)
counts &lt;- counts[,j]

# Compute the shifted log counts.
a &lt;- 1
s &lt;- rowSums(counts)
s &lt;- s/mean(s)
Y &lt;- MatrixExtra::mapSparse(counts/(a*s),log1p)
dat &lt;- Matrix::tcrossprod(Y)/ncol(Y) # cell by cells covariance matrix</code></pre>
<p>Now fit a gbcd model to the data, using the <code>fit_gbcd</code>
function.</p>
<pre class="r"><code>fit.gbcd = fit_gbcd(Y, Kmax = 20)</code></pre>
<pre><code>[1] &quot;Form cell by cell covariance matrix...&quot;
   user  system elapsed 
 22.574   0.273  23.241 
[1] &quot;Initialize GEP membership matrix L...&quot;
Adding factor 1 to flash object...
Wrapping up...
Done.
Adding factor 2 to flash object...
Adding factor 3 to flash object...
Adding factor 4 to flash object...
Adding factor 5 to flash object...
Adding factor 6 to flash object...
Adding factor 7 to flash object...
Adding factor 8 to flash object...
Adding factor 9 to flash object...
Adding factor 10 to flash object...
Adding factor 11 to flash object...
Adding factor 12 to flash object...
Adding factor 13 to flash object...
Adding factor 14 to flash object...
Adding factor 15 to flash object...
Adding factor 16 to flash object...
Adding factor 17 to flash object...
Adding factor 18 to flash object...
Adding factor 19 to flash object...
Adding factor 20 to flash object...
Wrapping up...
Done.
Backfitting 20 factors (tolerance: 7.78e-02)...
  Difference between iterations is within 1.0e+04...
  Difference between iterations is within 1.0e+03...
  --Maximum number of iterations reached!
Wrapping up...
Done.
Backfitting 20 factors (tolerance: 7.78e-02)...
  Difference between iterations is within 1.0e+04...
  Difference between iterations is within 1.0e+03...
  Difference between iterations is within 1.0e+02...
  Difference between iterations is within 1.0e+01...
  Difference between iterations is within 1.0e+00...
  --Maximum number of iterations reached!
Wrapping up...
Done.
   user  system elapsed 
503.151  28.499 569.685 
[1] &quot;Estimate GEP membership matrix L...&quot;
Backfitting 39 factors (tolerance: 7.78e-02)...
An update to factor 11 decreased the objective by 6.380e-01.
An update to factor 11 decreased the objective by 4.631e-01.
  Difference between iterations is within 1.0e+05...
An update to factor 11 decreased the objective by 1.929e-01.
An update to factor 11 decreased the objective by 2.469e-01.
An update to factor 11 decreased the objective by 3.434e-01.
  Difference between iterations is within 1.0e+04...
An update to factor 11 decreased the objective by 6.332e+00.
An update to factor 11 decreased the objective by 5.223e-01.
An update to factor 11 decreased the objective by 1.362e-03.
An update to factor 11 decreased the objective by 6.924e-04.
An update to factor 11 decreased the objective by 1.297e-03.
  Difference between iterations is within 1.0e+03...
An update to factor 11 decreased the objective by 2.509e-04.
An update to factor 11 decreased the objective by 2.270e-03.
An update to factor 11 decreased the objective by 2.364e-03.
An update to factor 11 decreased the objective by 5.532e-04.
An update to factor 11 decreased the objective by 7.934e-04.
An update to factor 11 decreased the objective by 1.087e-03.
An update to factor 11 decreased the objective by 1.522e-03.
  --Maximum number of iterations reached!
Wrapping up...
Done.
Backfitting 39 factors (tolerance: 7.78e-02)...
An update to factor 11 decreased the objective by 9.776e+00.
An update to factor 11 decreased the objective by 6.211e+00.
An update to factor 11 decreased the objective by 6.497e+00.
An update to factor 11 decreased the objective by 7.659e+00.
An update to factor 11 decreased the objective by 2.487e-03.
An update to factor 11 decreased the objective by 5.518e-04.
An update to factor 11 decreased the objective by 7.381e-04.
An update to factor 11 decreased the objective by 1.239e-03.
An update to factor 11 decreased the objective by 2.026e-03.
An update to factor 11 decreased the objective by 9.996e-05.
An update to factor 11 decreased the objective by 2.296e-04.
  Difference between iterations is within 1.0e+03...
An update to factor 11 decreased the objective by 6.140e-04.
An update to factor 11 decreased the objective by 4.609e-04.
An update to factor 11 decreased the objective by 3.069e-04.
An update to factor 11 decreased the objective by 2.173e-04.
An update to factor 11 decreased the objective by 1.859e-04.
An update to factor 11 decreased the objective by 1.387e-04.
An update to factor 11 decreased the objective by 2.506e-04.
An update to factor 11 decreased the objective by 2.012e-04.
An update to factor 11 decreased the objective by 3.130e-04.
An update to factor 11 decreased the objective by 3.199e-04.
An update to factor 11 decreased the objective by 1.714e-07.
An update to factor 11 decreased the objective by 1.207e-06.
An update to factor 11 decreased the objective by 2.896e-04.
An update to factor 11 decreased the objective by 3.017e-07.
An update to factor 11 decreased the objective by 1.717e-04.
An update to factor 11 decreased the objective by 1.937e-06.
An update to factor 11 decreased the objective by 2.834e-04.
An update to factor 11 decreased the objective by 3.580e-06.
  Difference between iterations is within 1.0e+02...
An update to factor 11 decreased the objective by 3.262e-04.
An update to factor 11 decreased the objective by 1.442e-06.
An update to factor 11 decreased the objective by 3.215e-06.
An update to factor 11 decreased the objective by 1.941e-06.
An update to factor 11 decreased the objective by 3.029e-06.
An update to factor 11 decreased the objective by 3.029e-04.
An update to factor 11 decreased the objective by 3.107e-06.
An update to factor 11 decreased the objective by 1.311e-06.
An update to factor 11 decreased the objective by 8.382e-07.
An update to factor 11 decreased the objective by 7.078e-08.
An update to factor 11 decreased the objective by 2.886e-04.
An update to factor 11 decreased the objective by 2.843e-04.
An update to factor 11 decreased the objective by 7.711e-07.
An update to factor 11 decreased the objective by 2.306e-06.
An update to factor 11 decreased the objective by 3.792e-03.
  --Estimate of factor 11 is numerically zero!
An update to factor 11 decreased the objective by 2.550e-04.
  --Estimate of factor 11 is numerically zero!
An update to factor 11 decreased the objective by 2.723e-04.
  --Estimate of factor 11 is numerically zero!
An update to factor 11 decreased the objective by 2.386e-04.
An update to factor 11 decreased the objective by 1.811e-05.
An update to factor 11 decreased the objective by 8.138e-05.
  --Estimate of factor 11 is numerically zero!
An update to factor 11 decreased the objective by 2.497e-04.
An update to factor 11 decreased the objective by 1.479e-05.
An update to factor 11 decreased the objective by 7.207e-05.
An update to factor 11 decreased the objective by 2.540e-02.
  --Estimate of factor 11 is numerically zero!
An update to factor 11 decreased the objective by 1.026e-04.
  --Estimate of factor 11 is numerically zero!
An update to factor 11 decreased the objective by 8.729e-05.
  --Estimate of factor 11 is numerically zero!
An update to factor 11 decreased the objective by 5.640e-05.
  --Estimate of factor 11 is numerically zero!
An update to factor 11 decreased the objective by 3.577e-05.
  --Estimate of factor 11 is numerically zero!
An update to factor 11 decreased the objective by 4.756e-05.
  --Estimate of factor 11 is numerically zero!
An update to factor 11 decreased the objective by 5.396e-05.
  --Estimate of factor 11 is numerically zero!
An update to factor 11 decreased the objective by 6.146e-05.
  --Estimate of factor 11 is numerically zero!
An update to factor 11 decreased the objective by 3.093e-05.
  --Estimate of factor 11 is numerically zero!
An update to factor 11 decreased the objective by 4.366e-05.
  --Estimate of factor 11 is numerically zero!
An update to factor 11 decreased the objective by 4.660e-05.
  --Estimate of factor 11 is numerically zero!
An update to factor 11 decreased the objective by 5.139e-05.
  --Estimate of factor 11 is numerically zero!
An update to factor 11 decreased the objective by 5.319e-05.
  --Estimate of factor 11 is numerically zero!
An update to factor 11 decreased the objective by 6.454e-05.
  --Estimate of factor 11 is numerically zero!
An update to factor 11 decreased the objective by 8.186e-05.
  --Estimate of factor 11 is numerically zero!
An update to factor 11 decreased the objective by 8.067e-05.
  --Estimate of factor 11 is numerically zero!
An update to factor 11 decreased the objective by 8.522e-05.
An update to factor 11 decreased the objective by 7.391e-04.
  --Estimate of factor 11 is numerically zero!
An update to factor 11 decreased the objective by 6.808e-05.
  --Estimate of factor 11 is numerically zero!
An update to factor 11 decreased the objective by 4.533e-05.
  --Estimate of factor 11 is numerically zero!
An update to factor 11 decreased the objective by 3.513e-05.
  --Estimate of factor 11 is numerically zero!
An update to factor 11 decreased the objective by 2.901e-05.
  --Estimate of factor 11 is numerically zero!
An update to factor 11 decreased the objective by 1.074e-05.
  --Estimate of factor 11 is numerically zero!
  Difference between iterations is within 1.0e+01...
  --Maximum number of iterations reached!
Wrapping up...
Done.
   user  system elapsed 
534.943  28.937 565.535 
[1] &quot;Estimate GEP signature matrix F...&quot;
Backfitting 28 factors (tolerance: 4.95e-01)...
  Difference between iterations is within 1.0e+03...
  Difference between iterations is within 1.0e+02...
  Difference between iterations is within 1.0e+01...
  Difference between iterations is within 1.0e+00...
Wrapping up...
Done.
   user  system elapsed 
709.598  42.865 755.112 </code></pre>
<p>Plot the structure plot</p>
<pre class="r"><code>sample_info &lt;- transform(sample_info,celltype = factor(celltype))
celltype &lt;- sample_info$celltype
celltype &lt;-
 factor(celltype,
        c(&quot;acinar&quot;,&quot;ductal&quot;,&quot;activated_stellate&quot;,&quot;quiescent_stellate&quot;,
          &quot;endothelial&quot;,&quot;macrophage&quot;,&quot;mast&quot;,&quot;schwann&quot;,&quot;alpha&quot;,&quot;beta&quot;,
          &quot;delta&quot;,&quot;gamma&quot;,&quot;epsilon&quot;))

L &lt;- fit.gbcd$L

p1 &lt;- structure_plot(L[,-(1:10)],grouping = celltype,
                     gap = 20,perplexity = 70,n = Inf) +
  labs(y = &quot;membership&quot;,fill = &quot;factor&quot;,color = &quot;factor&quot;)
plot(p1)</code></pre>
<p><img src="figure/pancreas_celseq2_gbcd.Rmd/structure-plots-celseq2-flashier-1.png" width="768" style="display: block; margin: auto;" /></p>
<p>Volcano plot for a few GEPs</p>
<pre class="r"><code>res.gbcd = fit.gbcd
for(GEP in c(&quot;GEP8&quot;,&quot;GEP10&quot;,&quot;GEP12&quot;,&quot;GEP15&quot;,&quot;GEP23&quot;)){

  pdat &lt;- data.frame(gene = rownames(res.gbcd$F$lfc), 
                   lfc = res.gbcd$F$lfc[, GEP], 
                   z = abs(res.gbcd$F$z_score[, GEP]), 
                   lfsr = res.gbcd$F$lfsr[, GEP],
                   stringsAsFactors = FALSE)
  pdat &lt;- transform(pdat, lfsr = cut(lfsr, c(-1, 0.001, 0.01, 0.05, Inf)))
  rows  &lt;- with(pdat, which(!(abs(lfc) &gt; quantile(abs(lfc), 0.998) | (z &gt; 10))))
  pdat[rows, &quot;gene&quot;] &lt;- &quot;&quot;
  p = ggplot(pdat, aes(x = lfc, y = z, color = lfsr, label = gene)) + geom_point() + 
    geom_text_repel(color = &quot;black&quot;, size = 2.5, segment.color = &quot;black&quot;,
                  segment.size = 0.25, min.segment.length = 0,
                  max.overlaps = Inf, na.rm = TRUE) +
  scale_color_manual(values = c(&quot;coral&quot;, &quot;orange&quot;, &quot;gold&quot;, &quot;deepskyblue&quot;)) +
  labs(x = &quot;log-fold change&quot;, y = &quot;|posterior z-score|&quot;) + 
  guides(colour = guide_legend(override.aes = list(size = 2))) + 
  theme(plot.title      = element_text(hjust = 0.5,size = 12),
        axis.text       = element_text(size = 10),
        axis.title      = element_text(size = 10),
        legend.title    = element_text(size = 12),
        legend.text     = element_text(size = 10),
        legend.position = &quot;bottom&quot;) +
  ggtitle(paste0(&quot;Volcano plot of gene signature for &quot;,GEP))
  plot(p)
}</code></pre>
<p><img src="figure/pancreas_celseq2_gbcd.Rmd/unnamed-chunk-4-1.png" width="672" style="display: block; margin: auto;" /><img src="figure/pancreas_celseq2_gbcd.Rmd/unnamed-chunk-4-2.png" width="672" style="display: block; margin: auto;" /><img src="figure/pancreas_celseq2_gbcd.Rmd/unnamed-chunk-4-3.png" width="672" style="display: block; margin: auto;" /><img src="figure/pancreas_celseq2_gbcd.Rmd/unnamed-chunk-4-4.png" width="672" style="display: block; margin: auto;" /><img src="figure/pancreas_celseq2_gbcd.Rmd/unnamed-chunk-4-5.png" width="672" style="display: block; margin: auto;" /></p>
</div>
<div id="try-regular-nmf" class="section level2">
<h2>Try regular NMF</h2>
<p>Since I could not find a convenient implementation of symmetric NMF
in R I used regular nmf on the Gram matrix (note that because it is
symmetric the result tends to be close to symmetric, but it would be
nice to have a version that properly respects symmetry.)</p>
<pre class="r"><code>temp = nmf(as.matrix(dat),10)
p1 &lt;- structure_plot(temp@w,grouping = celltype,
                     gap = 20,perplexity = 70,n = Inf) +
  labs(y = &quot;membership&quot;,fill = &quot;factor&quot;,color = &quot;factor&quot;)</code></pre>
<pre><code>Running tsne on 274 x 10 matrix.</code></pre>
<pre><code>Read the 274 x 10 data matrix successfully!
Using no_dims = 1, perplexity = 70.000000, and theta = 0.100000
Computing input similarities...
Building tree...
Done in 0.03 seconds (sparsity = 0.891337)!
Learning embedding...
Iteration 50: error is 45.202019 (50 iterations in 0.02 seconds)
Iteration 100: error is 45.050823 (50 iterations in 0.02 seconds)
Iteration 150: error is 45.058666 (50 iterations in 0.02 seconds)
Iteration 200: error is 45.050021 (50 iterations in 0.02 seconds)
Iteration 250: error is 45.059067 (50 iterations in 0.02 seconds)
Iteration 300: error is 0.714175 (50 iterations in 0.02 seconds)
Iteration 350: error is 0.709455 (50 iterations in 0.01 seconds)
Iteration 400: error is 0.709442 (50 iterations in 0.01 seconds)
Iteration 450: error is 0.709448 (50 iterations in 0.01 seconds)
Iteration 500: error is 0.709448 (50 iterations in 0.01 seconds)
Iteration 550: error is 0.709448 (50 iterations in 0.01 seconds)
Iteration 600: error is 0.709443 (50 iterations in 0.01 seconds)
Iteration 650: error is 0.709448 (50 iterations in 0.01 seconds)
Iteration 700: error is 0.709448 (50 iterations in 0.01 seconds)
Iteration 750: error is 0.709443 (50 iterations in 0.01 seconds)
Iteration 800: error is 0.709448 (50 iterations in 0.01 seconds)
Iteration 850: error is 0.709448 (50 iterations in 0.01 seconds)
Iteration 900: error is 0.709443 (50 iterations in 0.01 seconds)
Iteration 950: error is 0.709448 (50 iterations in 0.01 seconds)
Iteration 1000: error is 0.709448 (50 iterations in 0.01 seconds)
Fitting performed in 0.29 seconds.</code></pre>
<pre><code>Running tsne on 258 x 10 matrix.</code></pre>
<pre><code>Read the 258 x 10 data matrix successfully!
Using no_dims = 1, perplexity = 70.000000, and theta = 0.100000
Computing input similarities...
Building tree...
Done in 0.03 seconds (sparsity = 0.934018)!
Learning embedding...
Iteration 50: error is 44.614611 (50 iterations in 0.02 seconds)
Iteration 100: error is 44.025269 (50 iterations in 0.02 seconds)
Iteration 150: error is 44.021221 (50 iterations in 0.02 seconds)
Iteration 200: error is 44.017811 (50 iterations in 0.02 seconds)
Iteration 250: error is 44.021043 (50 iterations in 0.02 seconds)
Iteration 300: error is 0.669961 (50 iterations in 0.01 seconds)
Iteration 350: error is 0.660775 (50 iterations in 0.01 seconds)
Iteration 400: error is 0.660730 (50 iterations in 0.01 seconds)
Iteration 450: error is 0.660731 (50 iterations in 0.01 seconds)
Iteration 500: error is 0.660731 (50 iterations in 0.01 seconds)
Iteration 550: error is 0.660731 (50 iterations in 0.01 seconds)
Iteration 600: error is 0.660731 (50 iterations in 0.01 seconds)
Iteration 650: error is 0.660731 (50 iterations in 0.01 seconds)
Iteration 700: error is 0.660731 (50 iterations in 0.01 seconds)
Iteration 750: error is 0.660731 (50 iterations in 0.01 seconds)
Iteration 800: error is 0.660731 (50 iterations in 0.01 seconds)
Iteration 850: error is 0.660731 (50 iterations in 0.01 seconds)
Iteration 900: error is 0.660731 (50 iterations in 0.01 seconds)
Iteration 950: error is 0.660731 (50 iterations in 0.01 seconds)
Iteration 1000: error is 0.660731 (50 iterations in 0.01 seconds)
Fitting performed in 0.26 seconds.</code></pre>
<pre><code>Running tsne on 90 x 10 matrix.</code></pre>
<pre><code>Read the 90 x 10 data matrix successfully!
Using no_dims = 1, perplexity = 28.000000, and theta = 0.100000
Computing input similarities...
Building tree...
Done in 0.00 seconds (sparsity = 0.983210)!
Learning embedding...
Iteration 50: error is 48.674819 (50 iterations in 0.00 seconds)
Iteration 100: error is 49.477593 (50 iterations in 0.00 seconds)
Iteration 150: error is 49.413597 (50 iterations in 0.00 seconds)
Iteration 200: error is 50.498457 (50 iterations in 0.00 seconds)
Iteration 250: error is 51.643978 (50 iterations in 0.00 seconds)
Iteration 300: error is 2.106221 (50 iterations in 0.00 seconds)
Iteration 350: error is 0.741434 (50 iterations in 0.00 seconds)
Iteration 400: error is 0.472935 (50 iterations in 0.00 seconds)
Iteration 450: error is 0.468116 (50 iterations in 0.00 seconds)
Iteration 500: error is 0.468110 (50 iterations in 0.00 seconds)
Iteration 550: error is 0.468109 (50 iterations in 0.00 seconds)
Iteration 600: error is 0.468110 (50 iterations in 0.00 seconds)
Iteration 650: error is 0.468110 (50 iterations in 0.00 seconds)
Iteration 700: error is 0.468110 (50 iterations in 0.00 seconds)
Iteration 750: error is 0.468110 (50 iterations in 0.00 seconds)
Iteration 800: error is 0.468110 (50 iterations in 0.00 seconds)
Iteration 850: error is 0.468110 (50 iterations in 0.00 seconds)
Iteration 900: error is 0.468110 (50 iterations in 0.00 seconds)
Iteration 950: error is 0.468110 (50 iterations in 0.00 seconds)
Iteration 1000: error is 0.468110 (50 iterations in 0.00 seconds)
Fitting performed in 0.06 seconds.</code></pre>
<pre><code>Running tsne on 21 x 10 matrix.</code></pre>
<pre><code>Read the 21 x 10 data matrix successfully!
Using no_dims = 1, perplexity = 5.000000, and theta = 0.100000
Computing input similarities...
Building tree...
Done in 0.00 seconds (sparsity = 0.829932)!
Learning embedding...
Iteration 50: error is 77.359324 (50 iterations in 0.00 seconds)
Iteration 100: error is 76.429022 (50 iterations in 0.00 seconds)
Iteration 150: error is 74.820843 (50 iterations in 0.00 seconds)
Iteration 200: error is 70.418554 (50 iterations in 0.00 seconds)
Iteration 250: error is 72.070643 (50 iterations in 0.00 seconds)
Iteration 300: error is 2.531773 (50 iterations in 0.00 seconds)
Iteration 350: error is 2.072695 (50 iterations in 0.00 seconds)
Iteration 400: error is 2.066777 (50 iterations in 0.00 seconds)
Iteration 450: error is 2.060084 (50 iterations in 0.00 seconds)
Iteration 500: error is 2.057953 (50 iterations in 0.00 seconds)
Iteration 550: error is 2.057950 (50 iterations in 0.00 seconds)
Iteration 600: error is 2.057949 (50 iterations in 0.00 seconds)
Iteration 650: error is 2.057948 (50 iterations in 0.00 seconds)
Iteration 700: error is 2.057946 (50 iterations in 0.00 seconds)
Iteration 750: error is 2.057944 (50 iterations in 0.00 seconds)
Iteration 800: error is 2.057942 (50 iterations in 0.00 seconds)
Iteration 850: error is 2.057939 (50 iterations in 0.00 seconds)
Iteration 900: error is 2.057936 (50 iterations in 0.00 seconds)
Iteration 950: error is 2.057933 (50 iterations in 0.00 seconds)
Iteration 1000: error is 2.057930 (50 iterations in 0.00 seconds)
Fitting performed in 0.01 seconds.</code></pre>
<pre><code>Running tsne on 843 x 10 matrix.</code></pre>
<pre><code>Read the 843 x 10 data matrix successfully!
Using no_dims = 1, perplexity = 70.000000, and theta = 0.100000
Computing input similarities...
Building tree...
Done in 0.10 seconds (sparsity = 0.316556)!
Learning embedding...
Iteration 50: error is 57.542228 (50 iterations in 0.06 seconds)
Iteration 100: error is 53.921400 (50 iterations in 0.06 seconds)
Iteration 150: error is 53.894658 (50 iterations in 0.05 seconds)
Iteration 200: error is 53.894272 (50 iterations in 0.05 seconds)
Iteration 250: error is 53.894271 (50 iterations in 0.05 seconds)
Iteration 300: error is 0.981008 (50 iterations in 0.05 seconds)
Iteration 350: error is 0.922532 (50 iterations in 0.05 seconds)
Iteration 400: error is 0.912849 (50 iterations in 0.05 seconds)
Iteration 450: error is 0.911205 (50 iterations in 0.05 seconds)
Iteration 500: error is 0.910978 (50 iterations in 0.05 seconds)
Iteration 550: error is 0.910931 (50 iterations in 0.05 seconds)
Iteration 600: error is 0.910915 (50 iterations in 0.05 seconds)
Iteration 650: error is 0.910908 (50 iterations in 0.05 seconds)
Iteration 700: error is 0.910906 (50 iterations in 0.05 seconds)
Iteration 750: error is 0.910901 (50 iterations in 0.05 seconds)
Iteration 800: error is 0.910901 (50 iterations in 0.05 seconds)
Iteration 850: error is 0.910898 (50 iterations in 0.05 seconds)
Iteration 900: error is 0.910900 (50 iterations in 0.05 seconds)
Iteration 950: error is 0.910897 (50 iterations in 0.05 seconds)
Iteration 1000: error is 0.910898 (50 iterations in 0.05 seconds)
Fitting performed in 1.07 seconds.</code></pre>
<pre><code>Running tsne on 445 x 10 matrix.</code></pre>
<pre><code>Read the 445 x 10 data matrix successfully!
Using no_dims = 1, perplexity = 70.000000, and theta = 0.100000
Computing input similarities...
Building tree...
Done in 0.05 seconds (sparsity = 0.577604)!
Learning embedding...
Iteration 50: error is 49.814600 (50 iterations in 0.03 seconds)
Iteration 100: error is 48.553840 (50 iterations in 0.03 seconds)
Iteration 150: error is 48.553112 (50 iterations in 0.02 seconds)
Iteration 200: error is 48.553109 (50 iterations in 0.02 seconds)
Iteration 250: error is 48.553106 (50 iterations in 0.02 seconds)
Iteration 300: error is 0.540286 (50 iterations in 0.03 seconds)
Iteration 350: error is 0.533518 (50 iterations in 0.02 seconds)
Iteration 400: error is 0.533430 (50 iterations in 0.02 seconds)
Iteration 450: error is 0.533429 (50 iterations in 0.02 seconds)
Iteration 500: error is 0.533429 (50 iterations in 0.02 seconds)
Iteration 550: error is 0.533429 (50 iterations in 0.02 seconds)
Iteration 600: error is 0.533429 (50 iterations in 0.02 seconds)
Iteration 650: error is 0.533429 (50 iterations in 0.02 seconds)
Iteration 700: error is 0.533429 (50 iterations in 0.02 seconds)
Iteration 750: error is 0.533429 (50 iterations in 0.02 seconds)
Iteration 800: error is 0.533429 (50 iterations in 0.02 seconds)
Iteration 850: error is 0.533429 (50 iterations in 0.02 seconds)
Iteration 900: error is 0.533429 (50 iterations in 0.02 seconds)
Iteration 950: error is 0.533429 (50 iterations in 0.02 seconds)
Iteration 1000: error is 0.533429 (50 iterations in 0.02 seconds)
Fitting performed in 0.49 seconds.</code></pre>
<pre><code>Running tsne on 203 x 10 matrix.</code></pre>
<pre><code>Read the 203 x 10 data matrix successfully!
Using no_dims = 1, perplexity = 66.000000, and theta = 0.100000
Computing input similarities...
Building tree...
Done in 0.02 seconds (sparsity = 0.993763)!
Learning embedding...
Iteration 50: error is 42.513384 (50 iterations in 0.01 seconds)
Iteration 100: error is 42.308155 (50 iterations in 0.01 seconds)
Iteration 150: error is 42.349078 (50 iterations in 0.01 seconds)
Iteration 200: error is 42.418540 (50 iterations in 0.01 seconds)
Iteration 250: error is 42.492489 (50 iterations in 0.01 seconds)
Iteration 300: error is 0.493282 (50 iterations in 0.01 seconds)
Iteration 350: error is 0.488524 (50 iterations in 0.01 seconds)
Iteration 400: error is 0.488526 (50 iterations in 0.01 seconds)
Iteration 450: error is 0.488526 (50 iterations in 0.01 seconds)
Iteration 500: error is 0.488526 (50 iterations in 0.01 seconds)
Iteration 550: error is 0.488525 (50 iterations in 0.01 seconds)
Iteration 600: error is 0.488526 (50 iterations in 0.01 seconds)
Iteration 650: error is 0.488526 (50 iterations in 0.01 seconds)
Iteration 700: error is 0.488526 (50 iterations in 0.01 seconds)
Iteration 750: error is 0.488526 (50 iterations in 0.01 seconds)
Iteration 800: error is 0.488526 (50 iterations in 0.01 seconds)
Iteration 850: error is 0.488526 (50 iterations in 0.01 seconds)
Iteration 900: error is 0.488526 (50 iterations in 0.01 seconds)
Iteration 950: error is 0.488526 (50 iterations in 0.01 seconds)
Iteration 1000: error is 0.488526 (50 iterations in 0.01 seconds)
Fitting performed in 0.18 seconds.</code></pre>
<pre><code>Running tsne on 110 x 10 matrix.</code></pre>
<pre><code>Read the 110 x 10 data matrix successfully!
Using no_dims = 1, perplexity = 35.000000, and theta = 0.100000
Computing input similarities...
Building tree...
Done in 0.00 seconds (sparsity = 0.987603)!
Learning embedding...
Iteration 50: error is 49.397071 (50 iterations in 0.00 seconds)
Iteration 100: error is 52.113761 (50 iterations in 0.00 seconds)
Iteration 150: error is 51.577903 (50 iterations in 0.01 seconds)
Iteration 200: error is 53.032972 (50 iterations in 0.00 seconds)
Iteration 250: error is 52.656930 (50 iterations in 0.00 seconds)
Iteration 300: error is 1.721970 (50 iterations in 0.00 seconds)
Iteration 350: error is 1.081070 (50 iterations in 0.00 seconds)
Iteration 400: error is 0.934760 (50 iterations in 0.00 seconds)
Iteration 450: error is 0.933822 (50 iterations in 0.00 seconds)
Iteration 500: error is 0.933818 (50 iterations in 0.00 seconds)
Iteration 550: error is 0.933818 (50 iterations in 0.00 seconds)
Iteration 600: error is 0.933818 (50 iterations in 0.00 seconds)
Iteration 650: error is 0.933818 (50 iterations in 0.00 seconds)
Iteration 700: error is 0.933818 (50 iterations in 0.00 seconds)
Iteration 750: error is 0.933818 (50 iterations in 0.00 seconds)
Iteration 800: error is 0.933818 (50 iterations in 0.00 seconds)
Iteration 850: error is 0.933818 (50 iterations in 0.00 seconds)
Iteration 900: error is 0.933818 (50 iterations in 0.00 seconds)
Iteration 950: error is 0.933818 (50 iterations in 0.00 seconds)
Iteration 1000: error is 0.933818 (50 iterations in 0.00 seconds)
Fitting performed in 0.08 seconds.</code></pre>
<pre class="r"><code>plot(p1)</code></pre>
<p><img src="figure/pancreas_celseq2_gbcd.Rmd/unnamed-chunk-5-1.png" width="672" style="display: block; margin: auto;" /></p>
<br>
<p>
<button type="button" class="btn btn-default btn-workflowr btn-workflowr-sessioninfo" data-toggle="collapse" data-target="#workflowr-sessioninfo" style="display: block;">
<span class="glyphicon glyphicon-wrench" aria-hidden="true"></span>
Session information
</button>
</p>
<div id="workflowr-sessioninfo" class="collapse">
<pre class="r"><code>sessionInfo()</code></pre>
<pre><code>R version 4.4.2 (2024-10-31)
Platform: aarch64-apple-darwin20
Running under: macOS Sequoia 15.5

Matrix products: default
BLAS:   /Library/Frameworks/R.framework/Versions/4.4-arm64/Resources/lib/libRblas.0.dylib 
LAPACK: /Library/Frameworks/R.framework/Versions/4.4-arm64/Resources/lib/libRlapack.dylib;  LAPACK version 3.12.0

locale:
[1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8

time zone: America/Chicago
tzcode source: internal

attached base packages:
[1] stats     graphics  grDevices utils     datasets  methods   base     

other attached packages:
[1] RcppML_0.5.6      ggrepel_0.9.6     cowplot_1.1.3     ggplot2_3.5.2    
[5] Matrix_1.7-2      fastTopics_0.7-07 flashier_1.0.56   gbcd_0.2-08      
[9] ebnm_1.1-34      

loaded via a namespace (and not attached):
 [1] tidyselect_1.2.1     viridisLite_0.4.2    dplyr_1.1.4         
 [4] farver_2.1.2         fastmap_1.2.0        lazyeval_0.2.2      
 [7] promises_1.3.3       digest_0.6.37        lifecycle_1.0.4     
[10] MatrixExtra_0.1.15   invgamma_1.1         magrittr_2.0.3      
[13] compiler_4.4.2       rlang_1.1.6          sass_0.4.10         
[16] progress_1.2.3       tools_4.4.2          yaml_2.3.10         
[19] data.table_1.17.6    knitr_1.50           labeling_0.4.3      
[22] prettyunits_1.2.0    htmlwidgets_1.6.4    scatterplot3d_0.3-44
[25] RColorBrewer_1.1-3   Rtsne_0.17           workflowr_1.7.1     
[28] withr_3.0.2          purrr_1.0.4          grid_4.4.2          
[31] git2r_0.35.0         colorspace_2.1-1     scales_1.4.0        
[34] gtools_3.9.5         cli_3.6.5            rmarkdown_2.29      
[37] crayon_1.5.3         generics_0.1.4       RcppParallel_5.1.10 
[40] rstudioapi_0.17.1    httr_1.4.7           pbapply_1.7-2       
[43] cachem_1.1.0         stringr_1.5.1        splines_4.4.2       
[46] parallel_4.4.2       softImpute_1.4-3     vctrs_0.6.5         
[49] jsonlite_2.0.0       hms_1.1.3            mixsqp_0.3-54       
[52] irlba_2.3.5.1        horseshoe_0.2.0      trust_0.1-8         
[55] plotly_4.11.0        jquerylib_0.1.4      tidyr_1.3.1         
[58] glue_1.8.0           uwot_0.2.3           stringi_1.8.7       
[61] Polychrome_1.5.4     gtable_0.3.6         later_1.4.2         
[64] quadprog_1.5-8       tibble_3.3.0         pillar_1.10.2       
[67] htmltools_0.5.8.1    float_0.3-2          truncnorm_1.0-9     
[70] R6_2.6.1             rprojroot_2.0.4      evaluate_1.0.4      
[73] lattice_0.22-6       RhpcBLASctl_0.23-42  SQUAREM_2021.1      
[76] ashr_2.2-66          httpuv_1.6.15        bslib_0.9.0         
[79] Rcpp_1.0.14          deconvolveR_1.2-1    whisker_0.4.1       
[82] xfun_0.52            fs_1.6.6             pkgconfig_2.0.3     </code></pre>
</div>
</div>
</div>
</div>
</div>
</div>
</div>


<!-- Adjust MathJax settings so that all math formulae are shown using
TeX fonts only; see
https://docs.mathjax.org/en/latest/web/configuration.html. This will make
the presentation more consistent at the cost of the webpage sometimes
taking slightly longer to load. Note that this only works because the
footer is added to webpages before the MathJax javascript. -->
<script type="text/x-mathjax-config">
  MathJax.Hub.Config({
    "HTML-CSS": { availableFonts: ["TeX"] }
  });
</script>





</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.odd').parent('tbody').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- tabsets -->

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open');
  });
});
</script>

<!-- code folding -->


<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
