<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />


<meta name="author" content="Peter Carbonetto" />


<title>Preparation and initial exploration of the “pancreas cytokine” data set</title>

<script src="site_libs/header-attrs-2.26/header-attrs.js"></script>
<script src="site_libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/readable.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<style>h1 {font-size: 34px;}
       h1.title {font-size: 38px;}
       h2 {font-size: 30px;}
       h3 {font-size: 24px;}
       h4 {font-size: 18px;}
       h5 {font-size: 16px;}
       h6 {font-size: 12px;}
       code {color: inherit; background-color: rgba(0, 0, 0, 0.04);}
       pre:not([class]) { background-color: white }</style>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<link href="site_libs/highlightjs-9.12.0/textmate.css" rel="stylesheet" />
<script src="site_libs/highlightjs-9.12.0/highlight.js"></script>

<link rel="icon" href="https://github.com/workflowr/workflowr-assets/raw/main/img/reproducible.png">
<!-- Add a small amount of space between sections. -->
<style type="text/css">
div.section {
  padding-top: 12px;
}
</style>



<style type="text/css">
  code{white-space: pre-wrap;}
  span.smallcaps{font-variant: small-caps;}
  span.underline{text-decoration: underline;}
  div.column{display: inline-block; vertical-align: top; width: 50%;}
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  ul.task-list{list-style: none;}
    </style>

<style type="text/css">code{white-space: pre;}</style>
<script type="text/javascript">
if (window.hljs) {
  hljs.configure({languages: []});
  hljs.initHighlightingOnLoad();
  if (document.readyState && document.readyState === "complete") {
    window.setTimeout(function() { hljs.initHighlighting(); }, 0);
  }
}
</script>









<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
img {
  max-width:100%;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
details > summary > p:only-child {
  display: inline;
}
pre code {
  padding: 0;
}
</style>


<style type="text/css">
.dropdown-submenu {
  position: relative;
}
.dropdown-submenu>.dropdown-menu {
  top: 0;
  left: 100%;
  margin-top: -6px;
  margin-left: -1px;
  border-radius: 0 6px 6px 6px;
}
.dropdown-submenu:hover>.dropdown-menu {
  display: block;
}
.dropdown-submenu>a:after {
  display: block;
  content: " ";
  float: right;
  width: 0;
  height: 0;
  border-color: transparent;
  border-style: solid;
  border-width: 5px 0 5px 5px;
  border-left-color: #cccccc;
  margin-top: 5px;
  margin-right: -10px;
}
.dropdown-submenu:hover>a:after {
  border-left-color: #adb5bd;
}
.dropdown-submenu.pull-left {
  float: none;
}
.dropdown-submenu.pull-left>.dropdown-menu {
  left: -100%;
  margin-left: 10px;
  border-radius: 6px 0 6px 6px;
}
</style>

<script type="text/javascript">
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark the anchor link active (and if it's in a dropdown, also mark that active)
  var dropdown = menuAnchor.closest('li.dropdown');
  if (window.bootstrap) { // Bootstrap 4+
    menuAnchor.addClass('active');
    dropdown.find('> .dropdown-toggle').addClass('active');
  } else { // Bootstrap 3
    menuAnchor.parent().addClass('active');
    dropdown.addClass('active');
  }

  // Navbar adjustments
  var navHeight = $(".navbar").first().height() + 15;
  var style = document.createElement('style');
  var pt = "padding-top: " + navHeight + "px; ";
  var mt = "margin-top: -" + navHeight + "px; ";
  var css = "";
  // offset scroll position for anchor links (for fixed navbar)
  for (var i = 1; i <= 6; i++) {
    css += ".section h" + i + "{ " + pt + mt + "}\n";
  }
  style.innerHTML = "body {" + pt + "padding-bottom: 40px; }\n" + css;
  document.head.appendChild(style);
});
</script>

<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before, .tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "\e259";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "\e258";
  font-family: 'Glyphicons Halflings';
  border: none;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
  background-color: transparent;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<!-- code folding -->




</head>

<body>


<div class="container-fluid main-container">




<div class="navbar navbar-default  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-bs-toggle="collapse" data-target="#navbar" data-bs-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html">single-cell jamboree</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="index.html">Overview</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/stephenslab/single-cell-jamboree">source</a>
</li>
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<div id="header">



<h1 class="title toc-ignore">Preparation and initial exploration of the
“pancreas cytokine” data set</h1>
<h4 class="author">Peter Carbonetto</h4>

</div>


<p>
<button type="button" class="btn btn-default btn-workflowr btn-workflowr-report" data-toggle="collapse" data-target="#workflowr-report">
<span class="glyphicon glyphicon-list" aria-hidden="true"></span>
workflowr <span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span>
</button>
</p>
<div id="workflowr-report" class="collapse">
<ul class="nav nav-tabs">
<li class="active">
<a data-toggle="tab" href="#summary">Summary</a>
</li>
<li>
<a data-toggle="tab" href="#checks"> Checks <span
class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span>
</a>
</li>
<li>
<a data-toggle="tab" href="#versions">Past versions</a>
</li>
</ul>
<div class="tab-content">
<div id="summary" class="tab-pane fade in active">
<p>
<strong>Last updated:</strong> 2025-06-11
</p>
<p>
<strong>Checks:</strong> <span
class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> 7
<span class="glyphicon glyphicon-exclamation-sign text-danger"
aria-hidden="true"></span> 0
</p>
<p>
<strong>Knit directory:</strong>
<code>single-cell-jamboree/analysis/</code> <span
class="glyphicon glyphicon-question-sign" aria-hidden="true"
title="This is the local directory in which the code in this file was executed.">
</span>
</p>
<p>
This reproducible <a href="https://rmarkdown.rstudio.com">R Markdown</a>
analysis was created with <a
  href="https://github.com/workflowr/workflowr">workflowr</a> (version
1.7.1). The <em>Checks</em> tab describes the reproducibility checks
that were applied when the results were created. The <em>Past
versions</em> tab lists the development history.
</p>
<hr>
</div>
<div id="checks" class="tab-pane fade">
<div id="workflowr-checks" class="panel-group">
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongRMarkdownfilestronguptodate">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>R Markdown file:</strong> up-to-date
</a>
</p>
</div>
<div id="strongRMarkdownfilestronguptodate"
class="panel-collapse collapse">
<div class="panel-body">
<p>Great! Since the R Markdown file has been committed to the Git
repository, you know the exact version of the code that produced these
results.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongEnvironmentstrongempty">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Environment:</strong> empty </a>
</p>
</div>
<div id="strongEnvironmentstrongempty" class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! The global environment was empty. Objects defined in the
global environment can affect the analysis in your R Markdown file in
unknown ways. For reproduciblity it’s best to always run the code in an
empty environment.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongSeedstrongcodesetseed1code">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Seed:</strong>
<code>set.seed(1)</code> </a>
</p>
</div>
<div id="strongSeedstrongcodesetseed1code"
class="panel-collapse collapse">
<div class="panel-body">
<p>The command <code>set.seed(1)</code> was run prior to running the
code in the R Markdown file. Setting a seed ensures that any results
that rely on randomness, e.g. subsampling or permutations, are
reproducible.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongSessioninformationstrongrecorded">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Session information:</strong>
recorded </a>
</p>
</div>
<div id="strongSessioninformationstrongrecorded"
class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! Recording the operating system, R version, and package
versions is critical for reproducibility.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongCachestrongnone">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Cache:</strong> none </a>
</p>
</div>
<div id="strongCachestrongnone" class="panel-collapse collapse">
<div class="panel-body">
<p>Nice! There were no cached chunks for this analysis, so you can be
confident that you successfully produced the results during this
run.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongFilepathsstrongrelative">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>File paths:</strong> relative </a>
</p>
</div>
<div id="strongFilepathsstrongrelative" class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! Using relative paths to the files within your workflowr
project makes it easier to run your code on other machines.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongRepositoryversionstrongahrefhttpsgithubcomstephenslabsinglecelljamboreetreeffad5cdff21ce9cf92bdce3b7e2f58a178e56b6dtargetblankffad5cda">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Repository version:</strong>
<a href="https://github.com/stephenslab/single-cell-jamboree/tree/ffad5cdff21ce9cf92bdce3b7e2f58a178e56b6d" target="_blank">ffad5cd</a>
</a>
</p>
</div>
<div
id="strongRepositoryversionstrongahrefhttpsgithubcomstephenslabsinglecelljamboreetreeffad5cdff21ce9cf92bdce3b7e2f58a178e56b6dtargetblankffad5cda"
class="panel-collapse collapse">
<div class="panel-body">
<p>
Great! You are using Git for version control. Tracking code development
and connecting the code version to the results is critical for
reproducibility.
</p>
<p>
The results in this page were generated with repository version
<a href="https://github.com/stephenslab/single-cell-jamboree/tree/ffad5cdff21ce9cf92bdce3b7e2f58a178e56b6d" target="_blank">ffad5cd</a>.
See the <em>Past versions</em> tab to see a history of the changes made
to the R Markdown and HTML files.
</p>
<p>
Note that you need to be careful to ensure that all relevant files for
the analysis have been committed to Git prior to generating the results
(you can use <code>wflow_publish</code> or
<code>wflow_git_commit</code>). workflowr only checks the R Markdown
file, but you know if there are other scripts or data files that it
depends on. Below is the status of the Git repository when the results
were generated:
</p>
<pre><code>
Untracked files:
    Untracked:  analysis/lps_cache/
    Untracked:  analysis/mcf7_cache/
    Untracked:  analysis/pancreas_cytokine_S1_factors_cache/
    Untracked:  analysis/temp2.R
    Untracked:  data/GSE132188_adata.h5ad.h5
    Untracked:  data/GSE183010/
    Untracked:  data/Immune_ALL_human.h5ad
    Untracked:  data/pancreas_cytokine.RData
    Untracked:  data/pancreas_endocrine.RData
    Untracked:  data/pancreas_endocrine_alldays.h5ad

Unstaged changes:
    Modified:   analysis/pancreas_cytokine_S1_factors.Rmd

</code></pre>
<p>
Note that any generated files, e.g. HTML, png, CSS, etc., are not
included in this status report because it is ok for generated content to
have uncommitted changes.
</p>
</div>
</div>
</div>
</div>
<hr>
</div>
<div id="versions" class="tab-pane fade">

<p>
These are the previous versions of the repository in which changes were
made to the R Markdown (<code>analysis/pancreas_cytokine.Rmd</code>) and
HTML (<code>docs/pancreas_cytokine.html</code>) files. If you’ve
configured a remote Git repository (see <code>?wflow_git_remote</code>),
click on the hyperlinks in the table below to view the files as they
were in that past version.
</p>
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
File
</th>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
<th>
Message
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/stephenslab/single-cell-jamboree/blob/ffad5cdff21ce9cf92bdce3b7e2f58a178e56b6d/analysis/pancreas_cytokine.Rmd" target="_blank">ffad5cd</a>
</td>
<td>
Peter Carbonetto
</td>
<td>
2025-06-11
</td>
<td>
wflow_publish("pancreas_cytokine.Rmd", verbose = TRUE, view = TRUE)
</td>
</tr>
<tr>
<td>
html
</td>
<td>
<a href="https://rawcdn.githack.com/stephenslab/single-cell-jamboree/e876f200425815e4f28291a50d1ad111886998ef/docs/pancreas_cytokine.html" target="_blank">e876f20</a>
</td>
<td>
Peter Carbonetto
</td>
<td>
2025-06-10
</td>
<td>
Ran wflow_publish("pancreas_cytokine.Rmd").
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/stephenslab/single-cell-jamboree/blob/5249817791516a7b04d8a431b008b7bd6df82610/analysis/pancreas_cytokine.Rmd" target="_blank">5249817</a>
</td>
<td>
Peter Carbonetto
</td>
<td>
2025-06-10
</td>
<td>
wflow_publish("pancreas_cytokine.Rmd", verbose = TRUE)
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/stephenslab/single-cell-jamboree/blob/62b29346d015e2c1e651fbcfa785866f6dece3fc/analysis/pancreas_cytokine.Rmd" target="_blank">62b2934</a>
</td>
<td>
Peter Carbonetto
</td>
<td>
2025-06-10
</td>
<td>
Added steps to filter out mitochondrial and ribosomal protein genes from
the pancreas_cytokine data set.
</td>
</tr>
<tr>
<td>
html
</td>
<td>
<a href="https://rawcdn.githack.com/stephenslab/single-cell-jamboree/5941ced2fc05b6379038fc30c28b6fdf94da761c/docs/pancreas_cytokine.html" target="_blank">5941ced</a>
</td>
<td>
Peter Carbonetto
</td>
<td>
2025-06-06
</td>
<td>
Added a umap plot to the pancreas_cytokine analysis.
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/stephenslab/single-cell-jamboree/blob/8168143d10ffdd7cfdcafda94affc890f8694d9d/analysis/pancreas_cytokine.Rmd" target="_blank">8168143</a>
</td>
<td>
Peter Carbonetto
</td>
<td>
2025-06-06
</td>
<td>
wflow_publish("pancreas_cytokine.Rmd", verbose = TRUE, view = FALSE)
</td>
</tr>
<tr>
<td>
html
</td>
<td>
<a href="https://rawcdn.githack.com/stephenslab/single-cell-jamboree/af8e02d63c8bba11a42eaf2b2417b83cfaeef700/docs/pancreas_cytokine.html" target="_blank">af8e02d</a>
</td>
<td>
Peter Carbonetto
</td>
<td>
2025-06-06
</td>
<td>
Build site.
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/stephenslab/single-cell-jamboree/blob/a33032feb73c6246eadb9ccd9b504c26a24ae332/analysis/pancreas_cytokine.Rmd" target="_blank">a33032f</a>
</td>
<td>
Peter Carbonetto
</td>
<td>
2025-06-06
</td>
<td>
wflow_publish("pancreas_cytokine.Rmd", verbose = TRUE, view = FALSE)
</td>
</tr>
<tr>
<td>
html
</td>
<td>
<a href="https://rawcdn.githack.com/stephenslab/single-cell-jamboree/d9cd6103626726cebc5792a9195371527be6ecb5/docs/pancreas_cytokine.html" target="_blank">d9cd610</a>
</td>
<td>
Peter Carbonetto
</td>
<td>
2025-06-06
</td>
<td>
Ran wflow_publish("pancreas_cytokine.Rmd").
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/stephenslab/single-cell-jamboree/blob/c8b096b7874ccd25dc74956ee3119094e08a6535/analysis/pancreas_cytokine.Rmd" target="_blank">c8b096b</a>
</td>
<td>
Peter Carbonetto
</td>
<td>
2025-06-06
</td>
<td>
wflow_publish("pancreas_cytokine.Rmd", verbose = TRUE, view = TRUE)
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/stephenslab/single-cell-jamboree/blob/ea78d4f9a5acc1f981b4d71a8cb346a92b050201/analysis/pancreas_cytokine.Rmd" target="_blank">ea78d4f</a>
</td>
<td>
Peter Carbonetto
</td>
<td>
2025-06-06
</td>
<td>
Fixed a bug in computing the size factors in the pancreas_cytokine
analysis.
</td>
</tr>
<tr>
<td>
html
</td>
<td>
<a href="https://rawcdn.githack.com/stephenslab/single-cell-jamboree/8081d2c4f3cf5d6e4adec876ba88b05233d455e7/docs/pancreas_cytokine.html" target="_blank">8081d2c</a>
</td>
<td>
Peter Carbonetto
</td>
<td>
2025-06-05
</td>
<td>
Build site.
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/stephenslab/single-cell-jamboree/blob/281790b5bf0b7edc001123395e9faf0070461a4d/analysis/pancreas_cytokine.Rmd" target="_blank">281790b</a>
</td>
<td>
Peter Carbonetto
</td>
<td>
2025-06-05
</td>
<td>
wflow_publish("pancreas_cytokine.Rmd", verbose = TRUE)
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/stephenslab/single-cell-jamboree/blob/e98c763dfb718b3bc4e1f035fd2ca6e1e929a559/analysis/pancreas_cytokine.Rmd" target="_blank">e98c763</a>
</td>
<td>
Peter Carbonetto
</td>
<td>
2025-06-05
</td>
<td>
Fixed the clustering of the pancreas_cytokine data and added umap plots
to the analysis.
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/stephenslab/single-cell-jamboree/blob/debdaa79333fbd94da552a57ab29e5a3e0165614/analysis/pancreas_cytokine.Rmd" target="_blank">debdaa7</a>
</td>
<td>
Peter Carbonetto
</td>
<td>
2025-06-05
</td>
<td>
Added UMAP plot + clustering of pancreas_cytokine data.
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/stephenslab/single-cell-jamboree/blob/3cac0e6b3b7e072ebe6d639e346aecc0abd77e09/analysis/pancreas_cytokine.Rmd" target="_blank">3cac0e6</a>
</td>
<td>
Peter Carbonetto
</td>
<td>
2025-06-05
</td>
<td>
Added steps to the pancreas_cytokine analysis to filter cells and genes.
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/stephenslab/single-cell-jamboree/blob/2af2dfd2f8879c46e1e2390ca3253fe7ec8104d5/analysis/pancreas_cytokine.Rmd" target="_blank">2af2dfd</a>
</td>
<td>
Peter Carbonetto
</td>
<td>
2025-06-05
</td>
<td>
Added a few notes to the pancreas_cytokine analysis.
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/stephenslab/single-cell-jamboree/blob/9e1f1279beceee0612c9ca09ce3de52ba22c38f8/analysis/pancreas_cytokine.Rmd" target="_blank">9e1f127</a>
</td>
<td>
Peter Carbonetto
</td>
<td>
2025-06-05
</td>
<td>
Added code to pancreas_cytokine analysis to prepare the scrna-seq data
downloaded from geo.
</td>
</tr>
<tr>
<td>
html
</td>
<td>
<a href="https://rawcdn.githack.com/stephenslab/single-cell-jamboree/95aea28ff888aaebdd7ff23669d50fb6c8d92183/docs/pancreas_cytokine.html" target="_blank">95aea28</a>
</td>
<td>
Peter Carbonetto
</td>
<td>
2025-06-05
</td>
<td>
First build of the pancreas_cytokine analysis.
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/stephenslab/single-cell-jamboree/blob/773b4e42da1042b85d963ebe57f0af992f43b5a3/analysis/pancreas_cytokine.Rmd" target="_blank">773b4e4</a>
</td>
<td>
Peter Carbonetto
</td>
<td>
2025-06-05
</td>
<td>
wflow_publish("pancreas_cytokine.Rmd")
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/stephenslab/single-cell-jamboree/blob/0a7a69eb94acc1a9879055fc09b02a8945b608b6/analysis/pancreas_cytokine.Rmd" target="_blank">0a7a69e</a>
</td>
<td>
Peter Carbonetto
</td>
<td>
2025-06-04
</td>
<td>
Created placeholder analysis pancreas_cytokine.Rmd.
</td>
</tr>
</tbody>
</table>
</div>
<hr>
</div>
</div>
</div>
<p>Here we will prepare the single-cell RNA-seq data from <a
href="https://doi.org/10.1093/function/zqab063">Stancill et al 2021</a>
for analysis with fastTopics and flashier. The data files were obtained
by downloading and extracting tar file GSE183010_RAW.tar from <a
href="https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE183010">GEO
accession GSE183010</a>.</p>
<p>Load the R packages used to perform the data processing and
analysis:</p>
<pre class="r"><code>library(data.table)
library(Matrix)
library(tools)
library(rsvd)
library(uwot)
library(ggplot2)
library(cowplot)
# library(MatrixSparse)</code></pre>
<p>Set the seed for reproducibility:</p>
<pre class="r"><code>set.seed(1)</code></pre>
<p>Import the count data from the “matrix market” format:</p>
<pre class="r"><code>read_geo_data &lt;- function (i, r, prefix = &quot;GSM000000&quot;, dir = &quot;.&quot;) {
  infile   &lt;- sprintf(&quot;%s_Rep%d_S%d_barcodes.tsv.gz&quot;,prefix,r,i)
  barcodes &lt;- fread(file.path(dir,infile),quote = FALSE,header = FALSE,
                    stringsAsFactors = FALSE)
  class(barcodes) &lt;- &quot;data.frame&quot;
  barcodes &lt;- barcodes[,1]
  infile &lt;- sprintf(&quot;%s_Rep%d_S%d_features.tsv.gz&quot;,prefix,r,i)
  genes  &lt;- fread(file.path(dir,infile),sep = &quot;\t&quot;,quote = FALSE,
                  header = FALSE,stringsAsFactors = FALSE)
  class(genes) &lt;- &quot;data.frame&quot;
  names(genes) &lt;- c(&quot;ensembl&quot;,&quot;symbol&quot;,&quot;type&quot;)
  genes        &lt;- transform(genes,type = factor(type))
  infile &lt;- sprintf(&quot;%s_Rep%d_S%d_matrix.mtx.gz&quot;,prefix,r,i)
  counts &lt;- fread(file.path(dir,infile),quote = FALSE,header = FALSE,
                  skip = 2)
  class(counts) &lt;- &quot;data.frame&quot;
  names(counts) &lt;- c(&quot;row&quot;,&quot;col&quot;,&quot;value&quot;)
  n &lt;- max(counts$row)
  m &lt;- max(counts$col)
  counts &lt;- sparseMatrix(i = counts$row,j = counts$col,x = counts$value,
                         dims = c(n,m))
  rownames(counts) &lt;- genes$ensembl
  colnames(counts) &lt;- barcodes
  return(list(genes  = genes,
              counts = counts))
}
dat     &lt;- vector(&quot;list&quot;,8)
dataset &lt;- 0
samples &lt;- NULL
for (r in 1:2) {
  for (i in 1:4) {
    dataset &lt;- dataset + 1
    dat[[dataset]] &lt;-
      read_geo_data(i,r,prefix = paste0(&quot;GSM55486&quot;,23 + dataset),
                    dir = &quot;../data/GSE183010&quot;)
    samples &lt;- rbind(samples,
                     data.frame(barcode   = colnames(dat[[dataset]]$counts),
                                mouse     = paste0(&quot;S&quot;,i),
                                replicate = r,
                                stringsAsFactors = FALSE))
  }
}
samples &lt;- transform(samples,
                     mouse     = factor(mouse),
                     replicate = factor(replicate))
features &lt;- Reduce(intersect,lapply(dat,function (x) x$genes$ensembl))
genes    &lt;- subset(dat[[1]]$genes,is.element(ensembl,features))
counts   &lt;- do.call(&quot;cbind&quot;,lapply(dat,function (x) x$counts[features,]))
counts   &lt;- t(counts)</code></pre>
<p>A good fraction of cells have fewer expressed genes. Let’s remove
them:</p>
<pre class="r"><code>par(mar = c(4,4,1,1))
x &lt;- rowSums(counts &gt; 0)
i &lt;- which(x &gt; 2000)
samples &lt;- samples[i,]
counts  &lt;- counts[i,]
hist(x,n = 64,main = &quot;&quot;,xlab = &quot;number of genes&quot;,ylab = &quot;number of cells&quot;)</code></pre>
<p><img src="figure/pancreas_cytokine.Rmd/filter-genes-per-cell-1.png" width="300" style="display: block; margin: auto;" /></p>
<p>
<button type="button" class="btn btn-default btn-xs btn-workflowr btn-workflowr-fig" data-toggle="collapse" data-target="#fig-filter-genes-per-cell-1">
Past versions of filter-genes-per-cell-1.png
</button>
</p>
<div id="fig-filter-genes-per-cell-1" class="collapse">
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
<a href="https://github.com/stephenslab/single-cell-jamboree/blob/d9cd6103626726cebc5792a9195371527be6ecb5/docs/figure/pancreas_cytokine.Rmd/filter-genes-per-cell-1.png" target="_blank">d9cd610</a>
</td>
<td>
Peter Carbonetto
</td>
<td>
2025-06-06
</td>
</tr>
<tr>
<td>
<a href="https://github.com/stephenslab/single-cell-jamboree/blob/8081d2c4f3cf5d6e4adec876ba88b05233d455e7/docs/figure/pancreas_cytokine.Rmd/filter-genes-per-cell-1.png" target="_blank">8081d2c</a>
</td>
<td>
Peter Carbonetto
</td>
<td>
2025-06-05
</td>
</tr>
</tbody>
</table>
</div>
</div>
<p>A small number of cells have a large proportion of mitochondrial
genes. Let’s remove those cells as well.</p>
<pre class="r"><code>par(mar = c(4,4,1,1))
mito_genes &lt;- which(substr(genes$symbol,1,2) == &quot;mt&quot;)
s          &lt;- rowSums(counts)
s_mito     &lt;- counts[,mito_genes]
prop_mito  &lt;- rowSums(s_mito)/s
i          &lt;- which(prop_mito &lt; 0.1)
samples &lt;- samples[i,]
counts  &lt;- counts[i,]
hist(prop_mito,n = 64,main = &quot;&quot;,xlab = &quot;proportion mitochondrial&quot;,
     ylab = &quot;number of cells&quot;)</code></pre>
<p><img src="figure/pancreas_cytokine.Rmd/filter-mito-1.png" width="300" style="display: block; margin: auto;" /></p>
<p>
<button type="button" class="btn btn-default btn-xs btn-workflowr btn-workflowr-fig" data-toggle="collapse" data-target="#fig-filter-mito-1">
Past versions of filter-mito-1.png
</button>
</p>
<div id="fig-filter-mito-1" class="collapse">
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
<a href="https://github.com/stephenslab/single-cell-jamboree/blob/d9cd6103626726cebc5792a9195371527be6ecb5/docs/figure/pancreas_cytokine.Rmd/filter-mito-1.png" target="_blank">d9cd610</a>
</td>
<td>
Peter Carbonetto
</td>
<td>
2025-06-06
</td>
</tr>
<tr>
<td>
<a href="https://github.com/stephenslab/single-cell-jamboree/blob/8081d2c4f3cf5d6e4adec876ba88b05233d455e7/docs/figure/pancreas_cytokine.Rmd/filter-mito-1.png" target="_blank">8081d2c</a>
</td>
<td>
Peter Carbonetto
</td>
<td>
2025-06-05
</td>
</tr>
</tbody>
</table>
</div>
</div>
<p>It seems that it might further be beneficial for some of the matrix
factorization analyses to remove the mitochondrial and ribosomal protein
genes, particularly since these genes are unlikely to be interesting.
This code is adapted from <a
href="https://nbisweden.github.io/workshop-scRNAseq/labs/seurat/seurat_01_qc.html">here</a>.</p>
<pre class="r"><code>j &lt;- which(!(grepl(&quot;^mt-&quot;,genes$symbol) |
             grepl(&quot;^Rp[sl]&quot;,genes$symbol)))
genes  &lt;- genes[j,]
counts &lt;- counts[,j]</code></pre>
<p>Remove <em>Malat1</em>, which typically shows variation for technical
reasons:</p>
<pre class="r"><code>j      &lt;- which(genes$symbol != &quot;Malat1&quot;)
genes  &lt;- genes[j,]
counts &lt;- counts[,j]</code></pre>
<p>Finally, a bunch of genes are not expressed in any cells. Let’s
remove those as well:</p>
<pre class="r"><code>x      &lt;- colSums(counts &gt; 0)
j      &lt;- which(x &gt; 0)
genes  &lt;- genes[j,]
counts &lt;- counts[,j]</code></pre>
<p>Here’s an overview of the scRNA-seq data after these data filtering
steps:</p>
<pre class="r"><code>nrow(samples)
nrow(genes)
dim(counts)
mean(counts &gt; 0)
# [1] 9354
# [1] 21792
# [1]  9354 21792
# [1] 0.2031521</code></pre>
<p>Now let’s generate a 2-d nonlinear embedding of the cells using
t-SNE. First, transform the counts into “shifted log counts”:</p>
<pre class="r"><code>a &lt;- 1
s &lt;- rowSums(counts)
s &lt;- s/mean(s)
shifted_log_counts &lt;- MatrixExtra::mapSparse(counts/(a*s),log1p)</code></pre>
<p>Then project the cells onto the top 30 PCs:</p>
<pre class="r"><code>U &lt;- rsvd(shifted_log_counts,k = 30)$u</code></pre>
<p>Now run UMAP on the 30 PCs:</p>
<pre class="r"><code>Y &lt;- umap(U,n_neighbors = 30,metric = &quot;cosine&quot;,min_dist = 0.3,
          n_threads = 8,verbose = FALSE)
x &lt;- Y[,1]
y &lt;- Y[,2]
samples$umap1 &lt;- x
samples$umap2 &lt;- y</code></pre>
<p>Specify the clusters based on the UMAP plot:</p>
<pre class="r"><code>samples$cluster &lt;- &quot;none&quot;
samples$cluster[x &gt; -3 &amp; y &gt; 5]   &lt;- &quot;islet1&quot; 
samples &lt;- transform(samples,cluster = factor(cluster))</code></pre>
<p>UMAP plot by cluster:</p>
<pre class="r"><code>cluster_colors &lt;- c(&quot;darkmagenta&quot;,&quot;royalblue&quot;,&quot;limegreen&quot;,&quot;red&quot;,&quot;darkorange&quot;)
samples$cluster          &lt;- &quot;islet&quot; # Gcg, Sst, Ins1, Ins2, Ppy, # Mafa, Mafb
samples$cluster[y &lt; -16] &lt;- &quot;duct&quot;  # Krt19
samples$cluster[x &gt; 11 &amp; y &gt; 4] &lt;- &quot;macrophage&quot; # Ccr5
samples$cluster[x &gt; 10 &amp; y &gt; -6 &amp; y &lt; 0] &lt;- 
  &quot;endothelial-mesenchymal&quot; # Esam, Pecam1, Col1a1
samples &lt;- transform(samples,cluster = factor(cluster))
ggplot(samples,aes(x = umap1,y = umap2,color = cluster)) +
  geom_point(size = 1) +
  scale_color_manual(values = cluster_colors) +
  theme_cowplot(font_size = 10)</code></pre>
<p><img src="figure/pancreas_cytokine.Rmd/umap-plot-1-1.png" width="540" style="display: block; margin: auto;" /></p>
<p>
<button type="button" class="btn btn-default btn-xs btn-workflowr btn-workflowr-fig" data-toggle="collapse" data-target="#fig-umap-plot-1-1">
Past versions of umap-plot-1-1.png
</button>
</p>
<div id="fig-umap-plot-1-1" class="collapse">
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
<a href="https://github.com/stephenslab/single-cell-jamboree/blob/e876f200425815e4f28291a50d1ad111886998ef/docs/figure/pancreas_cytokine.Rmd/umap-plot-1-1.png" target="_blank">e876f20</a>
</td>
<td>
Peter Carbonetto
</td>
<td>
2025-06-10
</td>
</tr>
<tr>
<td>
<a href="https://github.com/stephenslab/single-cell-jamboree/blob/5941ced2fc05b6379038fc30c28b6fdf94da761c/docs/figure/pancreas_cytokine.Rmd/umap-plot-1-1.png" target="_blank">5941ced</a>
</td>
<td>
Peter Carbonetto
</td>
<td>
2025-06-06
</td>
</tr>
<tr>
<td>
<a href="https://github.com/stephenslab/single-cell-jamboree/blob/d9cd6103626726cebc5792a9195371527be6ecb5/docs/figure/pancreas_cytokine.Rmd/umap-plot-1-1.png" target="_blank">d9cd610</a>
</td>
<td>
Peter Carbonetto
</td>
<td>
2025-06-06
</td>
</tr>
<tr>
<td>
<a href="https://github.com/stephenslab/single-cell-jamboree/blob/8081d2c4f3cf5d6e4adec876ba88b05233d455e7/docs/figure/pancreas_cytokine.Rmd/umap-plot-1-1.png" target="_blank">8081d2c</a>
</td>
<td>
Peter Carbonetto
</td>
<td>
2025-06-05
</td>
</tr>
</tbody>
</table>
</div>
</div>
<p>UMAP plot for the first (untreated) mouse only:</p>
<pre class="r"><code>ggplot(subset(samples,mouse == &quot;S1&quot;),
       aes(x = umap1,y = umap2,color = cluster)) +
  geom_point(size = 1) +
  scale_color_manual(values = cluster_colors) +
  theme_cowplot(font_size = 10)</code></pre>
<p><img src="figure/pancreas_cytokine.Rmd/umap-plot-2-1.png" width="540" style="display: block; margin: auto;" /></p>
<p>
<button type="button" class="btn btn-default btn-xs btn-workflowr btn-workflowr-fig" data-toggle="collapse" data-target="#fig-umap-plot-2-1">
Past versions of umap-plot-2-1.png
</button>
</p>
<div id="fig-umap-plot-2-1" class="collapse">
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
<a href="https://github.com/stephenslab/single-cell-jamboree/blob/e876f200425815e4f28291a50d1ad111886998ef/docs/figure/pancreas_cytokine.Rmd/umap-plot-2-1.png" target="_blank">e876f20</a>
</td>
<td>
Peter Carbonetto
</td>
<td>
2025-06-10
</td>
</tr>
<tr>
<td>
<a href="https://github.com/stephenslab/single-cell-jamboree/blob/5941ced2fc05b6379038fc30c28b6fdf94da761c/docs/figure/pancreas_cytokine.Rmd/umap-plot-2-1.png" target="_blank">5941ced</a>
</td>
<td>
Peter Carbonetto
</td>
<td>
2025-06-06
</td>
</tr>
<tr>
<td>
<a href="https://github.com/stephenslab/single-cell-jamboree/blob/af8e02d63c8bba11a42eaf2b2417b83cfaeef700/docs/figure/pancreas_cytokine.Rmd/umap-plot-2-1.png" target="_blank">af8e02d</a>
</td>
<td>
Peter Carbonetto
</td>
<td>
2025-06-06
</td>
</tr>
<tr>
<td>
<a href="https://github.com/stephenslab/single-cell-jamboree/blob/d9cd6103626726cebc5792a9195371527be6ecb5/docs/figure/pancreas_cytokine.Rmd/umap-plot-2-1.png" target="_blank">d9cd610</a>
</td>
<td>
Peter Carbonetto
</td>
<td>
2025-06-06
</td>
</tr>
<tr>
<td>
<a href="https://github.com/stephenslab/single-cell-jamboree/blob/8081d2c4f3cf5d6e4adec876ba88b05233d455e7/docs/figure/pancreas_cytokine.Rmd/umap-plot-2-1.png" target="_blank">8081d2c</a>
</td>
<td>
Peter Carbonetto
</td>
<td>
2025-06-05
</td>
</tr>
</tbody>
</table>
</div>
</div>
<p>UMAP plot by mouse:</p>
<pre class="r"><code>sample_colors &lt;- c(&quot;limegreen&quot;,&quot;orange&quot;,&quot;darkblue&quot;,&quot;magenta&quot;)
ggplot(samples,aes(x = umap1,y = umap2,color = mouse)) +
  geom_point(size = 1) +
  scale_color_manual(values = sample_colors) +
  theme_cowplot(font_size = 10)</code></pre>
<p><img src="figure/pancreas_cytokine.Rmd/umap-plot-3-1.png" width="390" style="display: block; margin: auto;" /></p>
<p>
<button type="button" class="btn btn-default btn-xs btn-workflowr btn-workflowr-fig" data-toggle="collapse" data-target="#fig-umap-plot-3-1">
Past versions of umap-plot-3-1.png
</button>
</p>
<div id="fig-umap-plot-3-1" class="collapse">
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
<a href="https://github.com/stephenslab/single-cell-jamboree/blob/e876f200425815e4f28291a50d1ad111886998ef/docs/figure/pancreas_cytokine.Rmd/umap-plot-3-1.png" target="_blank">e876f20</a>
</td>
<td>
Peter Carbonetto
</td>
<td>
2025-06-10
</td>
</tr>
<tr>
<td>
<a href="https://github.com/stephenslab/single-cell-jamboree/blob/af8e02d63c8bba11a42eaf2b2417b83cfaeef700/docs/figure/pancreas_cytokine.Rmd/umap-plot-3-1.png" target="_blank">af8e02d</a>
</td>
<td>
Peter Carbonetto
</td>
<td>
2025-06-06
</td>
</tr>
</tbody>
</table>
</div>
</div>
<p>Save the processed data to an RData file:</p>
<pre class="r"><code>save(list = c(&quot;samples&quot;,&quot;genes&quot;,&quot;counts&quot;),
     file = &quot;pancreas_cytokine.RData&quot;)
resaveRdaFiles(&quot;pancreas_cytokine.RData&quot;)</code></pre>
<br>
<p>
<button type="button" class="btn btn-default btn-workflowr btn-workflowr-sessioninfo" data-toggle="collapse" data-target="#workflowr-sessioninfo" style="display: block;">
<span class="glyphicon glyphicon-wrench" aria-hidden="true"></span>
Session information
</button>
</p>
<div id="workflowr-sessioninfo" class="collapse">
<pre class="r"><code>sessionInfo()
# R version 4.3.3 (2024-02-29)
# Platform: aarch64-apple-darwin20 (64-bit)
# Running under: macOS 15.4.1
# 
# Matrix products: default
# BLAS:   /Library/Frameworks/R.framework/Versions/4.3-arm64/Resources/lib/libRblas.0.dylib 
# LAPACK: /Library/Frameworks/R.framework/Versions/4.3-arm64/Resources/lib/libRlapack.dylib;  LAPACK version 3.11.0
# 
# locale:
# [1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8
# 
# time zone: America/Chicago
# tzcode source: internal
# 
# attached base packages:
# [1] tools     stats     graphics  grDevices utils     datasets  methods  
# [8] base     
# 
# other attached packages:
# [1] cowplot_1.1.3     ggplot2_3.5.0     uwot_0.2.3        rsvd_1.0.5       
# [5] Matrix_1.6-5      data.table_1.17.4
# 
# loaded via a namespace (and not attached):
#  [1] sass_0.4.9          utf8_1.2.4          generics_0.1.3     
#  [4] stringi_1.8.3       lattice_0.22-5      digest_0.6.34      
#  [7] magrittr_2.0.3      evaluate_1.0.3      grid_4.3.3         
# [10] float_0.3-2         fastmap_1.1.1       R.oo_1.26.0        
# [13] rprojroot_2.0.4     workflowr_1.7.1     jsonlite_1.8.8     
# [16] R.utils_2.12.3      whisker_0.4.1       promises_1.2.1     
# [19] fansi_1.0.6         scales_1.3.0        RhpcBLASctl_0.23-42
# [22] codetools_0.2-19    jquerylib_0.1.4     cli_3.6.4          
# [25] rlang_1.1.5         R.methodsS3_1.8.2   RcppAnnoy_0.0.22   
# [28] munsell_0.5.0       withr_3.0.2         cachem_1.0.8       
# [31] yaml_2.3.8          parallel_4.3.3      dplyr_1.1.4        
# [34] colorspace_2.1-0    httpuv_1.6.14       vctrs_0.6.5        
# [37] R6_2.5.1            lifecycle_1.0.4     git2r_0.33.0       
# [40] stringr_1.5.1       fs_1.6.5            irlba_2.3.5.1      
# [43] MatrixExtra_0.1.15  pkgconfig_2.0.3     pillar_1.9.0       
# [46] bslib_0.6.1         later_1.3.2         gtable_0.3.4       
# [49] glue_1.8.0          Rcpp_1.0.12         highr_0.10         
# [52] xfun_0.42           tibble_3.2.1        tidyselect_1.2.1   
# [55] knitr_1.45          farver_2.1.1        htmltools_0.5.8.1  
# [58] labeling_0.4.3      rmarkdown_2.26      compiler_4.3.3</code></pre>
</div>


<!-- Adjust MathJax settings so that all math formulae are shown using
TeX fonts only; see
https://docs.mathjax.org/en/latest/web/configuration.html. This will make
the presentation more consistent at the cost of the webpage sometimes
taking slightly longer to load. Note that this only works because the
footer is added to webpages before the MathJax javascript. -->
<script type="text/x-mathjax-config">
  MathJax.Hub.Config({
    "HTML-CSS": { availableFonts: ["TeX"] }
  });
</script>





</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.odd').parent('tbody').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- tabsets -->

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open');
  });
});
</script>

<!-- code folding -->


<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
