<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />


<meta name="author" content="Eric Weine and Peter Carbonetto" />


<title>Preparation and initial exploration of the LSA 2021 pancreas data set</title>

<script src="site_libs/header-attrs-2.29/header-attrs.js"></script>
<script src="site_libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/readable.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<style>h1 {font-size: 34px;}
       h1.title {font-size: 38px;}
       h2 {font-size: 30px;}
       h3 {font-size: 24px;}
       h4 {font-size: 18px;}
       h5 {font-size: 16px;}
       h6 {font-size: 12px;}
       code {color: inherit; background-color: rgba(0, 0, 0, 0.04);}
       pre:not([class]) { background-color: white }</style>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<link href="site_libs/highlightjs-9.12.0/textmate.css" rel="stylesheet" />
<script src="site_libs/highlightjs-9.12.0/highlight.js"></script>

<link rel="icon" href="https://github.com/workflowr/workflowr-assets/raw/main/img/reproducible.png">
<!-- Add a small amount of space between sections. -->
<style type="text/css">
div.section {
  padding-top: 12px;
}
</style>



<style type="text/css">
  code{white-space: pre-wrap;}
  span.smallcaps{font-variant: small-caps;}
  span.underline{text-decoration: underline;}
  div.column{display: inline-block; vertical-align: top; width: 50%;}
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  ul.task-list{list-style: none;}
    </style>

<style type="text/css">code{white-space: pre;}</style>
<script type="text/javascript">
if (window.hljs) {
  hljs.configure({languages: []});
  hljs.initHighlightingOnLoad();
  if (document.readyState && document.readyState === "complete") {
    window.setTimeout(function() { hljs.initHighlighting(); }, 0);
  }
}
</script>









<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
img {
  max-width:100%;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
details > summary > p:only-child {
  display: inline;
}
pre code {
  padding: 0;
}
</style>


<style type="text/css">
.dropdown-submenu {
  position: relative;
}
.dropdown-submenu>.dropdown-menu {
  top: 0;
  left: 100%;
  margin-top: -6px;
  margin-left: -1px;
  border-radius: 0 6px 6px 6px;
}
.dropdown-submenu:hover>.dropdown-menu {
  display: block;
}
.dropdown-submenu>a:after {
  display: block;
  content: " ";
  float: right;
  width: 0;
  height: 0;
  border-color: transparent;
  border-style: solid;
  border-width: 5px 0 5px 5px;
  border-left-color: #cccccc;
  margin-top: 5px;
  margin-right: -10px;
}
.dropdown-submenu:hover>a:after {
  border-left-color: #adb5bd;
}
.dropdown-submenu.pull-left {
  float: none;
}
.dropdown-submenu.pull-left>.dropdown-menu {
  left: -100%;
  margin-left: 10px;
  border-radius: 6px 0 6px 6px;
}
</style>

<script type="text/javascript">
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark the anchor link active (and if it's in a dropdown, also mark that active)
  var dropdown = menuAnchor.closest('li.dropdown');
  if (window.bootstrap) { // Bootstrap 4+
    menuAnchor.addClass('active');
    dropdown.find('> .dropdown-toggle').addClass('active');
  } else { // Bootstrap 3
    menuAnchor.parent().addClass('active');
    dropdown.addClass('active');
  }

  // Navbar adjustments
  var navHeight = $(".navbar").first().height() + 15;
  var style = document.createElement('style');
  var pt = "padding-top: " + navHeight + "px; ";
  var mt = "margin-top: -" + navHeight + "px; ";
  var css = "";
  // offset scroll position for anchor links (for fixed navbar)
  for (var i = 1; i <= 6; i++) {
    css += ".section h" + i + "{ " + pt + mt + "}\n";
  }
  style.innerHTML = "body {" + pt + "padding-bottom: 40px; }\n" + css;
  document.head.appendChild(style);
});
</script>

<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before, .tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "\e259";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "\e258";
  font-family: 'Glyphicons Halflings';
  border: none;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
  background-color: transparent;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<!-- code folding -->




</head>

<body>


<div class="container-fluid main-container">




<div class="navbar navbar-default  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-bs-toggle="collapse" data-target="#navbar" data-bs-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html">single-cell jamboree</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="index.html">Overview</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/stephenslab/single-cell-jamboree">source</a>
</li>
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<div id="header">



<h1 class="title toc-ignore">Preparation and initial exploration of the
LSA 2021 pancreas data set</h1>
<h4 class="author">Eric Weine and Peter Carbonetto</h4>

</div>


<p>
<button type="button" class="btn btn-default btn-workflowr btn-workflowr-report" data-toggle="collapse" data-target="#workflowr-report">
<span class="glyphicon glyphicon-list" aria-hidden="true"></span>
workflowr <span class="glyphicon glyphicon-exclamation-sign text-danger"
aria-hidden="true"></span>
</button>
</p>
<div id="workflowr-report" class="collapse">
<ul class="nav nav-tabs">
<li class="active">
<a data-toggle="tab" href="#summary">Summary</a>
</li>
<li>
<a data-toggle="tab" href="#checks"> Checks <span
class="glyphicon glyphicon-exclamation-sign text-danger"
aria-hidden="true"></span> </a>
</li>
<li>
<a data-toggle="tab" href="#versions">Past versions</a>
</li>
</ul>
<div class="tab-content">
<div id="summary" class="tab-pane fade in active">
<p>
<strong>Last updated:</strong> 2025-07-11
</p>
<p>
<strong>Checks:</strong> <span
class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> 6
<span class="glyphicon glyphicon-exclamation-sign text-danger"
aria-hidden="true"></span> 1
</p>
<p>
<strong>Knit directory:</strong>
<code>single-cell-jamboree/analysis/</code> <span
class="glyphicon glyphicon-question-sign" aria-hidden="true"
title="This is the local directory in which the code in this file was executed.">
</span>
</p>
<p>
This reproducible <a href="https://rmarkdown.rstudio.com">R Markdown</a>
analysis was created with <a
  href="https://github.com/workflowr/workflowr">workflowr</a> (version
1.7.1). The <em>Checks</em> tab describes the reproducibility checks
that were applied when the results were created. The <em>Past
versions</em> tab lists the development history.
</p>
<hr>
</div>
<div id="checks" class="tab-pane fade">
<div id="workflowr-checks" class="panel-group">
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongRMarkdownfilestronguptodate">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>R Markdown file:</strong> up-to-date
</a>
</p>
</div>
<div id="strongRMarkdownfilestronguptodate"
class="panel-collapse collapse">
<div class="panel-body">
<p>Great! Since the R Markdown file has been committed to the Git
repository, you know the exact version of the code that produced these
results.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongEnvironmentstrongempty">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Environment:</strong> empty </a>
</p>
</div>
<div id="strongEnvironmentstrongempty" class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! The global environment was empty. Objects defined in the
global environment can affect the analysis in your R Markdown file in
unknown ways. For reproduciblity it’s best to always run the code in an
empty environment.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongSeedstrongcodesetseed1code">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Seed:</strong>
<code>set.seed(1)</code> </a>
</p>
</div>
<div id="strongSeedstrongcodesetseed1code"
class="panel-collapse collapse">
<div class="panel-body">
<p>The command <code>set.seed(1)</code> was run prior to running the
code in the R Markdown file. Setting a seed ensures that any results
that rely on randomness, e.g. subsampling or permutations, are
reproducible.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongSessioninformationstrongrecorded">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Session information:</strong>
recorded </a>
</p>
</div>
<div id="strongSessioninformationstrongrecorded"
class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! Recording the operating system, R version, and package
versions is critical for reproducibility.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongCachestrongdetected">
<span class="glyphicon glyphicon-exclamation-sign text-danger"
aria-hidden="true"></span> <strong>Cache:</strong> detected </a>
</p>
</div>
<div id="strongCachestrongdetected" class="panel-collapse collapse">
<div class="panel-body">
The following chunks had caches available:
<ul>
<li>
fit-topic-model
</li>
</ul>
<p>To ensure reproducibility of the results, delete the cache directory
<code>pancreas_cytokine_lsa_clustering_cache</code> and re-run the
analysis. To have workflowr automatically delete the cache directory
prior to building the file, set <code>delete_cache = TRUE</code> when
running <code>wflow_build()</code> or <code>wflow_publish()</code>.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongFilepathsstrongrelative">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>File paths:</strong> relative </a>
</p>
</div>
<div id="strongFilepathsstrongrelative" class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! Using relative paths to the files within your workflowr
project makes it easier to run your code on other machines.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongRepositoryversionstrongahrefhttpsgithubcomstephenslabsinglecelljamboreetree17a796d815221fcdcbeb6b64bd19f93524112d32targetblank17a796da">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Repository version:</strong>
<a href="https://github.com/stephenslab/single-cell-jamboree/tree/17a796d815221fcdcbeb6b64bd19f93524112d32" target="_blank">17a796d</a>
</a>
</p>
</div>
<div
id="strongRepositoryversionstrongahrefhttpsgithubcomstephenslabsinglecelljamboreetree17a796d815221fcdcbeb6b64bd19f93524112d32targetblank17a796da"
class="panel-collapse collapse">
<div class="panel-body">
<p>
Great! You are using Git for version control. Tracking code development
and connecting the code version to the results is critical for
reproducibility.
</p>
<p>
The results in this page were generated with repository version
<a href="https://github.com/stephenslab/single-cell-jamboree/tree/17a796d815221fcdcbeb6b64bd19f93524112d32" target="_blank">17a796d</a>.
See the <em>Past versions</em> tab to see a history of the changes made
to the R Markdown and HTML files.
</p>
<p>
Note that you need to be careful to ensure that all relevant files for
the analysis have been committed to Git prior to generating the results
(you can use <code>wflow_publish</code> or
<code>wflow_git_commit</code>). workflowr only checks the R Markdown
file, but you know if there are other scripts or data files that it
depends on. Below is the status of the Git repository when the results
were generated:
</p>
<pre><code>
Untracked files:
    Untracked:  analysis/lps_cache/
    Untracked:  analysis/mcf7_cache/
    Untracked:  analysis/pancreas_cytokine_S1_factors_cache/
    Untracked:  analysis/pancreas_cytokine_lsa_clustering_cache/
    Untracked:  data/GSE132188_adata.h5ad.h5
    Untracked:  data/GSE156175_RAW/
    Untracked:  data/GSE183010/
    Untracked:  data/Immune_ALL_human.h5ad
    Untracked:  data/pancreas_cytokine.RData
    Untracked:  data/pancreas_cytokine_lsa.RData
    Untracked:  data/pancreas_endocrine.RData
    Untracked:  data/pancreas_endocrine_alldays.h5ad
    Untracked:  output/panc_cyto_lsa_res/

</code></pre>
<p>
Note that any generated files, e.g. HTML, png, CSS, etc., are not
included in this status report because it is ok for generated content to
have uncommitted changes.
</p>
</div>
</div>
</div>
</div>
<hr>
</div>
<div id="versions" class="tab-pane fade">

<p>
These are the previous versions of the repository in which changes were
made to the R Markdown
(<code>analysis/pancreas_cytokine_lsa_clustering.Rmd</code>) and HTML
(<code>docs/pancreas_cytokine_lsa_clustering.html</code>) files. If
you’ve configured a remote Git repository (see
<code>?wflow_git_remote</code>), click on the hyperlinks in the table
below to view the files as they were in that past version.
</p>
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
File
</th>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
<th>
Message
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/stephenslab/single-cell-jamboree/blob/17a796d815221fcdcbeb6b64bd19f93524112d32/analysis/pancreas_cytokine_lsa_clustering.Rmd" target="_blank">17a796d</a>
</td>
<td>
Peter Carbonetto
</td>
<td>
2025-07-11
</td>
<td>
wflow_publish("pancreas_cytokine_lsa_clustering.Rmd", verbose = T,
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/stephenslab/single-cell-jamboree/blob/6f9bee5bfc6f24558a7426918c39120f906df05a/analysis/pancreas_cytokine_lsa_clustering.Rmd" target="_blank">6f9bee5</a>
</td>
<td>
Peter Carbonetto
</td>
<td>
2025-07-11
</td>
<td>
Added back the Seurat analysis to pancreas_cytokine_lsa_clustering.Rmd.
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/stephenslab/single-cell-jamboree/blob/932597099a88745ff2267e2efa9261b05cf4066f/analysis/pancreas_cytokine_lsa_clustering.Rmd" target="_blank">9325970</a>
</td>
<td>
Peter Carbonetto
</td>
<td>
2025-07-09
</td>
<td>
A few edits to pancreas_cytokine_lsa_clustering.Rmd.
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/stephenslab/single-cell-jamboree/blob/746b8f3ba53f627e26452d2981cdcc314118d4c9/analysis/pancreas_cytokine_lsa_clustering.Rmd" target="_blank">746b8f3</a>
</td>
<td>
Peter Carbonetto
</td>
<td>
2025-07-09
</td>
<td>
A few small revisions to the pancreas_cytokine_lsa_clustering analysis.
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/stephenslab/single-cell-jamboree/blob/5ec605ffa13f606ab306c786e71a3e26a62efe18/analysis/pancreas_cytokine_lsa_clustering.Rmd" target="_blank">5ec605f</a>
</td>
<td>
Peter Carbonetto
</td>
<td>
2025-07-09
</td>
<td>
A few more edits to the pancreas_cytokine_lsa_clustering analysis.
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/stephenslab/single-cell-jamboree/blob/ac2f5f982536c9c381e2897afa6ff0e36827d98a/analysis/pancreas_cytokine_lsa_clustering.Rmd" target="_blank">ac2f5f9</a>
</td>
<td>
Peter Carbonetto
</td>
<td>
2025-07-09
</td>
<td>
A few revisions/fixes to the pancreas_cytokine_lsa_clustering analysis.
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/stephenslab/single-cell-jamboree/blob/7af0ecd58c483b71bcaf350ad008aacb3a3edc3e/analysis/pancreas_cytokine_lsa_clustering.Rmd" target="_blank">7af0ecd</a>
</td>
<td>
Eric Weine
</td>
<td>
2025-06-17
</td>
<td>
added umap plot to clustering
</td>
</tr>
<tr>
<td>
html
</td>
<td>
<a href="https://rawcdn.githack.com/stephenslab/single-cell-jamboree/7af0ecd58c483b71bcaf350ad008aacb3a3edc3e/docs/pancreas_cytokine_lsa_clustering.html" target="_blank">7af0ecd</a>
</td>
<td>
Eric Weine
</td>
<td>
2025-06-17
</td>
<td>
added umap plot to clustering
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/stephenslab/single-cell-jamboree/blob/a7e6cd59545ebb400c5e99ce03e7f500e13fd71b/analysis/pancreas_cytokine_lsa_clustering.Rmd" target="_blank">a7e6cd5</a>
</td>
<td>
Eric Weine
</td>
<td>
2025-06-16
</td>
<td>
added initial clustering of LSA pancreas cytokine paper
</td>
</tr>
<tr>
<td>
html
</td>
<td>
<a href="https://rawcdn.githack.com/stephenslab/single-cell-jamboree/a7e6cd59545ebb400c5e99ce03e7f500e13fd71b/docs/pancreas_cytokine_lsa_clustering.html" target="_blank">a7e6cd5</a>
</td>
<td>
Eric Weine
</td>
<td>
2025-06-16
</td>
<td>
added initial clustering of LSA pancreas cytokine paper
</td>
</tr>
</tbody>
</table>
</div>
<hr>
</div>
</div>
</div>
<p>Here we will prepare the single-cell RNA-seq data from <a
href="https://www.life-science-alliance.org/content/4/6/e202000949">Stancill
et al 2021</a> for analysis with the topic model and other matrix
factorization methods. The data files were obtained by downloading and
the <code>GSE156175_RAW.tar</code> file from <a
href="https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE156175">GEO
accession GSE156175</a>. To run this script, download this file from GEO
and extract the contents into a folder called “GSE156175_RAW” in the
data directory.</p>
<p>Load the R packages used to perform the data processing and initial
analyses:</p>
<pre class="r"><code>library(Matrix)
library(tools)
library(readr)
library(cowplot)
library(fastTopics)
library(dplyr)
library(Seurat)</code></pre>
<p>Set the seed for reproducibility:</p>
<pre class="r"><code>set.seed(1)</code></pre>
<p>Import the sample information:</p>
<pre class="r"><code>barcodes1&lt;-read_tsv(&quot;../data/GSE156175_RAW/GSM5842388_Rep1_S1_barcodes.tsv.gz&quot;,
                    col_names = &quot;barcode&quot;)
barcodes2 &lt;- read_tsv(&quot;../data/GSE156175_RAW/GSM4726017_S2_barcodes.tsv.gz&quot;,
                      col_names = &quot;barcode&quot;)
barcodes3 &lt;- read_tsv(&quot;../data/GSE156175_RAW/GSM4726018_S3_barcodes.tsv.gz&quot;,
                      col_names = &quot;barcode&quot;)
barcodes4 &lt;- read_tsv(&quot;../data/GSE156175_RAW/GSM4726019_S4_barcodes.tsv.gz&quot;,
                      col_names = &quot;barcode&quot;)
barcodes1$barcode &lt;- substr(barcodes1$barcode,1,nchar(barcodes1$barcode) - 2)
barcodes2$barcode &lt;- substr(barcodes2$barcode,1,nchar(barcodes2$barcode) - 2)
barcodes3$barcode &lt;- substr(barcodes3$barcode,1,nchar(barcodes3$barcode) - 2)
barcodes4$barcode &lt;- substr(barcodes4$barcode,1,nchar(barcodes4$barcode) - 2)
barcodes1$condition &lt;- &quot;Untreated&quot;
barcodes2$condition &lt;- &quot;IL-1B&quot;
barcodes3$condition &lt;- &quot;IFNg&quot;
barcodes4$condition &lt;- &quot;IL-1B_IFNg&quot;
barcodes &lt;- rbind(barcodes1, barcodes2, barcodes3, barcodes4)
barcodes$cell_bc &lt;- paste0(barcodes$barcode, &quot;_&quot;, barcodes$condition)
barcodes$condition &lt;- factor(barcodes$condition)</code></pre>
<p>Import the gene information:</p>
<pre class="r"><code>genes1 &lt;- read_tsv(&quot;../data/GSE156175_RAW/GSM5842388_Rep1_S1_features.tsv.gz&quot;,
                   col_names = c(&quot;ensembl&quot;, &quot;symbol&quot;, &quot;type&quot;))
genes2 &lt;- read_tsv(&quot;../data/GSE156175_RAW/GSM4726017_S2_features.tsv.gz&quot;,
                   col_names = c(&quot;ensembl&quot;, &quot;symbol&quot;, &quot;type&quot;))
genes3 &lt;- read_tsv(&quot;../data/GSE156175_RAW/GSM4726018_S3_features.tsv.gz&quot;,
                   col_names = c(&quot;ensembl&quot;, &quot;symbol&quot;, &quot;type&quot;))
genes4 &lt;- read_tsv(&quot;../data/GSE156175_RAW/GSM4726019_S4_features.tsv.gz&quot;,
                   col_names = c(&quot;ensembl&quot;, &quot;symbol&quot;, &quot;type&quot;))
genes &lt;- genes1
genes$type &lt;- factor(genes$type)</code></pre>
<p>Now import in the read counts:</p>
<pre class="r"><code>counts1 &lt;- t(readMM(&quot;../data/GSE156175_RAW/GSM5842388_Rep1_S1_matrix.mtx.gz&quot;))
counts2 &lt;- t(readMM(&quot;../data/GSE156175_RAW/GSM4726017_S2_matrix.mtx.gz&quot;))
counts3 &lt;- t(readMM(&quot;../data/GSE156175_RAW/GSM4726018_S3_matrix.mtx.gz&quot;))
counts4 &lt;- t(readMM(&quot;../data/GSE156175_RAW/GSM4726019_S4_matrix.mtx.gz&quot;))
rownames(counts1) &lt;- paste0(barcodes1$barcode, &quot;_&quot;, barcodes1$condition)
rownames(counts2) &lt;- paste0(barcodes2$barcode, &quot;_&quot;, barcodes2$condition)
rownames(counts3) &lt;- paste0(barcodes3$barcode, &quot;_&quot;, barcodes3$condition)
rownames(counts4) &lt;- paste0(barcodes4$barcode, &quot;_&quot;, barcodes4$condition)
colnames(counts1) &lt;- genes1$symbol
colnames(counts2) &lt;- genes2$symbol
colnames(counts3) &lt;- genes3$symbol
colnames(counts4) &lt;- genes4$symbol
counts &lt;- rbind(
  as.matrix(counts1),
  as.matrix(counts2),
  as.matrix(counts3),
  as.matrix(counts4)
)
counts &lt;- as(counts, &quot;CsparseMatrix&quot;)</code></pre>
<p>Remove genes that are not expressed in any cell:</p>
<pre class="r"><code>x      &lt;- colSums(counts)
j      &lt;- which(x &gt; 0)
genes  &lt;- genes[j,]
counts &lt;- counts[,j]</code></pre>
<p>Remove cells in terms (i) total UMI count and (ii) very few expressed
genes:</p>
<pre class="r"><code>x        &lt;- rowSums(counts &gt; 0)
i        &lt;- which(x &gt; 2000)
barcodes &lt;- barcodes[i,]
counts   &lt;- counts[i,]
x        &lt;- rowSums(counts)
i        &lt;- which(x &lt;= 60000)
barcodes &lt;- barcodes[i,]
counts   &lt;- counts[i,]</code></pre>
<p>PC: The filtering step <code>x &gt; 2000</code> seems a bit
aggressive. I think we could have retained some more cells.</p>
<p>Remove cells with high mitochondrial counts:</p>
<pre class="r"><code>mito_genes &lt;- which(substr(genes$symbol,1,2) == &quot;mt&quot;)
s          &lt;- rowSums(counts)
s_mito     &lt;- rowSums(counts[,mito_genes])
prop_mito  &lt;- s_mito/s
i          &lt;- which(prop_mito &lt; 0.1)
barcodes   &lt;- barcodes[i,]
counts     &lt;- counts[i,]</code></pre>
<p>PC: Again, the filtering step <code>prop_mito &lt; 0.1</code> seems a
bit aggressive.</p>
<p>Remove <em>MALAT1</em>, ribosomal genes, and mitochondrial genes:</p>
<pre class="r"><code>j &lt;- which(!(grepl(&quot;^mt-&quot;,genes$symbol) | grepl(&quot;^Rp[sl]&quot;,genes$symbol)))
genes  &lt;- genes[j,]
counts &lt;- counts[,j]
j      &lt;- which(genes$symbol != &quot;Malat1&quot;)
genes  &lt;- genes[j,]
counts &lt;- counts[,j]</code></pre>
<p>Remove other genes that are expressed in only 1 or 2 cells:</p>
<pre class="r"><code>x      &lt;- colSums(counts &gt; 0)
j      &lt;- which(x &gt; 2)
genes  &lt;- genes[j,]
counts &lt;- counts[,j]</code></pre>
<p>Fit a topic model with <span class="math inline">\(K =
12\)</span>.</p>
<pre class="r"><code>set.seed(1)
tm_fit &lt;- fit_poisson_nmf(X = counts,k = 12,control = list(nc = 7),
                          verbose = &quot;none&quot;)</code></pre>
<div class="alert alert-warning">
<p>
<strong>Warning:</strong> The above code chunk cached its results, but
it won’t be re-run if previous chunks it depends on are updated. If you
need to use caching, it is highly recommended to also set
<code>knitr::opts_chunk$set(autodep = TRUE)</code> at the top of the
file (in a chunk that is not cached). Alternatively, you can customize
the option <code>dependson</code> for each individual chunk that is
cached. Using either <code>autodep</code> or <code>dependson</code> will
remove this warning. See the
<a href="https://yihui.name/knitr/options/#cache"
>knitr cache options</a> for more details.
</p>
</div>
<p>Here is the Structure plot:</p>
<pre class="r"><code>set.seed(1)
structure_plot(tm_fit,n = Inf)
# Running tsne on 7606 x 12 matrix.</code></pre>
<p><img src="figure/pancreas_cytokine_lsa_clustering.Rmd/structure-plot-1.png" width="840" style="display: block; margin: auto;" /></p>
<p>Below is a heatmap annotation of the marker genes given in the
Stancill paper:</p>
<pre class="r"><code>scale_rows &lt;- function (A)
  A / apply(A,1,max)
marker_genes &lt;- c(&quot;Ins1&quot;,&quot;Ins2&quot;,&quot;Gcg&quot;,&quot;Sst&quot;,&quot;Ppy&quot;,&quot;Krt17&quot;,&quot;Ccr5&quot;,
                  &quot;Col1a1&quot;,&quot;Cd34&quot;,&quot;Cpa1&quot;)
F &lt;- poisson2multinom(tm_fit)$F
F &lt;- F[marker_genes,]
F &lt;- scale_rows(F)
rownames(F) &lt;- marker_genes
annotation_heatmap(F,select_features = &quot;all&quot;,verbose = FALSE)</code></pre>
<p><img src="figure/pancreas_cytokine_lsa_clustering.Rmd/annotation-heatmap-1.png" width="480" style="display: block; margin: auto;" /></p>
<p>Below are some comments relating my guesses about factor
associations. Many of the beta associated factors are non-specific:</p>
<ul>
<li>k1: ?</li>
<li>k2: acinar</li>
<li>k3: endothelial/mesenchymal</li>
<li>k4: ? (presumably beta)</li>
<li>k5: ductal</li>
<li>k6: deta</li>
<li>k7: ? (presumably beta)</li>
<li>k8: delta</li>
<li>k9: alpha</li>
<li>k10: gamma (PP)</li>
<li>k11: macrophage</li>
<li>k12: ? (presumably beta)</li>
</ul>
<p>Now, I do celltype assignments. I used 1/3 as a threshold on the
topic proportions to determine celltype because this seemed to look
reasonable visually. Below is the structure plot with cells grouped by
type.</p>
<pre class="r"><code>L &lt;- poisson2multinom(tm_fit)$L
L_df &lt;- as.data.frame(L)
L_df &lt;- L_df %&gt;%
  mutate(
    cluster = case_when(
      k2 &gt; (1/3) ~ &quot;Acinar&quot;,
      k3 &gt; (1/3) ~ &quot;Endothelial/Mesnchymal&quot;,
      k5 &gt; (1/3) ~ &quot;Ductal&quot;,
      k8 &gt; (1/3) ~ &quot;Delta&quot;,
      k9 &gt; (1/3) ~ &quot;Alpha&quot;,
      k10 &gt; (1/3) ~ &quot;Gamma&quot;,
      k11 &gt; (1/3) ~ &quot;Macrophage&quot;,
      TRUE ~ &quot;Beta&quot;
    )
  )
structure_plot(tm_fit, grouping = L_df$cluster, gap = 30, n = Inf)
# Running tsne on 860 x 12 matrix.
# Running tsne on 5780 x 12 matrix.
# Running tsne on 359 x 12 matrix.
# Running tsne on 101 x 12 matrix.
# Running tsne on 244 x 12 matrix.
# Running tsne on 145 x 12 matrix.
# Running tsne on 107 x 12 matrix.</code></pre>
<p><img src="figure/pancreas_cytokine_lsa_clustering.Rmd/clustering-1.png" width="840" style="display: block; margin: auto;" /></p>
<p>Overall, this clustering looks quite good, but there is a small
pocket of beta cells that have loadings on the green, black, and grey
factors, which indicate that they could also be alpha or delta cells.
Perhaps additional investigation into these cells (e.g. with a UMAP)
could help distinguish their precise types. Or, we could exclude them as
not having a clear type.</p>
<p>Next, I investigated a UMAP plot to see how our clusters looked.
Below is an elbow plot of the PCs created by Seurat:</p>
<pre class="r"><code>barcodes$celltype &lt;- L_df$cluster
counts &lt;- counts[,!duplicated(colnames(counts))]
so &lt;- CreateSeuratObject(counts = Matrix::t(counts), meta.data = barcodes)
# Warning: Setting row names on a tibble is deprecated.
so &lt;- SCTransform(so, verbose = FALSE)
so &lt;- RunPCA(so, verbose = FALSE)
ElbowPlot(so)</code></pre>
<p><img src="figure/pancreas_cytokine_lsa_clustering.Rmd/elbo-plot-1.png" width="360" style="display: block; margin: auto;" /></p>
<p>And here is a UMAP plot:</p>
<pre class="r"><code>so &lt;- RunUMAP(so, dims = 1:10, verbose = FALSE)
# Warning: The default method for RunUMAP has changed from calling Python UMAP via reticulate to the R-native UWOT using the cosine metric
# To use Python UMAP via reticulate, set umap.method to &#39;umap-learn&#39; and metric to &#39;correlation&#39;
# This message will be shown once per session
DimPlot(so, group.by = &quot;celltype&quot;) +
  theme_cowplot(font_size = 10)</code></pre>
<p><img src="figure/pancreas_cytokine_lsa_clustering.Rmd/umap-plot-1.png" width="570" style="display: block; margin: auto;" /></p>
<p>There are clearly a few cells on the UMAP plot that appear to lie in
a different cluster than their assignment. However, I’m not convinced
that this plot is any more reliable than the topic model structure
plot.</p>
<p>Save the processed data to an RData file:</p>
<pre class="r"><code>save(list = c(&quot;barcodes&quot;,&quot;genes&quot;,&quot;counts&quot;),
     file = &quot;pancreas_cytokine_lsa.RData&quot;)
# resaveRdaFiles(&quot;pancreas_cytokine.RData&quot;)</code></pre>
<br>
<p>
<button type="button" class="btn btn-default btn-workflowr btn-workflowr-sessioninfo" data-toggle="collapse" data-target="#workflowr-sessioninfo" style="display: block;">
<span class="glyphicon glyphicon-wrench" aria-hidden="true"></span>
Session information
</button>
</p>
<div id="workflowr-sessioninfo" class="collapse">
<pre class="r"><code>sessionInfo()
# R version 4.3.3 (2024-02-29)
# Platform: aarch64-apple-darwin20 (64-bit)
# Running under: macOS 15.5
# 
# Matrix products: default
# BLAS:   /Library/Frameworks/R.framework/Versions/4.3-arm64/Resources/lib/libRblas.0.dylib 
# LAPACK: /Library/Frameworks/R.framework/Versions/4.3-arm64/Resources/lib/libRlapack.dylib;  LAPACK version 3.11.0
# 
# locale:
# [1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8
# 
# time zone: America/Chicago
# tzcode source: internal
# 
# attached base packages:
# [1] tools     stats     graphics  grDevices utils     datasets  methods  
# [8] base     
# 
# other attached packages:
# [1] Seurat_5.0.3       SeuratObject_5.0.1 sp_2.1-4           dplyr_1.1.4       
# [5] fastTopics_0.7-25  cowplot_1.1.3      readr_2.1.5        Matrix_1.6-5      
# 
# loaded via a namespace (and not attached):
#   [1] RcppAnnoy_0.0.22            splines_4.3.3              
#   [3] later_1.4.2                 bitops_1.0-7               
#   [5] tibble_3.3.0                polyclip_1.10-6            
#   [7] fastDummies_1.7.3           lifecycle_1.0.4            
#   [9] mixsqp_0.3-54               rprojroot_2.0.4            
#  [11] globals_0.16.3              lattice_0.22-5             
#  [13] vroom_1.6.5                 MASS_7.3-60.0.1            
#  [15] magrittr_2.0.3              plotly_4.11.0              
#  [17] sass_0.4.10                 rmarkdown_2.29             
#  [19] jquerylib_0.1.4             yaml_2.3.10                
#  [21] httpuv_1.6.14               glmGamPoi_1.14.3           
#  [23] sctransform_0.4.1           spam_2.10-0                
#  [25] spatstat.sparse_3.0-3       reticulate_1.40.0          
#  [27] pbapply_1.7-2               RColorBrewer_1.1-3         
#  [29] abind_1.4-5                 zlibbioc_1.48.2            
#  [31] Rtsne_0.17                  quadprog_1.5-8             
#  [33] GenomicRanges_1.54.1        purrr_1.0.4                
#  [35] BiocGenerics_0.48.1         RCurl_1.98-1.14            
#  [37] git2r_0.33.0                GenomeInfoDbData_1.2.11    
#  [39] IRanges_2.36.0              S4Vectors_0.40.2           
#  [41] ggrepel_0.9.6               irlba_2.3.5.1              
#  [43] listenv_0.9.1               spatstat.utils_3.0-4       
#  [45] goftest_1.2-3               RSpectra_0.16-2            
#  [47] spatstat.random_3.2-3       fitdistrplus_1.1-11        
#  [49] parallelly_1.37.1           DelayedMatrixStats_1.24.0  
#  [51] DelayedArray_0.28.0         leiden_0.4.3.1             
#  [53] codetools_0.2-19            tidyselect_1.2.1           
#  [55] farver_2.1.2                matrixStats_1.2.0          
#  [57] stats4_4.3.3                spatstat.explore_3.2-7     
#  [59] jsonlite_2.0.0              ellipsis_0.3.2             
#  [61] progressr_0.14.0            ggridges_0.5.6             
#  [63] survival_3.5-8              progress_1.2.3             
#  [65] ica_1.0-3                   Rcpp_1.1.0                 
#  [67] glue_1.8.0                  SparseArray_1.2.4          
#  [69] gridExtra_2.3               xfun_0.52                  
#  [71] MatrixGenerics_1.14.0       GenomeInfoDb_1.38.8        
#  [73] withr_3.0.2                 fastmap_1.2.0              
#  [75] digest_0.6.37               truncnorm_1.0-9            
#  [77] R6_2.6.1                    mime_0.13                  
#  [79] colorspace_2.1-0            scattermore_1.2            
#  [81] gtools_3.9.5                tensor_1.5                 
#  [83] dichromat_2.0-0.1           spatstat.data_3.0-4        
#  [85] RhpcBLASctl_0.23-42         tidyr_1.3.1                
#  [87] generics_0.1.4              data.table_1.17.6          
#  [89] S4Arrays_1.2.1              prettyunits_1.2.0          
#  [91] httr_1.4.7                  htmlwidgets_1.6.4          
#  [93] whisker_0.4.1               uwot_0.2.3                 
#  [95] pkgconfig_2.0.3             gtable_0.3.6               
#  [97] workflowr_1.7.1             lmtest_0.9-40              
#  [99] XVector_0.42.0              htmltools_0.5.8.1          
# [101] dotCall64_1.1-1             Biobase_2.62.0             
# [103] scales_1.4.0                png_0.1-8                  
# [105] ashr_2.2-66                 knitr_1.50                 
# [107] tzdb_0.4.0                  reshape2_1.4.4             
# [109] nlme_3.1-164                cachem_1.1.0               
# [111] zoo_1.8-12                  stringr_1.5.1              
# [113] KernSmooth_2.23-22          parallel_4.3.3             
# [115] miniUI_0.1.1.1              pillar_1.11.0              
# [117] grid_4.3.3                  vctrs_0.6.5                
# [119] RANN_2.6.1                  promises_1.3.3             
# [121] xtable_1.8-4                cluster_2.1.6              
# [123] evaluate_1.0.4              invgamma_1.2               
# [125] cli_3.6.5                   compiler_4.3.3             
# [127] rlang_1.1.6                 crayon_1.5.3               
# [129] SQUAREM_2021.1              future.apply_1.11.2        
# [131] labeling_0.4.3              plyr_1.8.9                 
# [133] fs_1.6.6                    stringi_1.8.7              
# [135] viridisLite_0.4.2           deldir_2.0-4               
# [137] lazyeval_0.2.2              spatstat.geom_3.2-9        
# [139] RcppHNSW_0.6.0              hms_1.1.3                  
# [141] patchwork_1.2.0             sparseMatrixStats_1.14.0   
# [143] bit64_4.0.5                 future_1.33.2              
# [145] ggplot2_3.5.2               shiny_1.8.0                
# [147] SummarizedExperiment_1.32.0 ROCR_1.0-11                
# [149] igraph_2.0.3                RcppParallel_5.1.10        
# [151] bslib_0.9.0                 bit_4.0.5</code></pre>
</div>


<!-- Adjust MathJax settings so that all math formulae are shown using
TeX fonts only; see
https://docs.mathjax.org/en/latest/web/configuration.html. This will make
the presentation more consistent at the cost of the webpage sometimes
taking slightly longer to load. Note that this only works because the
footer is added to webpages before the MathJax javascript. -->
<script type="text/x-mathjax-config">
  MathJax.Hub.Config({
    "HTML-CSS": { availableFonts: ["TeX"] }
  });
</script>





</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.odd').parent('tbody').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- tabsets -->

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open');
  });
});
</script>

<!-- code folding -->


<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
