---
title: Initial exploration of the budding yeast data set
author: Peter Carbonetto
output: workflowr::wflow_html
---

The paper is Jackson et al *eLife* 2020
([doi:10.7554/eLife.51254][jackson-2020]).  First, I downloaded the
file `GSE125162_ALL-fastqTomat0-Counts.tsv.gz` from GEO, accession
GSE125162. (To read the data correctly using `read_tsv`, I first added
"id" to the beginning of the header.)

```{r knitr-opts, include=FALSE}
knitr::opts_chunk$set(comment = "#",collapse = TRUE,results = "hold",
                      fig.align = "center",dpi = 120)
```

First, load the packages needed for this analysis:

```{r load-pkgs}
library(Matrix)
library(tools)
library(readr)
library(rsvd)
library(uwot)
library(ggplot2)
library(cowplot)
```

Load the count data:

```{r load-data, message=FALSE}
counts <- read_tsv("../data/GSE125162_ALL-fastqTomat0-Counts.tsv.gz",
                   col_names = TRUE)
counts <- as.data.frame(counts)
ids    <- counts$id
counts <- counts[-1]
sample_info <- cbind(data.frame(id = ids),counts[6830:6834])
counts <- counts[1:6829]
counts <- as.matrix(counts)
counts <- as(counts,"CsparseMatrix")
rownames(counts) <- ids
sample_info <- transform(sample_info,
                         Genotype       = factor(Genotype),
				      	 Genotype_Group = factor(Genotype_Group),
					     Replicate      = factor(Replicate),
					     Condition      = factor(Condition))
```

This is the number of cells and genes before any filtering:

```{r examine-data-1}
nrow(counts)
ncol(counts)
```

Many cells have much smaller sequencing depth so, out of caution, I
filter out the cells with the smaller sequencing depths:

```{r filter-cells, fig.height=2.5, fig.width=3}
par(mar = c(4,4,1,1))
x <- rowSums(counts)
i <- which(x >= 1000)
sample_info <- sample_info[i,]
counts <- counts[i,]
hist(log10(x),n = 64,xlab = "",ylab = "",main = "")
```

Further, a handful of genes are expressed in only a very small number
of cells. I filter out genes that are expressed in fewer than 10
cells:

```{r filter-genes, fig.height=2.75, fig.width=3}
par(mar = c(4,4,1,1))
x <- colSums(counts > 0)
j <- which(x >= 10)
counts <- counts[,j]
hist(log10(x),n = 64,xlab = "",ylab = "",main = "")
```

This gives the number of cells, the number of genes, and the
proportion of counts that are nonzero after these filtering steps:

```{r examine-data-2}
nrow(counts)
ncol(counts)
mean(counts > 0)
```

And here is the number of cells in each of the 11 growth conditions:

```{r examine-data-3}
table(sample_info$Condition)
```

To get an initial sense for the structure underlying the data, I
generate a 2-d nonlinear embedding of the cells using UMAP. First,
I transform the counts into "shifted log counts":

```{r shifted-log-counts}
a <- 1
s <- rowSums(counts)
s <- s/mean(s)
shifted_log_counts <- MatrixExtra::mapSparse(counts/(a*s),log1p)
```

Next, I project the cells onto the top 50 PCs:

```{r pca}
set.seed(1)
U <- rsvd(shifted_log_counts,k = 50)$u
```

Then I run UMAP on the 50 PCs:

```{r embeddings}
Y <- umap(U,n_neighbors = 20,metric = "cosine",min_dist = 0.3,
          n_threads = 8,verbose = FALSE)
sample_info$umap1 <- Y[,1]
sample_info$umap2 <- Y[,2]
```

UMAP with the cells colored by growth condition:

```{r umap-vs-condition, fig.height=3.75, fig.width=5}
umap_colors <- c("#a6cee3","#1f78b4","#b2df8a","#33a02c","#fb9a99","#e31a1c",
                 "#fdbf6f","#ff7f00","#cab2d6","#6a3d9a","#ffff99")
p <- ggplot(sample_info,aes(x = umap1,y = umap2,color = Condition)) +
  geom_point(size = 1.5) +
  scale_color_manual(values = umap_colors) +
  labs(x = "UMAP 1",y = "UMAP 2") + 
  theme_cowplot(font_size = 10)
print(p)
```

From this result, it is quite clear that the predominant structure
corresponds to the different growth conditions. (This UMAP plot closely
replicates Fig. 2B of the *eLife* paper.)

Finally, save the data and UMAP results to an .Rdata file for more
convenient analysis in R:

```{r save-data, eval=FALSE}
save(list = c("sample_info","counts"),file = "yeast.RData")
resaveRdaFiles("yeast.RData")
```

TO DO:

+ How to get info on genes?

+ Save the data to an .Rdata file.

[jackson-2020]: https://doi.org/10.7554/eLife.51254
