---
title: NMF analysis of the MCF-7 data set
author: Peter Carbonetto
output: workflowr::wflow_html
---

The MCF-7 data set from [Sanford et al 2020 eLife paper][sanford2020elife]
is helpful for illustrating some of the basic ideas behind NMF
analysis of single-cell data (even though it is actually a bulk
RNA-seq data set!). This paper analyzed the transcriptional response
of human MCF-7 cells to retinoic acid and TGF-$\beta$, applied
individually and in combination. These data were downloaded from
[GEO accession GSE152749][mcf7-geo].

The following abbreviations are used in analysis:
EtOH = ethanol; RA = retinoic acid; TGFb = TGF-$\beta$.

```{r knitr-opts, include=FALSE}
knitr::opts_chunk$set(comment = "#",collapse = TRUE,results = "hold",
                      fig.align = "center",dpi = 120)
```

Load packages used to process the data, perform the analyses, and
create the plots.

```{r load-pkgs, message=FALSE}
library(rsvd)
library(fastglmpca)
library(fastTopics)
library(flashier)
library(data.table)
library(GEOquery)
library(singlecelljamboreeR)
library(ggplot2)
library(cowplot)
```

Set the seed for reproducibility:

```{r set-seed}
set.seed(1)
```

## Prepare the data for analysis with fastTopics and flashier

Load the RNA-seq counts:

```{r load-rnaseq}
counts <- fread("../data/GSE152749_raw_counts_GRCh38.p13_NCBI.tsv.gz",
                sep = "\t",header = TRUE,stringsAsFactors = FALSE)
class(counts) <- "data.frame"
rownames(counts) <- counts$GeneID
counts <- counts[,-1]
counts <- as.matrix(counts)
storage.mode(counts) <- "double"
counts <- t(counts)
ids <- rownames(counts)
```

Load the gene information:

```{r load-genes}
genes <- fread("../data/Human.GRCh38.p13.annot.tsv.gz",sep = "\t",
               header = TRUE,stringsAsFactors = FALSE)
class(genes) <- "data.frame"
genes <- genes[1:10]
genes <- transform(genes,
                   GeneType = factor(GeneType),
				   Status   = factor(Status))
```

Load the sample information:

```{r load-sample-info, message=FALSE}
geo <- getGEO(filename = "../data/GSE152749_family.soft.gz")
samples <- data.frame(id = names(GSMList(geo)),
                      treatment = sapply(GSMList(geo),
                                         function (x) Meta(x)$title))
samples <- samples[ids,]
rownames(samples) <- NULL
samples <- transform(samples,
                     EtOH = grepl("EtOH",treatment,fixed = TRUE),
                     RA   = grepl("RA",treatment,fixed = TRUE),
                     TGFb = grepl("TGFb",treatment,fixed = TRUE))
samples$label                 <- "EtOH"
samples[samples$RA,"label"]   <- "RA"
samples[samples$TGFb,"label"] <- "TGFb"
samples[with(samples,RA & TGFb),"label"] <- "RA+TGFb"
samples <- transform(samples,
                     label = factor(label,c("EtOH","RA","TGFb","RA+TGFb")))
```

Remove the non-protein-coding genes, and the genes that are expressed
in fewer than 4 samples:

```{r filter-genes}
x <- colSums(counts > 0)
i <- which(x > 3 &
           genes$GeneType == "protein-coding" &
		   genes$Status == "active")
genes  <- genes[i,]
counts <- counts[,i]
```

(Some background: it turns out that there is some structure in the
non-coding RNA genes that is is picked up by the Poisson-based methods
but not by the Gaussian-based methods, so to avoid this complication
in the comparisons I have removed these genes from the data, which
aren't of interest anyhow.)

This is the dimension of the data set we will analyze:

```{r dim-counts}
dim(counts)
```

For the Gaussian-based analyses, we will need the shifted log counts:

```{r shifted-log-counts}
a <- 1
s <- rowSums(counts)
s <- s/mean(s)
shifted_log_counts <- log1p(counts/(a*s))
```

## PCA and GLM-PCA

Let's see what happens when we apply PCA to the shifted log counts and
GLM-PCA to the counts. As we will see, PCA and GLM-PCA both show a
clear clustering of the data that corresponds to the different
treatments. However, the PCs are not individually interpretable.

Run PCA on the shifted log counts, and plot the first 2 PCs:

```{r pca, fig.height=2, fig.width=3}
pca <- rpca(shifted_log_counts,k = 2,center = TRUE,scale = FALSE)
colnames(pca$x) <- c("PC1","PC2")
pdat <- data.frame(samples,pca$x)
ggplot(pdat,aes(x = PC1,y = PC2,color = label)) +
  geom_point() +
  scale_color_manual(values = c("dodgerblue","tomato","darkblue",
                                "limegreen")) + 
  theme_cowplot(font_size = 10)
```

GLM-PCA applied to the counts essentially produces the same result
(aside from an arbitrary rotation):

```{r fastglmpca, fig.height=2, fig.width=3, message=FALSE}
fit_glmpca <- init_glmpca_pois(t(counts),K = 2)
fit_glmpca <- fit_glmpca_pois(t(counts),fit0 = fit_glmpca,verbose = FALSE,
                              control = list(maxiter = 50))
colnames(fit_glmpca$V) <- c("k1","k2")
pdat <- data.frame(samples,fit_glmpca$V)
ggplot(pdat,aes(x = k1,y = k2,color = label)) +
  geom_point() +
  scale_color_manual(values = c("dodgerblue","tomato","darkblue",
                                "limegreen")) + 
  theme_cowplot(font_size = 10)
```

## Topic model (fastTopics)

Fit a topic model with $K = 3$ topics to the counts:

```{r fit-topic-model}
tm0 <- fit_poisson_nmf(counts,k = 3,init.method = "random",
                       numiter = 50,verbose = "none",
     				   control = list(nc = 4,extrapolate = FALSE))
tm <- fit_poisson_nmf(counts,fit0 = tm0,numiter = 50,verbose = "none",
                      control = list(nc = 4,extrapolate = TRUE))
```

Structure plot comparing the topics to the clusters or treatment
conditions:

```{r structure-plot-topic-model, fig.height=1.5, fig.width=5}
topic_colors <- c("tomato","darkblue","dodgerblue")
n    <- nrow(counts)
L    <- poisson2multinom(tm)$L
structure_plot(L,grouping = samples$label,topics = 1:3,
               loadings_order = 1:n,colors = topic_colors) +
  theme(axis.text.x = element_text(angle = 0,hjust = 0.5))
```

## EBNMF (flashier)

Next fit an NMF to the shifted log counts using flashier, with $K = 4$:

```{r flashier-nmf, cache=TRUE}
n  <- nrow(counts)
x  <- rpois(1e7,1/n)
s1 <- sd(log(x + 1))
fl_nmf <- flash(shifted_log_counts,ebnm_fn = ebnm_point_exponential,
                var_type = 2, greedy_Kmax = 4,S = s1,backfit = TRUE,
				verbose = 0)
```

EBNMF applied to the shifted log count decomposes the samples in a
similar way as the topic model:

```{r structure-plot-nmf, fig.height=2.25, fig.width=5}
topic_colors <- c("olivedrab","dodgerblue","darkblue","tomato")
L <- ldf(fl_nmf,type = "i")$L
structure_plot(L,grouping = samples$label,topics = 4:1,
               loadings_order = 1:n,colors = topic_colors) +
  labs(y = "membership") +
  theme(axis.text.x = element_text(angle = 0,hjust = 0.5))
```

[sanford2020elife]: https://doi.org/10.7554/eLife.59388
[mcf7-geo]: https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE152749

