---
title: "Analysis of Single-cell (Spatial) Transcriptomics From Mouse Embryos"
author: "Ziang Zhang"
date: "2024-11-15"
output: workflowr::wflow_html
---

## Data

This single cell RNA data is available [here](https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE166692), as studied in [Srivatsan et al, 2021](https://www.science.org/doi/10.1126/science.abb9536).

The dataset considers $108725$ cells and $39198$ genes (after QC), measured with spatial locations in the mouse embryo. 

```{r eval=TRUE, message=FALSE, warning=FALSE}
library(Seurat)
library(Matrix)
library(data.table)
library(flashier)
library(ggplot2)
library(patchwork)
library(cowplot)
library(RColorBrewer)
library(Biobase)
library(ggpubr)
library(gridExtra)
library(fastTopics)
source('../code/plot_loadings_on_umap.R')
source("../code/Customized_Plots.R")
seurat_object <- readRDS("../data/mouse_embryo/processed_seurat/seurat_object.rds")
Y <- t(seurat_object$RNA$data)
```

The data contains the UMAP information that we can directly use for visualization, and compare with the "anatomical annotation" and the "cluster label" in [Srivatsan et al, 2021](https://www.science.org/doi/10.1126/science.abb9536).

```{r eval=TRUE, fig.width=12, fig.height=5, message=FALSE, results='hide', warning=FALSE}
umap_original_embeddings <- cbind(seurat_object$umap1, seurat_object$umap2)
p1 <- DimPlotSagnik(umap_original_embeddings, group.by = seurat_object$anatomical_annotation, pt.size = 1) +
  theme(plot.title = element_text(hjust = 0.5)) +
  ggtitle("UMAP by Anatomical Annotation")

p2 <- DimPlotSagnik(umap_original_embeddings, group.by = seurat_object$final_cluster_label, pt.size = 1) +
  theme(plot.title = element_text(hjust = 0.5)) +
  ggtitle("UMAP by Cluster Label")

# Combine plots side-by-side
combined_plot <- p1 + p2 + plot_layout(ncol = 2)
combined_plot
```


## Trying EBMF

Let's try to implement the EBMF algorithm with the `flashier` package, with different constraints and prior distributions.

```{r eval=FALSE,message=FALSE, warning=FALSE}
cols_to_keep <- colSums(Y != 0, na.rm = TRUE) > 0
Y <- Y[, cols_to_keep]
n  <- nrow(Y)
x  <- rpois(1e7, 1/n)
s1 <- sd(log(x + 1))
a <- 1
size_factors <- rowSums(Y)
size_factors <- size_factors / mean(size_factors)
# shifted_log_counts <- log1p(Y / (a * size_factors))
shifted_log_counts <- MatrixExtra::mapSparse(Y/(a*size_factors),log1p)
```

### Non-Negative EBMF

For the first EBMF problem, we consider non-negative EBMF with `point_exponential` priors:

```{r eval=FALSE}
flashier_fit_nn <- flash(shifted_log_counts,
                      ebnm_fn = ebnm_point_exponential,
                      var_type = 2,
                      greedy_Kmax = 25,
                      S = s1,
                      backfit = T)
plot(flashier_fit_nn, 
     plot_type = "structure",
     pm_which = "loadings", 
     pm_groups = seurat_object$final_cluster_label,
     bins = 20, gap = 70)
```

```{r, eval=TRUE, message=FALSE, results='hide',fig.keep='all', echo=FALSE}
# flashier_fit_nn <- readRDS("../output/mouse_embryo/flashier_fit_nn.rds")
flashier_fit_nn <- readRDS("../output/mouse_embryo/flashier_fit_nn_backfit.rds")
plot(flashier_fit_nn, 
     plot_type = "structure",
     pm_which = "loadings", 
     pm_groups = seurat_object$final_cluster_label,
     bins = 20, gap = 70)
```


### Semi Non-Negative EBMF

```{r eval=FALSE}
ebnm_fn_list <- list()
ebnm_fn_list[[1]] <- flash_ebnm(
  prior_family = "point_exponential"
)
ebnm_fn_list[[2]] <- flash_ebnm(
  prior_family = "point_normal",
  mode = "estimate"
)
flashier_fit_semi <- flash(shifted_log_counts,
                      ebnm_fn = ebnm_fn_list,
                      var_type = 2,
                      greedy_Kmax = 25,
                      S = s1,
                      backfit = T)
plot(flashier_fit_semi, 
     plot_type = "structure",
     pm_which = "loadings", 
     pm_groups = seurat_object$final_cluster_label,
     bins = 20, gap = 70)
```

```{r eval=TRUE, message=FALSE, results='hide',fig.keep='all', echo=FALSE}
# flashier_fit_semi <- readRDS("../output/mouse_embryo/flashier_fit_semi.rds")
flashier_fit_semi <- readRDS("../output/mouse_embryo/flashier_fit_semi_backfit.rds")
ebnm_fn_list <- list()
ebnm_fn_list[[1]] <- flash_ebnm(
  prior_family = "point_exponential"
)
ebnm_fn_list[[2]] <- flash_ebnm(
  prior_family = "point_normal",
  mode = "estimate"
)
plot(flashier_fit_semi, 
     plot_type = "structure",
     pm_which = "loadings", 
     pm_groups = seurat_object$final_cluster_label,
     bins = 20, gap = 70)
```


### GB Semi Non-Negative EBMF

```{r eval=FALSE}
ebnm_fn_list <- list()
ebnm_fn_list[[1]] <- flash_ebnm(
  prior_family = "generalized_binary"
)
ebnm_fn_list[[2]] <- flash_ebnm(
  prior_family = "point_laplace",
  mode = "estimate"
)
flashier_fit_gbcd <- flash(shifted_log_counts,
                      ebnm_fn = ebnm_fn_list,
                      var_type = 2,
                      greedy_Kmax = 25,
                      S = s1,
                      backfit = T)
plot(flashier_fit_gbcd, 
     plot_type = "structure",
     pm_which = "loadings", 
     pm_groups = seurat_object$final_cluster_label,
     bins = 20, gap = 70)
```

```{r eval=TRUE, message=FALSE, results='hide',fig.keep='all', echo=FALSE}
# flashier_fit_gbcd <- readRDS("../output/mouse_embryo/flashier_fit_gbcd.rds")
flashier_fit_gbcd <- readRDS("../output/mouse_embryo/flashier_fit_gbcd_backfit.rds")
ebnm_fn_list <- list()
ebnm_fn_list[[1]] <- flash_ebnm(
  prior_family = "generalized_binary"
)
ebnm_fn_list[[2]] <- flash_ebnm(
  prior_family = "point_laplace",
  mode = "estimate"
)
plot(flashier_fit_gbcd, 
     plot_type = "structure",
     pm_which = "loadings", 
     pm_groups = seurat_object$final_cluster_label,
     bins = 20, gap = 70)
```

### GBCD

```{r eval=FALSE}
res.gbcd <- fit_gbcd(Y = Y, Kmax = 25, maxiter1 = 100,
                     maxiter2 = 50, maxiter3 = 50, 
                     prior = flash_ebnm(prior_family = "generalized_binary",
                                        scale = 0.04))
saveRDS(res.gbcd, "../output/mouse_embryo/res.gbcd.rds")
```

```{r include=FALSE}
# res.gbcd <- readRDS("../output/mouse_embryo/res.gbcd.rds")
# anno <- data.frame(sample = seurat_object$sample, subtype = seurat_object$final_cluster_label)
# rownames(anno) <- rownames(res.gbcd$L)
# anno_colors <- list(sample = seurat_object$sample, subtype = seurat_object$final_cluster_label)
# cols <- colorRampPalette(c("gray96", "red"))(50)
# brks <- seq(0, 1, 0.02)
```



## Trying Topic Models

Then, let's try fitting a topic model to this dataset using `FastTopics`.

```{r eval = FALSE}
fasttopics_fit <- fit_topic_model(Y, k = 25)
```

```{r include=FALSE}
fasttopics_fit <- readRDS("../output/mouse_embryo/fasttopics_fit.rds")
```

```{r eval=TRUE, message=FALSE, results='hide',fig.keep='all', echo=FALSE}
structure_plot(fasttopics_fit, grouping = seurat_object$final_cluster_label, gap = 70)
```


## Summarize Structures from Loadings

There are some interesting structures of the loading based on the result of EBMF and FastTopics, that add new insights to the original UMAP visualization.

### Non-negative EBMF

First, based on the non-negative EBMF result, there are some factors that are unique to certain cell types. For example:

- The factor $k9$ (mostly) is unique to Endothelial Cells.

- The factor $k13$ is unique to Choroid Plexus.

- The factor $k15$ is unique to White Blood Cells.

```{r eval=TRUE, message=FALSE, results='hide',fig.keep='all', echo=FALSE}
# Collect all plots in a list
plots <- list()
colnames(umap_original_embeddings) <- c("UMAP_1", "UMAP_2")

for (f in c(9, 13, 15)) {
  loading <- flashier_fit_nn$L_pm[, f]
  p <- plot_loadings_on_umap(umap = umap_original_embeddings, loading = loading, factor_num = f, size = 1)
  plots[[length(plots) + 1]] <- ggplotGrob(p)
}

# Combine the plots in a 2x2 grid
combined_plot <- wrap_plots(plots, ncol = 2, nrow = 2)

# Display the combined plot
print(combined_plot)
```


Shared factors include:

- The factor $k2$ is shared across Erythroid Progenitors and Hepatocytes.

- The factors $k3$ and $k10$ are shared across multiple types of neurons and glial cells.

- The factor $k4$ is shared across Cardiac Muscle lineages and Myocytes.

- The factor $k7$ is shared across Erythroid Progenitors and White Blood Cells.



```{r eval=TRUE, message=FALSE, results='hide',fig.keep='all', echo=FALSE}
# Collect all plots in a list
plots <- list()
colnames(umap_original_embeddings) <- c("UMAP_1", "UMAP_2")

for (f in c(2, 3, 4, 7, 10)) {
  loading <- flashier_fit_nn$L_pm[, f]
  p <- plot_loadings_on_umap(umap = umap_original_embeddings, loading = loading, factor_num = f, size = 1)
  plots[[length(plots) + 1]] <- ggplotGrob(p)
}

# Combine the plots in a 2x3 grid
combined_plot <- wrap_plots(plots, ncol = 3, nrow = 2)

# Display the combined plot
print(combined_plot)
```



### Semi-negative EBMF

The loadings from the semi-negative EBMF are more diverse, where each cell is composed of a larger number of factors.
However, there are still both "unique" and "shared" factors across different cell types. For examples:

Specific factors include $k3$, $k12$, $k14$, $k16$, $k19$, $k23$:

- Factor $k3$ is (mostly) unique to Myocytes.

- Factor $k12$ is (mostly) unique to Chondrocytes (also appears in Connective Tissue Progenitors).

- Factor $k14$ is unique to Hepatocytes.

- Factor $k16$ is unique to Cardiac Muscle lineages.

- Factor $k19$ is unique to Choroid Plexus.

- Factor $k23$ is unique to White Blood Cells.


```{r eval=TRUE, message=FALSE, results='hide',fig.keep='all', echo=FALSE}
# Collect all plots in a list
plots <- list()
colnames(umap_original_embeddings) <- c("UMAP_1", "UMAP_2")

for (f in c(3, 12, 14, 16, 19, 23)) {
  loading <- flashier_fit_semi$L_pm[, f]
  p <- plot_loadings_on_umap(umap = umap_original_embeddings, loading = loading, factor_num = f, size = 1)
  plots[[length(plots) + 1]] <- ggplotGrob(p)
}

# Combine the plots in a 2x3 grid
combined_plot <- wrap_plots(plots, ncol = 3, nrow = 2)

# Display the combined plot
print(combined_plot)
```


Shared factors include $k2$, $k4$, $k9$, $k11$:

- Factor $k2$ is shared across Erythroid Progenitors and Hepatocytes.

- Factor $k4$ (as well as $k11$) is shared across multiple types of neurons and glial cells.

- Factor $k9$ is shared across some clusters of the white blood cells and Erythroid cells.


```{r eval=TRUE, message=FALSE, results='hide',fig.keep='all', echo=FALSE}
# Collect all plots in a list
plots <- list()
colnames(umap_original_embeddings) <- c("UMAP_1", "UMAP_2")

for (f in c(2, 4, 9, 11)) {
  loading <- flashier_fit_semi$L_pm[, f]
  p <- plot_loadings_on_umap(umap = umap_original_embeddings, loading = loading, factor_num = f, size = 1)
  plots[[length(plots) + 1]] <- ggplotGrob(p)
}

# Combine the plots in a 2x2 grid
combined_plot <- wrap_plots(plots, ncol = 2, nrow = 2)

# Display the combined plot
print(combined_plot)
```



### FastTopics

The result from FastTopics looks quite similar to the semi-negative EBMF, where each cell is composed of a larger number of factors. 

For examples, some shared factors include:

- Factor $k3$ is shared between Erythroid Lineage and Hepatocytes.

- Factor $k7$ is shared between Peripheral Neuron and Schwann Cells.

- Factors $k10$ is shared between Developing Guts and Epithelial Cells.

- Factor $k13$ is shared between Erythroid Progenitors, Hepatocytes and White blood cells.

- Factor $k15$ is shared between Chondrocytes and Connective Tissue Progenitors.

- Factor $k16$ is shared between Glial cells and Peripheral Neuron.

```{r eval=TRUE, message=FALSE, results='hide', fig.keep='all', echo=FALSE}
# Collect all plots in a list
plots <- list()
colnames(umap_original_embeddings) <- c("UMAP_1", "UMAP_2")

for (f in c(3,7,10,13,15,16)) {
  loading <- fasttopics_fit$L[, f]
  p <- plot_loadings_on_umap(umap = umap_original_embeddings, loading = loading, factor_num = f, size = 1)
  plots[[length(plots) + 1]] <- ggplotGrob(p)
}

# Combine the plots in a 2x3 grid
combined_plot <- wrap_plots(plots, ncol = 3, nrow = 2)

# Display the combined plot
print(combined_plot)
```

For unique factors, we can see:

- Factor $k4$ is (mostly) unique to Myocytes.

- Factor $k5$ is (mostly) unique to Neuron.

- Factor $k8$ is unique to White Blood Cells.

- Factor $k9$ is unique to Endothelial Cells.

- Factor $k14$ is unique to Neuron.

- Factor $k18$ is unique to Cardiac Muscle lineages.

```{r eval=TRUE, message=FALSE, results='hide', fig.keep='all', echo=FALSE}
# Collect all plots in a list
plots <- list()
colnames(umap_original_embeddings) <- c("UMAP_1", "UMAP_2")

for (f in c(4,5,8,9,14,18)) {
  loading <- fasttopics_fit$L[, f]
  p <- plot_loadings_on_umap(umap = umap_original_embeddings, loading = loading, factor_num = f, size = 1)
  plots[[length(plots) + 1]] <- ggplotGrob(p)
}

# Combine the plots in a 2x3 grid
combined_plot <- wrap_plots(plots, ncol = 3, nrow = 2)

# Display the combined plot
print(combined_plot)
```



## Compare these methods

- The non-negative EBMF provides a kind of "cleaner composition" of each cell, where each cell is mostly composed of one base line factor and another factor that is " kind of specific" to the cell type. This in a way demonstrates the advantage of the part-based representation (plus sparsity).

- The semi non-negative EBMF provides a more "diverse" composition of each cell, where each cell is composed of a base line factor and multiple other factors that may appear in different cell types. The compensate of the "diversity" is the computation of this problem tends to be a bit more stable than the non-negative EBMF.

- The FastTopics also provides a kind of diverse composition of each cell, where each cell is composed of multiple factors. This might be due to there is no sparsity constraint in the FastTopics model.

- Overall, they all found some quite interesting structures (shared or unique) that are not obvious from the original UMAP visualization. 


At the same time, non-negative EBMF also has smaller number of "interpretable" factors compared to the semi-negative EBMF and FastTopics. For example, if we count the number of factors that appear (with loading value greater than say 1e-10) at least in $10$ cells, we have:

```{r}
nn_factors <- colSums(abs(flashier_fit_nn$L_pm) > 1e-10) > 10
table(nn_factors)

semi_factors <- colSums(abs(flashier_fit_semi$L_pm) > 1e-10) > 10
table(semi_factors)

ft_factors <- colSums(abs(fasttopics_fit$L) > 1e-10) > 10
table(ft_factors)
```

This is a good thing if we want factors and loadings with very clear interpretation. However, it might also lead to a loss of information for some more subtle structures in the data.


## More detailed exploration of NN-EBMF

```{r}
gene_ids <- rownames(flashier_fit_nn$F_pm)
res <- ldf(flashier_fit_nn, type = "i")
F <- with(res, F %*% diag(D))
# let's make the gene names more readable
# gene_ids <- gsub("\\..*", "", gene_ids)
rownames(flashier_fit_nn$F_pm) <- gene_ids
```

For now, let's focus on the result from the non-negative EBMF. First, let's take a look at the structure plot of each factor from nn-EBMF. To make the visualization more clear, we only show the top 4 genes that contribute the most to each factor.


```{r}
factor_of_interest <- c(2, 3, 9, 13, 15)
top_genes_mat <- apply(F, 2, order, decreasing = TRUE)[1:4, factor_of_interest]
top_genes <- unique(rownames(flashier_fit_nn$F_pm)[top_genes_mat])
plot(flashier_fit_nn,
     plot_type = "heatmap",
     pm_which = "factors",
     pm_subset = top_genes,
     pm_groups = factor(top_genes, levels = rev(top_genes)),
     kset = factor_of_interest,
     gap = 0.2)
```

For the factor `k2` that is shared across Erythroid Progenitors and Hepatocytes, the leading gene is ENSMUSG00000052305, which corresponds to the Hbb-bs gene in Mus musculus (mouse), encoding the hemoglobin, beta adult s chain. 
This gene is a part of the beta-globin cluster and plays a crucial role in oxygen transport from the lungs to peripheral tissues. It is predominantly expressed in tissues involved in hematopoiesis.
It is reasonable that this gene is highly expressed in Erythroid Progenitors, but it is kind of surprising that it is also highly expressed in Hepatocytes. Maybe this implies there exists some annotation error in the cell types of the original data.

For the factor `k3` that is shared across multiple types of neurons and glial cells, the leading gene is ENSMUSG00000072235.
ENSMUSG00000072235, corresponds to Tuba1a, encoding the alpha tubulin protein in Mus musculus (mouse).
This gene is known for its ubiquitous expression, with particularly high levels in brain and lungs.

For the factor `k9` that is unique to Endothelial Cells, the leading gene is ENSMUSG00000031502, which corresponds to the Col4a1 gene in Mus musculus (mouse), encoding the collagen, type IV, alpha 1 protein.
This protein is a crucial component of the basement membrane.

For the factor `k13` that is unique to Choroid Plexus, the leading gene is ENSMUSG00000061808.
It corresponds to the Ttr gene in Mus musculus (mouse), encoding the protein transthyretin, a transport protein primarily involved in the distribution of the thyroid hormone thyroxine and retinol (vitamin A). It is predominantly synthesized in the liver and the choroid plexus of the brain.

For the factor `k15` that is unique to White Blood Cells, the leading gene is ENSMUSG00000049744, which corresponds to the ArhGAP15 gene in Mus musculus (mouse), encoding the Rho GTPase-activating protein 15.
Existing literature has reported that "Knock-out of Arhgap15 function demonstrates that this gene is required to regulate multiple functions in macrophages and neutrophils."








