---
title: Preparation and initial exploration of the "pancreas cytokine" data set
author: Peter Carbonetto
output: workflowr::wflow_html
---

Here we will prepare the single-cell RNA-seq data from
[Stancill et al 2021][stancill2021function] for analysis with
fastTopics and flashier. The data files were obtained by downloading
and extracting tar file GSE183010_RAW.tar from
[GEO accession GSE183010][stancill2021function-geo].

```{r knitr-opts, include=FALSE}
knitr::opts_chunk$set(comment = "#",collapse = TRUE,results = "hold",
                      fig.align = "center",dpi = 120)
```

Load the R packages used to perform the data processing and analysis:

```{r load-pkgs, message=FALSE}
library(data.table)
library(Matrix)
library(tools)
library(rsvd)
library(uwot)
```

Set the seed for reproducibility:

```{r set-seed}
set.seed(1)
```

Import the count data from the "matrix market" format:

```{r load-data}
read_geo_data <- function (i, r, prefix = "GSM000000", dir = ".") {
  infile   <- sprintf("%s_Rep%d_S%d_barcodes.tsv.gz",prefix,r,i)
  barcodes <- fread(file.path(dir,infile),quote = FALSE,header = FALSE,
                    stringsAsFactors = FALSE)
  class(barcodes) <- "data.frame"
  barcodes <- barcodes[,1]
  infile <- sprintf("%s_Rep%d_S%d_features.tsv.gz",prefix,r,i)
  genes  <- fread(file.path(dir,infile),sep = "\t",quote = FALSE,
                  header = FALSE,stringsAsFactors = FALSE)
  class(genes) <- "data.frame"
  names(genes) <- c("ensembl","symbol","type")
  genes        <- transform(genes,type = factor(type))
  infile <- sprintf("%s_Rep%d_S%d_matrix.mtx.gz",prefix,r,i)
  counts <- fread(file.path(dir,infile),quote = FALSE,header = FALSE,
                  skip = 2)
  class(counts) <- "data.frame"
  names(counts) <- c("row","col","value")
  n <- max(counts$row)
  m <- max(counts$col)
  counts <- sparseMatrix(i = counts$row,j = counts$col,x = counts$value,
                         dims = c(n,m))
  rownames(counts) <- genes$ensembl
  colnames(counts) <- barcodes
  return(list(genes  = genes,
			  counts = counts))
}
dat     <- vector("list",8)
dataset <- 0
samples <- NULL
for (r in 1:2) {
  for (i in 1:4) {
    dataset <- dataset + 1
    dat[[dataset]] <-
	  read_geo_data(i,r,prefix = paste0("GSM55486",23 + dataset),
	                dir = "../data/GSE183010")
	samples <- rbind(samples,
	                 data.frame(barcode   = colnames(dat[[dataset]]$counts),
					            mouse     = paste0("S",i),
					            replicate = r,
								stringsAsFactors = FALSE))
  }
}
samples <- transform(samples,
                     mouse     = factor(mouse),
					 replicate = factor(replicate))
features <- Reduce(intersect,lapply(dat,function (x) x$genes$ensembl))
genes    <- subset(dat[[1]]$genes,is.element(ensembl,features))
counts   <- do.call("cbind",lapply(dat,function (x) x$counts[features,]))
counts   <- t(counts)
```

A good fraction of cells have fewer expressed genes. Let's remove them:

```{r filter-genes-per-cell, fig.height=3, fig.width=3}
par(mar = c(4,4,1,1))
x <- rowSums(counts > 0)
i <- which(x > 2000)
samples <- samples[i,]
counts  <- counts[i,]
hist(x,n = 64,main = "",xlab = "number of genes",ylab = "number of cells")
```

A small number of cells have a large proportion of mitochondrial genes.
Let's remove those cells as well.

```{r filter-mito, fig.height=3, fig.width=3}
par(mar = c(4,4,1,1))
mito_genes <- which(substr(genes$symbol,1,2) == "mt")
s          <- rowSums(counts)
s_mito     <- counts[,mito_genes]
prop_mito  <- rowSums(s_mito)/s
i          <- which(prop_mito < 0.1)
samples <- samples[i,]
counts  <- counts[i,]
hist(prop_mito,n = 64,main = "",xlab = "proportion mitochondrial",
     ylab = "number of cells")
```

Finally, a bunch of genes are not expressed in any cells. Let's remove
those as well:

```{r filter-genes}
x      <- colSums(counts > 0)
j      <- which(x > 0)
genes  <- genes[j,]
counts <- counts[,j]
```

Here's an overview of the scRNA-seq data after these data filtering
steps:

```{r data-overview}
nrow(samples)
nrow(genes)
dim(counts)
mean(counts > 0)
```

Now let's generate a 2-d nonlinear embedding of the cells using
t-SNE. First, transform the counts into "shifted log counts":

```{r shifted-log-counts}
a <- 1
s <- s/mean(s)
shifted_log_counts <- MatrixExtra::mapSparse(counts/(a*s),log1p)
```

Then project the cells onto the top 30 PCs:

```{r pca}
set.seed(1)
U <- rsvd(shifted_log_counts,k = 30)$u
```

Now run UMAP on the 30 PCs:

```{r umap}
set.seed(1)
Y <- umap(U,n_neighbors = 30,metric = "cosine",min_dist = 0.3,
          n_threads = 8,verbose = TRUE)
x <- Y[,1]
y <- Y[,2]
samples$umap1 <- x
samples$umap2 <- y
```

Specify the clusters based on the UMAP plot:

```{r find-clusters}
samples$cluster <- "c7"
samples$cluster[x < -3 & y > 5]   <- "c1"
samples$cluster[x < -6 & y < -7]  <- "c2"
samples$cluster[x < -2 & y < -15] <- "c3"
samples$cluster[x > 14] <- "c4"
samples$cluster[sqrt((x - 9)^2 + (y + 8)^2) < 2] <- "c5"
samples$cluster[sqrt((x + 1)^2 + (y + 13)^2) < 3] <- "c6"
samples <- transform(samples,cluster = factor(cluster))
```

UMAP plot:

```{r umap-vs-cell-type, fig.height=3.75, fig.width=5}
p <- ggplot(samples,aes(x = umap1,y = umap2,color = cluster)) +
  geom_point(size = 1.5) +
  theme_cowplot(font_size = 10)
print(p)
```

Save the processed data to an RData file:

```{r save-data, eval=FALSE}
save(list = c("samples","genes","counts"),
     file = "pancreas_cytokine.RData")
resaveRdaFiles("pancreas_cytokine.RData")
```

[stancill2021function]: https://doi.org/10.1093/function/zqab063
[stancill2021function-geo]: https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE183010
