---
title: NMF analysis of the "pancreas cytokine" data set (untreated mouse only)
author: Peter Carbonetto
output: workflowr::wflow_html
---

Here we perform a NMF analyses of the
["pancreas cytokine" data set](pancreas_cytokine.html), focussing on
the scRNA-seq data from untreated mouse only.

```{r knitr-opts, include=FALSE}
knitr::opts_chunk$set(comment = "#",collapse = TRUE,results = "hold",
                      fig.align = "center",dpi = 120)
```

Load packages used to process the data, perform the analyses, and
create the plots.

```{r load-pkgs, message=FALSE}
library(Matrix)
library(fastTopics)
library(flashier)
library(singlecelljamboreeR)
library(ggplot2)
library(cowplot)
```

Set the seed for reproducibility:

```{r set-seed}
set.seed(1)
```

Load the prepared data set:

```{r load-data}
load("../data/pancreas_cytokine.RData")
```

Here we will analyze the cells from the untreated mouse only:

```{r filter-cells-1}
i       <- which(samples$mouse == "S1")
samples <- samples[i,]
counts  <- counts[i,]
```

Remove three cells that appear to be outliers (one of them appears to
be an acinar cell based on Eric's analysis):

```{r filter-cells-2}
outliers <- c("TTTGTTGTCGTTAGTG-1","TTTGTTGGTAGAGCTG-1","CCCAACTCACTCATAG-1")
i        <- which(!is.element(samples$barcode,outliers))
samples  <- samples[i,]
counts   <- counts[i,]
```

Remove genes that are expressed in fewer than 5 cells:

```{r filter-genes}
j      <- which(colSums(counts > 0) > 4)
genes  <- genes[j,]
counts <- counts[,j]
```

This is the dimension of the data set we will analyze:

```{r dim-counts}
dim(counts)
```

For the Gaussian-based analyses (later), we will need the shifted log
counts:

```{r shifted-log-counts}
a <- 1
s <- rowSums(counts)
s <- s/mean(s)
shifted_log_counts <- log1p(counts/(a*s))
rownames(shifted_log_counts) <- NULL
```

## Topic model (fastTopics)

Fit a topic model to the counts (with $K = 13$ topics):

```{r fit-topic-model, cache=TRUE}
set.seed(1)
tm <- fit_poisson_nmf(counts,k = 13,init.method = "random",method = "em",
                      numiter = 40,verbose = "none",
					  control = list(numiter = 4,nc = 8,extrapolate = FALSE))
tm <- fit_poisson_nmf(counts,fit0 = tm,method = "scd",numiter = 40,
                      control = list(numiter = 4,nc = 8,extrapolate = TRUE),
					  verbose = "none")
```

Structure plots comparing the topics to the clusters (some of which
are inferred from the topics):

```{r structure-plot-topic-model, fig.height=4.25, fig.width=6.5, results="hide", message=FALSE}
set.seed(1)
celltype_topics <- paste0("k",c(2,3,5,7:9,13))
other_topics <- paste0("k",c(4,1,6,10:12))
L <- poisson2multinom(tm)$L
clusters <- as.character(samples$cluster)
clusters[clusters == "islet"]                  <- "beta"
clusters[clusters == "beta" & L[,"k3"] > 0.25] <- "alpha"
clusters[clusters == "beta" & L[,"k8"] > 0.25] <- "delta+epsilon"
clusters[clusters == "beta" & L[,"k9"] > 0.25] <- "gamma"
clusters <- factor(clusters,c("beta","alpha","delta+epsilon","gamma","duct",
                              "endothelial-mesenchymal","macrophage"))
i <- c(sample(which(clusters == "beta"),400),
       which(clusters != "beta"))
p1 <- structure_plot(L[i,],grouping = clusters[i],topics = celltype_topics,
                     gap = 10,n = Inf) +
  labs(fill = "")
p2 <- structure_plot(L[i,],grouping = clusters[i],topics = other_topics,
                     gap = 10,n = Inf) +
  labs(fill = "")
plot_grid(p1,p2,nrow = 2,ncol = 1)
```

Select genes in two ways: (a) genes that are different from the mean
expression level, (b) genes that are "distinctive".

```{r de-analysis, cache=TRUE}
set.seed(1)
tm_merged <- merge_topics(poisson2multinom(tm),paste0("k",c(1,2,4,11)))
de_vsnull <- de_analysis(tm_merged,counts,pseudocount = 0.1,lfc.stat = "vsnull",
                         verbose = FALSE,control = list(ns = 1e4,nc = 8))
de_le <- de_analysis(tm_merged,counts,pseudocount = 0.1,lfc.stat = "le",
                     verbose = FALSE,control = list(ns = 1e4,nc = 8))
```

Based on the estimated $\mathbf{F}$, we have the following potential
interpretation of the topics:

+ k5 = $\beta$ cells (*Ins1*, *Ins2*, *Mafa*)
+ k4 = *Ins1-* $\beta$ cells?
+ k6 = $\alpha$ cells (*Gcg*, *Mafb*)
+ k7 = $\delta$, $\gamma$ and $\epsilon$ cells (*Sst*, *Ghrl*)
+ k2 = a non-$\alpha$ islet process
+ k3 = macrophages, endothelial cells, mesenchymal cells and duct cells

```{r annotation-heatmap-topic-model, fig.height=3, fig.width=3, eval=FALSE}
scale_rows <- function (A)
  A / apply(A,1,max)
marker_genes <- c("Ins1","Ins2","Mafa","Gcg","Mafb","Sst","Ghrl",
                  "Ppy","Chga","Iapp","Krt19","Ccr5","Pecam1","Esam",
				  "Col1a1","Ghrl")
j <- match(marker_genes,genes$symbol)
F <- poisson2multinom(tm)$F
F <- F[j,]
F <- scale_rows(F)
rownames(F) <- marker_genes
topics <- paste0("k",c(5,4,6,7,2,3))
annotation_heatmap(F[,topics],select_features = "all",verbose = FALSE)
```

## EBNMF (flashier)

Next fit an NMF to the shifted log counts using flashier, with $K =
9$:

```{r flashier-nmf, cache=TRUE, warning=FALSE, cache=TRUE}
set.seed(1)
n  <- nrow(samples)
x  <- rpois(1e7,1/n)
s1 <- sd(log(x + 1))
fl_nmf <- flash(shifted_log_counts,S = s1,ebnm_fn = ebnm_point_exponential,
                var_type = 2,greedy_Kmax = 9,backfit = FALSE,
				nullcheck = FALSE,verbose = 0)
fl_nmf <- flash_backfit(fl_nmf,extrapolate = FALSE,maxiter = 40,verbose = 0)
fl_nmf <- flash_backfit(fl_nmf,extrapolate = TRUE,maxiter = 80,verbose = 0)
```

(Note that I tried setting `greedy_Kmax > 9`, but this didn't result
in more interesting fitsâ€”the additional factors loaded on only a
single cell.)

Structure plot comparing the factors to the clusters:

```{r structure-plot-flashier-nmf, fig.height=2.25, fig.width=6.5, results="hide", message=FALSE}
set.seed(1)
L <- ldf(fl_nmf,type = "i")$L
colnames(L) <- paste0("k",1:9)
i <- c(sample(which(clusters == "beta"),400),
       which(clusters != "beta"))
p <- structure_plot(L[i,],grouping = clusters[i],topics = 2:9,
                    gap = 10,n = Inf) +
  labs(y = "membership",fill = "")
print(p)
```

Possible interpretation of the factors:

+ k4 = $\beta$ cells (*Ins1*, *Ins2*, *Mafa*)
+ k7 = *Ins1-* $\beta$ cells?
+ k3 = $\alpha$, $\delta$ and $\gamma$ cells
+ k5 = macrophages (*Ccr5*)
+ k2 = endothelial cells (*Pecam1*, *Esam*)
+ k6 = mesenchymal cells (*Col1a1*)

```{r annotation-heatmap-flashier-nmf, fig.height=3, fig.width=3, eval=FALSE}
scale_cols <- function (A) {
  b <- apply(A,2,max)
  return(t(t(A) * b))
}
marker_genes <- c("Ins1","Ins2","Mafa","Gcg","Mafb","Sst","Ghrl",
                  "Ppy","Chga","Iapp","Krt19",
                  "Ccr5","Pecam1","Esam","Col1a1")
j <- match(marker_genes,genes$symbol)
F <- ldf(fl_nmf,type = "i")$F
F <- scale_cols(F)
F <- F[j,]
rownames(F) <- marker_genes
colnames(F) <- paste0("k",1:7)
factors <- paste0("k",c(4,7,3,5,2,6))
annotation_heatmap(F[,factors],select_eatures = "all",verbose = FALSE)
```
