---
title: Preparation and initial exploration of the LSA 2021 pancreas data set
author: Eric Weine and Peter Carbonetto
output: workflowr::wflow_html
---

Here we will prepare the single-cell RNA-seq data from
[Stancill et al 2021][stancill_2021_lsa] for analysis with the topic
model and other matrix factorization methods. The data files were
obtained by downloading and the `GSE156175_RAW.tar` file from
[GEO accession GSE156175][stancill_2021_lsa_geo]. To run this script,
download this file from GEO and extract the contents into a folder
called "GSE156175_RAW" in the data directory.

```{r knitr-opts, include=FALSE}
knitr::opts_chunk$set(comment = "#",collapse = TRUE,results = "hold",
                      fig.align = "center",dpi = 120)
```

Load the R packages used to perform the data processing and initial
analyses:

```{r load-pkgs, message=FALSE}
library(Matrix)
library(tools)
library(readr)
library(cowplot)
library(fastTopics)
library(dplyr)
library(Seurat)
```

Set the seed for reproducibility:

```{r set-seed}
set.seed(1)
```

Import the sample information:

```{r import-barcodes, message=FALSE}
barcodes1<-read_tsv("../data/GSE156175_RAW/GSM5842388_Rep1_S1_barcodes.tsv.gz",
                    col_names = "barcode")
barcodes2 <- read_tsv("../data/GSE156175_RAW/GSM4726017_S2_barcodes.tsv.gz",
                      col_names = "barcode")
barcodes3 <- read_tsv("../data/GSE156175_RAW/GSM4726018_S3_barcodes.tsv.gz",
                      col_names = "barcode")
barcodes4 <- read_tsv("../data/GSE156175_RAW/GSM4726019_S4_barcodes.tsv.gz",
                      col_names = "barcode")
barcodes1$barcode <- substr(barcodes1$barcode,1,nchar(barcodes1$barcode) - 2)
barcodes2$barcode <- substr(barcodes2$barcode,1,nchar(barcodes2$barcode) - 2)
barcodes3$barcode <- substr(barcodes3$barcode,1,nchar(barcodes3$barcode) - 2)
barcodes4$barcode <- substr(barcodes4$barcode,1,nchar(barcodes4$barcode) - 2)
barcodes1$condition <- "Untreated"
barcodes2$condition <- "IL-1B"
barcodes3$condition <- "IFNg"
barcodes4$condition <- "IL-1B_IFNg"
barcodes <- rbind(barcodes1, barcodes2, barcodes3, barcodes4)
barcodes$cell_bc <- paste0(barcodes$barcode, "_", barcodes$condition)
barcodes$condition <- factor(barcodes$condition)
```

Import the gene information:

```{r import-gene-info, message=FALSE}
genes1 <- read_tsv("../data/GSE156175_RAW/GSM5842388_Rep1_S1_features.tsv.gz",
                   col_names = c("ensembl", "symbol", "type"))
genes2 <- read_tsv("../data/GSE156175_RAW/GSM4726017_S2_features.tsv.gz",
                   col_names = c("ensembl", "symbol", "type"))
genes3 <- read_tsv("../data/GSE156175_RAW/GSM4726018_S3_features.tsv.gz",
                   col_names = c("ensembl", "symbol", "type"))
genes4 <- read_tsv("../data/GSE156175_RAW/GSM4726019_S4_features.tsv.gz",
                   col_names = c("ensembl", "symbol", "type"))
genes <- genes1
genes$type <- factor(genes$type)
```

Now import in the read counts:

```{r import-counts, message=FALSE}
counts1 <- t(readMM("../data/GSE156175_RAW/GSM5842388_Rep1_S1_matrix.mtx.gz"))
counts2 <- t(readMM("../data/GSE156175_RAW/GSM4726017_S2_matrix.mtx.gz"))
counts3 <- t(readMM("../data/GSE156175_RAW/GSM4726018_S3_matrix.mtx.gz"))
counts4 <- t(readMM("../data/GSE156175_RAW/GSM4726019_S4_matrix.mtx.gz"))
rownames(counts1) <- paste0(barcodes1$barcode, "_", barcodes1$condition)
rownames(counts2) <- paste0(barcodes2$barcode, "_", barcodes2$condition)
rownames(counts3) <- paste0(barcodes3$barcode, "_", barcodes3$condition)
rownames(counts4) <- paste0(barcodes4$barcode, "_", barcodes4$condition)
colnames(counts1) <- genes1$symbol
colnames(counts2) <- genes2$symbol
colnames(counts3) <- genes3$symbol
colnames(counts4) <- genes4$symbol
counts <- rbind(
  as.matrix(counts1),
  as.matrix(counts2),
  as.matrix(counts3),
  as.matrix(counts4)
)
counts <- as(counts, "CsparseMatrix")
```

Remove genes that are not expressed in any cell:

```{r filter-genes-1}
x      <- colSums(counts)
j      <- which(x > 0)
genes  <- genes[j,]
counts <- counts[,j]
```

Remove cells in terms (i) total UMI count and (ii) very few
expressed genes:

```{r filter-cells}
x        <- rowSums(counts > 0)
i        <- which(x > 2000)
barcodes <- barcodes[i,]
counts   <- counts[i,]
x        <- rowSums(counts)
i        <- which(x <= 60000)
barcodes <- barcodes[i,]
counts   <- counts[i,]
```

PC: The filtering step `x > 2000` seems a bit aggressive. I think we
could have retained some more cells.

Remove cells with high mitochondrial counts:

```{r filter-cells-2}
mito_genes <- which(substr(genes$symbol,1,2) == "mt")
s          <- rowSums(counts)
s_mito     <- rowSums(counts[,mito_genes])
prop_mito  <- s_mito/s
i          <- which(prop_mito < 0.1)
barcodes   <- barcodes[i,]
counts     <- counts[i,]
```

PC: Again, the filtering step `prop_mito < 0.1` seems a bit
aggressive.

Remove *MALAT1*, ribosomal genes, and mitochondrial genes:

```{r filter-genes-2}
j <- which(!(grepl("^mt-",genes$symbol) | grepl("^Rp[sl]",genes$symbol)))
genes  <- genes[j,]
counts <- counts[,j]
j      <- which(genes$symbol != "Malat1")
genes  <- genes[j,]
counts <- counts[,j]
```

Remove other genes that are expressed in only 1 or 2 cells:

```{r filter-genes-3}
x      <- colSums(counts > 0)
j      <- which(x > 2)
genes  <- genes[j,]
counts <- counts[,j]
```

Fit a topic model with $K = 12$.

```{r fit-topic-model, cache=TRUE}
set.seed(1)
tm_fit <- fit_poisson_nmf(X = counts,k = 12,control = list(nc = 7),
                          verbose = "none")
```

Here is the Structure plot:

```{r structure-plot, fig.height=1.75, fig.width=7, results="hide"}
set.seed(1)
structure_plot(tm_fit,n = Inf)
```

Below is a heatmap annotation of the marker genes given in the
Stancill paper:

```{r annotation-heatmap, fig.height=2.5, fig.width=4}
scale_rows <- function (A)
  A / apply(A,1,max)
marker_genes <- c("Ins1","Ins2","Gcg","Sst","Ppy","Krt17","Ccr5",
                  "Col1a1","Cd34","Cpa1")
F <- poisson2multinom(tm_fit)$F
F <- F[marker_genes,]
F <- scale_rows(F)
rownames(F) <- marker_genes
annotation_heatmap(F,select_features = "all",verbose = FALSE)
```

Below are some comments relating my guesses about factor
associations. Many of the beta associated factors are non-specific:

+ k1: ?
+ k2: acinar
+ k3: endothelial/mesenchymal
+ k4: ? (presumably beta)
+ k5: ductal
+ k6: deta
+ k7: ? (presumably beta)
+ k8: delta
+ k9: alpha
+ k10: gamma (PP) 
+ k11: macrophage
+ k12: ? (presumably beta)

Now, I do celltype assignments. I assigned cells to the celltype corresponding to the factor it is most loaded on. Below is the structure plot with cells grouped by
type.

```{r clustering, fig.height=2.5, fig.width=7, results="hide"}
L <- poisson2multinom(tm_fit)$L
L_df <- as.data.frame(L)
L_df <- L_df %>% 
  dplyr::select(
    k2, k3, k5, k6, k7, k8, k9, k10, k11
  )

cluster_map <- c(
  k7 = "Beta",
  k6  = "Beta",
  k2  = "Acinar",
  k3  = "Endothelial/Mesenchymal",
  k5  = "Ductal",
  k8  = "Delta",
  k9  = "Alpha",
  k10 = "Gamma",
  k11 = "Macrophage"
)

L_df <- L_df %>%
  rowwise() %>%
  mutate(
    # 1. Identify which column is maximal in this row
    max_col = names(cluster_map)[which.max(c_across(all_of(names(cluster_map))))],
    # 2. Map that column name to the cluster label
    cluster = cluster_map[[max_col]]
  ) %>%
  ungroup()
structure_plot(tm_fit, grouping = L_df$cluster, gap = 30, n = Inf)
```

Overall, this clustering looks quite good.

Save the processed data to an RData file:

```{r write-outputs}
save(list = c("barcodes","genes","counts"),
     file = "pancreas_cytokine_lsa.RData")
# resaveRdaFiles("pancreas_cytokine.RData")
```

[stancill_2021_lsa]: https://www.life-science-alliance.org/content/4/6/e202000949
[stancill_2021_lsa_geo]: https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE156175
