---
title: Initial exploration of the pancreas data set
author: Peter Carbonetto
output: workflowr::wflow_html
---

"Pancreas" data set featured in in the
[Luecken et al 2022 benchmarking paper][luecken-2022].  Widely used to
benchmark single-cell data integration methods. See Supplementary
Fig. 13, Supplementary Note 3 andSupplementary Data 7 for more info on
this data set.

```{r knitr-opts, include=FALSE}
knitr::opts_chunk$set(comment = "#",collapse = TRUE,results = "hold",
                      fig.align = "center",dpi = 120)
```

First, I downloaded the file `human_pancreas_norm_complexBatch.h5ad`
from [figshare][figshare] and copied it to the "data" subdirectory of
this git repository.

Load the packages needed for this analysis.

```{r load-pkgs, message=FALSE}
library(Matrix)
library(hdf5r)
library(uwot)
library(ggplot2)
library(cowplot)
```

Load the counts and convert to a sparse matrix.

```{r load-data}
dat <- H5File$new("../data/human_pancreas_norm_complexBatch.h5ad",mode = "r")
counts <- dat[["layers"]][["counts"]][,]
counts <- as(counts,"CsparseMatrix")
counts <- t(counts)
sample_info <- data.frame(id          = dat[["obs"]][["_index"]][],
                          tech        = dat[["obs"]][["tech"]][],
						  celltype    = dat[["obs"]][["celltype"]][],
						  size_factor = dat[["obs"]][["size_factors"]][],
                          stringsAsFactors = FALSE)
sample_info <- transform(sample_info,
                         tech     = factor(tech),
                         celltype = factor(celltype))
levels(sample_info$tech)     <- dat[["obs"]][["__categories"]][["tech"]][]
levels(sample_info$celltype) <- dat[["obs"]][["__categories"]][["celltype"]][]
genes <- dat[["var"]][["_index"]][]
rownames(counts) <- sample_info$id
colnames(counts) <- genes
```

```{r examine-data-1}
nrow(counts)
ncol(counts)
mean(counts > 0)
```

Notes:

+ Some entries are UMI counts ("UMI-count-like"), whereas others are not.

+ 16,382 cells.

+ About 18% nonzeros.

+ 13 cell types. Some are very rare.

+ 9 "batches"

+ What to do about the very large counts? I think it is okay.

```{r temp}
normalize_rows <- function (A)
  A / rowSums(A)
X <- normalize_rows(counts)
```

TO DO: Save the data in an .Rdata file for more convenient analysis
with the matrix factorization methods.

```{r save-data}
```

[figshare]: (https://figshare.com/articles/dataset/Benchmarking_atlas-level_data_integration_in_single-cell_genomics_-_integration_task_datasets_Immune_and_pancreas_/12420968)
[luecken-2022]: https://doi.org/10.1038/s41592-021-01336-8
