---
title: Initial exploration of the pancreas data set
author: Peter Carbonetto
output: workflowr::wflow_html
---

The aim of this analysis is to take an initial look at the "pancreas"
data set that was featured in in the
[Luecken et al 2022 benchmarking paper][luecken-2022], and prepare the
data in a convenient form for subsequent analyses in R.

In addition to being featured in the Luecken et al paper, it has been
used in several papers on "data integration" methods for single-cell
data (also known as "batch correction" or "harmonization" methods).
See for example the [MNN paper][mnn-paper]. The Supplementary Note in the Luecken et al paper has additional references.

See Supplementary Fig. 13, Supplementary Note 3 and Supplementary Data
7 of the Luecken et al paper for more information on this data set.

```{r knitr-opts, include=FALSE}
knitr::opts_chunk$set(comment = "#",collapse = TRUE,results = "hold",
                      fig.align = "center",dpi = 120)
```

First, load the packages needed for this analysis. Note that
MatrixExtra is also used in one of the steps below.

```{r load-pkgs, message=FALSE}
library(tools)
library(Matrix)
library(hdf5r)
library(rsvd)
library(Rtsne)
library(ggplot2)
library(cowplot)
library(Seurat)
```

Download the file "human_pancreas_norm_complexBatch.h5ad" from
[figshare][figshare] and copy it to the "data" subdirectory of this
git repository. Then load the count data, and encode them as a sparse
matrix:

```{r load-data}
dat <- H5File$new("../data/human_pancreas_norm_complexBatch.h5ad",mode = "r")
counts <- dat[["layers"]][["counts"]][,]
counts <- t(counts)
counts <- as(counts,"CsparseMatrix")
sample_info <- data.frame(id          = dat[["obs"]][["_index"]][],
                          tech        = dat[["obs"]][["tech"]][],
						  celltype    = dat[["obs"]][["celltype"]][],
						  size_factor = dat[["obs"]][["size_factors"]][],
                          stringsAsFactors = FALSE)
sample_info <- transform(sample_info,
                         tech     = factor(tech),
                         celltype = factor(celltype))
levels(sample_info$tech)     <- dat[["obs"]][["__categories"]][["tech"]][]
levels(sample_info$celltype) <- dat[["obs"]][["__categories"]][["celltype"]][]
genes <- dat[["var"]][["_index"]][]
rownames(counts) <- sample_info$id
colnames(counts) <- genes
```

Note that some of the data are not actually counts, so perhaps calling
this matrix "counts" is a bit misleading. Regardless, in some of our
analyses we will model these data as counts.

Also note that in Luecken et al the counts were log-transformed, but
here we taking the untransformed data.

The matrix has 16,382 rows (cells) and 19,093 columns (genes), and
about 18% of the entries are nonzeros:

```{r examine-data-1}
nrow(counts)
ncol(counts)
mean(counts > 0)
```

The pancreas data are actually a combination of several scRNA-seq data
sets that are from different sequencing technologies or were processed
in different ways:

```{r batches}
table(sample_info$tech)
```

The cells were previously annotated by cell type:

```{r cell-types}
table(sample_info$celltype)
```

Some of the cell types occur in only a very small number of cells.

The "size factors" (here, total counts per cell) vary across a very
wide range:

```{r size-factors, fig.height=2.5, fig.width=3.5}
s <- rowSums(counts)
pdat <- data.frame(log_size_factor = log10(s))
ggplot(pdat,aes(log_size_factor)) +
  geom_histogram(bins = 64,col = "black",fill = "black") +
  labs(x = "log(size factor)") +
  theme_cowplot(font_size = 10)
```

Most genes are expressed in at least one cell:

```{r gene-expression-levels-1}
p <- colSums(counts)/sum(s)
sum(p > 0)
```

The (relative) gene expression levels also vary across a very wide
range:

```{r gene-expression-levels-2, fig.height=2.5, fig.width=3.5}
p <- p[p > 0]
pdat <- data.frame(log_rel_expression_level = log10(p))
ggplot(pdat,aes(log_rel_expression_level)) +
  geom_histogram(bins = 64,col = "black",fill = "black") +
  labs(x = "log-expression level (relative)") +
  theme_cowplot(font_size = 10)
```

Let's now generate a 2-d nonlinear embedding of the cells using
*t*-SNE. First, transform the counts into "shifted log counts":

```{r shifted-log-counts}
a <- 1
s <- rowSums(counts)
s <- s/mean(s)
Y <- MatrixExtra::mapSparse(counts/(a*s),log1p)
```

Next, project the cells onto the top 50 PCs:

```{r pca}
set.seed(1)
U <- rsvd(Y,k = 50)$u
```

Now run t-SNE on the 50 PCs:

```{r embeddings, results="hide"}
tsne <- Rtsne(U,dims = 2,perplexity = 100,pca = FALSE,
              num_threads = 8,verbose = TRUE)
sample_info$tsne1 <- tsne$Y[,1]
sample_info$tsne2 <- tsne$Y[,2]
```

*t*-SNE with cells colored by cell-type:

```{r umap-vs-cell-type, fig.height=3.75, fig.width=5}
tsne_colors <- rep(c("#E69F00","#56B4E9","#009E73","#F0E442",
                     "#0072B2","#D55E00","#CC79A7"),times = 2)
tsne_shapes <- rep(c(19,17),each = 7)
ggplot(sample_info,aes(x = tsne1,y = tsne2,color = celltype,
                       shape = celltype)) +
  geom_point(size = 1.5) +
  scale_color_manual(values = tsne_colors) +
  scale_shape_manual(values = tsne_shapes) +
  labs(x = "tSNE 1",y = "tSNE 2") + 
  theme_cowplot(font_size = 10)
```

*t*-SNE with cells colored by batch:

```{r umap-vs-batch, fig.height=3.75, fig.width=4.75}
ggplot(sample_info,aes(x = tsne1,y = tsne2,color = tech,shape = tech)) +
  geom_point(size = 1.5) +
  scale_color_manual(values = tsne_colors) +
  scale_shape_manual(values = tsne_shapes) +
  labs(x = "tSNE 1",y = "tSNE 2") + 
  theme_cowplot(font_size = 10)
```

Both batch and cell-type contribute to structure in the data.

Seurat:

```{r seurat-1}
pancreas <- CreateSeuratObject(counts = t(counts),project = "pancreas",
                               meta.data = sample_info)
pancreas <- NormalizeData(pancreas)
pancreas <- ScaleData(pancreas)
pancreas <- FindVariableFeatures(pancreas)
pancreas <- RunPCA(pancreas,npcs = 50,features = VariableFeatures(pancreas))
pancreas <- RunTSNE(pancreas)
```

More Seurat:

```{r seurat-2, fig.height=3.75, fig.width=5}
DimPlot(pancreas,reduction = "tsne",group.by = "celltype") +
  scale_color_manual(values = tsne_colors) +
  scale_shape_manual(values = tsne_shapes)
```

```{r seurat-3, fig.height=3.75, fig.width=5}
DimPlot(pancreas,reduction = "tsne",group.by = "tech") +
  scale_color_manual(values = tsne_colors) +
  scale_shape_manual(values = tsne_shapes)
```

Finally, save the data and t-SNE results to an .Rdata file for more
convenient analysis in R:

```{r save-data, eval=FALSE}
save(list = c("sample_info","counts"),file = "pancreas.RData")
resaveRdaFiles("pancreas.RData")
```

[figshare]: (https://figshare.com/articles/dataset/Benchmarking_atlas-level_data_integration_in_single-cell_genomics_-_integration_task_datasets_Immune_and_pancreas_/12420968)
[luecken-2022]: https://doi.org/10.1038/s41592-021-01336-8
[mnn-paper]: https://doi.org/10.1038/nbt.4091
