---
title: Initial exploration of the pancreas data set
author: Peter Carbonetto
output: workflowr::wflow_html
---

The aim of this analysis is to take an initial look at the "pancreas"
data set that was featured in in the
[Luecken et al 2022 benchmarking paper][luecken-2022], and prepare the
data in a convenient form for subsequent analyses in R.

In addition to being featured in the Luecken et al paper, it has been
used in several papers on "data integration" methods for single-cell
data (also known as "batch correction" or "harmonization" methods).
See for example the [MNN paper][mnn-paper]. The Supplementary Note in the Luecken et al paper has additional references.

See Supplementary Fig. 13, Supplementary Note 3 and Supplementary Data
7 of the Luecken et al paper for more information on this data set.

```{r knitr-opts, include=FALSE}
knitr::opts_chunk$set(comment = "#",collapse = TRUE,results = "hold",
                      fig.align = "center",dpi = 120)
```

First, load the packages needed for this analysis.

```{r load-pkgs, message=FALSE}
library(tools)
library(Matrix)
library(MatrixExtra)
library(hdf5r)
library(uwot)
library(ggplot2)
library(cowplot)
```

Download the file "human_pancreas_norm_complexBatch.h5ad" from
[figshare][figshare] and copy it to the "data" subdirectory of this
git repository. Then load the count data, and encode them as a sparse
matrix:

```{r load-data}
dat <- H5File$new("../data/human_pancreas_norm_complexBatch.h5ad",mode = "r")
counts <- dat[["layers"]][["counts"]][,]
counts <- t(counts)
counts <- as(counts,"CsparseMatrix")
sample_info <- data.frame(id          = dat[["obs"]][["_index"]][],
                          tech        = dat[["obs"]][["tech"]][],
						  celltype    = dat[["obs"]][["celltype"]][],
						  size_factor = dat[["obs"]][["size_factors"]][],
                          stringsAsFactors = FALSE)
sample_info <- transform(sample_info,
                         tech     = factor(tech),
                         celltype = factor(celltype))
levels(sample_info$tech)     <- dat[["obs"]][["__categories"]][["tech"]][]
levels(sample_info$celltype) <- dat[["obs"]][["__categories"]][["celltype"]][]
genes <- dat[["var"]][["_index"]][]
rownames(counts) <- sample_info$id
colnames(counts) <- genes
```

Note that some of the data are not actually counts, so perhaps calling
this matrix "counts" is a bit misleading. Also note that in Luecken et
al the counts were log-transformed, but here we taking the
untransformed counts.

The matrix has 16,382 rows (cells) and 19,093 columns (genes), and
about 18% of the entries are nonzeros:

```{r examine-data-1}
nrow(counts)
ncol(counts)
mean(counts > 0)
```

The pancreas data are actually several scRNA-seq data sets that were
combined together:

```{r batches}
table(sample_info$tech)
```

The cells were previous annotated by cell type:

```{r cell-types}
table(sample_info$celltype)
```

Some of the cell types occur in only a very small number of cells.

Distribution of the size factors:

```{r size-factors, fig.height=2.5, fig.width=3.5}
s <- rowSums(counts)
ggplot(data.frame(log_size_factor = log10(s)),aes(log_size_factor)) +
  geom_histogram(bins = 64,col = "black",fill = "black") +
  theme_cowplot(font_size = 10)
```

Genes with nonzero expression:

```{r gene-expression-levels-1}
p <- colSums(counts)/sum(s)
sum(p > 0)
```

Distribution of the (relative) expression levels:

```{r gene-expression-levels-2, fig.height=2.5, fig.width=3.5}
p <- p[p > 0]
ggplot(data.frame(log_rel_expression_level = log10(p)),
                  aes(log_rel_expression_level)) +
  geom_histogram(bins = 64,col = "black",fill = "black") +
  theme_cowplot(font_size = 10)
```

Now do UMAP... Compute the shifted log counts:

```{r shifted-log-counts}
a <- 1
s <- rowSums(counts)
s <- s/mean(s)
Y <- mapSparse(counts/(a*s),log1p)
```

UMAP:

```{r umap, eval=FALSE}
Y <- as.matrix(Y)
res <- umap(Y,n_components = 2,n_threads = 8,verbose = TRUE)
```

TO DO: Save the data and UMAP result in an .Rdata file for more
convenient analyses with the matrix factorization methods.

```{r save-data, eval=FALSE}
save(list = c("sample_info","counts"),file = "pancreas.RData")
resaveRdaFiles("pancreas.RData")
```

TO DO: Upload the prepared data file pancreas.RData to Box and add the
[link][box] here.

[figshare]: (https://figshare.com/articles/dataset/Benchmarking_atlas-level_data_integration_in_single-cell_genomics_-_integration_task_datasets_Immune_and_pancreas_/12420968)
[luecken-2022]: https://doi.org/10.1038/s41592-021-01336-8
[mnn-paper]: https://doi.org/10.1038/nbt.4091
[box]: https://uchicago.box.com/s/l0emrjcqpw5yat1zaygtr1v3akq7de1t
